<html>
  <head>
    <!-- for datatables -->
    <link rel="stylesheet" type="text/css" href="https://interactivereport.github.io/cellxgene_VIP/static/DataTables/datatables.min.css">
    <script src="https://interactivereport.github.io/cellxgene_VIP/static/jquery-ui.min.js"></script>
    <!-- for d3 volcano -->
    <script src="https://d3js.org/d3.v4.min.js"></script>
    <script src="static/d3plot/volcanoPlot.js"></script>
    <link rel="stylesheet" href="static/d3plot/volcanoPlot.css">
    <!-- for gene sets -->
    <link href="https://bxaf.net/genesets/bootstrap-4/css/bootstrap.min.css" rel="stylesheet" type="text/css">
	  <script src="https://bxaf.net/genesets/bootstrap-4/js/bootstrap.min.js"></script>
  	<script>
		  var GENESET_ACTION_URL = "https://bxaf.net/genesets/genesets.php";
    </script>
	  <link href="https://bxaf.net/genesets/genesets.css" rel="stylesheet">
	  <script src="https://bxaf.net/genesets/genesets.js"></script>
    <!-- end -->
    <style>
      button.link{background:none;border:none;}
      #sortable { list-style-type: none; margin: 0; padding: 0; width: 60%; }
      #sortable li { margin: 0 3px 3px 3px; padding: 0.4em; padding-left: 1.5em; font-size: 1.4em; height: 18px; cursor:move; }
      #sortable li span { position: absolute; margin-left: -1.3em; }
      #sortable li.fixed{cursor:default; color:#959595; opacity:0.5;}

      .splitter {
          display: flex;
          width: 100%;
      }
      
      #separator {  
          cursor: col-resize;
          min-width: 6px;
          background-color: #aaa;
          background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAhCAQAAABOpSL+AAAAIklEQVR4AWMwbb/PdR+JZDD9f1/oPhI5sgVGBSruc9xHIgGdSQqqQJGkRgAAAABJRU5ErkJggg==') center center no-repeat #F1F1F1;
      /* prevent browser's built-in drag from interfering */
          -moz-user-select: none;
          -ms-user-select: none;
          user-select: none;
      }
      
      #panel-left {
          background-color: #dde;
          width: 20%;
          white-space: nowrap;
      }
      
      #panel-right {
          background-color: #fff;
          width: 80%;
      }
      #separator, #panel-left, #panel-right {
          vertical-align: top; 
      }
      #STACBARplot {
          margin: 40px;
          position: relative;
      }
      .axis path,
      .axis line {
        fill: none;
        stroke: #000;
        shape-rendering: crispEdges;
      }
      .fly-in {
        font-size: 8px;
  
      }
      .axis .tick line {
        stroke: #ccc;
        /*opacity: 0.5;*/
        pointer-events: none;
      }
      /*.axis .minor  line{*/
          /*stroke: red;*/
      /*}*/
      .highlight {
        font-weight: bold ;
      }
      svg {
        overflow: visible;
      }

      .block {
        display: block;
        width: 100%;
        border: none;
        background-color: #B6B6B6;
        color: white;
        padding: 4px 28px;
        font-size: 18px;
        cursor: pointer;
        text-align: center;
      }
      .block:hover {
        background-color: #ffff33;
        color: black;
      }

      .tab {
        float: left;
        width: 100%;
        height: 100%;
        
        border: 1px solid #ccc;
        background-color: #f1f1f1;
      }
      .tab button {
        display: block;
        width: 100%;
        text-align: left;
        
        background-color: inherit;
        border: none;
        outline: none;
        cursor: pointer;
        padding: 5px 10px;
        transition: 0.3s;
      }
      .tab button:hover {
        background-color: #ddd;
      }
      .tab button.active {
        background-color: #ccc;
      }
      .tabcontent {
        float: left;
        border-left: none;
        width:100%;
        
        display: none;
        padding: 0px 10px;
      }

      h3 {
        display: block;
        font-size: 1.17em;
        margin-top: 1em;
        margin-bottom: 1em;
        margin-left: 0;
        margin-right: 0;
        font-weight: bold;
      }
      .loading {
          width: 50%;
          height: auto;
          position: relative;
      }
      .loading-wheel {
          width: 70px;
          height: 70px;
          margin-top: -60px;
          margin-left: -60px;
          
          position: absolute;
          top: 50%;
          left: 50%;
          
          border-width: 30px;
          border-radius: 50%;
          -webkit-animation: spin 2s linear infinite;
      }
      .style-2 .loading-wheel {
          border-style: double;
          border-color: #99f transparent;
      }
      @-webkit-keyframes spin {
          0% {
              -webkit-transform: rotate(0);
          }
          100% {
              -webkit-transform: rotate(-360deg);
          }
      }
      
      tr:nth-child(even) {background-color: #F1F1F1;}
      input[type=number]{
        width: 60px;
      }
      input,textarea,select{ font-size: 14px; }
      .ace_editor p,.ace_editor div,.ace_editor input,.ace_editor label,.ace_editor span,.ace_editor text {font-family: 'Monaco', 'Menlo', 'Droid Sans Mono', 'Courier New', monospace !important;}
      pre span {font-size: 12px; font-family: 'Arial', monospace !important;} 
      table {
        border-collapse: collapse;
      }
      th {
        background-color: #D4D8DE;
        color: black;
      }
    </style>
  </head>
<body>
<div class="splitter">
  <div id="panel-left">
    <div class="tab">
      <button class="tablinks" onclick="selFun(event, 'ADDG')"><b>Add Genes</b></button>
      <button class="tablinks" onclick="selFun(event, 'SGV');$('#SGVnote').text('');if($('#SGVgene').val() == null) $('#SGVnote').text('Please add genes either from the VIP menu or the main window');">Violin</button>
      <button class="tablinks" onclick="selFun(event, 'PGV')">Stacked Violin</button>
      <button class="tablinks" onclick="selFun(event, 'HEAT')">Heatmap</button>
      <button class="tablinks" onclick="selFun(event, 'EMBED')">UMAP/tSNE</button>
      <button class="tablinks" onclick="selFun(event, 'DOT')">Dot Plot</button>
      <button class="tablinks" onclick="selFun(event, 'TRAK')">Track Plot</button>
      <button class="tablinks" onclick="selFun(event, 'DENS')">Density Plot</button>
      <button class="tablinks" onclick="selFun(event, 'DENS2D')">2D Density</button>
      <button class="tablinks" onclick="selFun(event, 'DUAL')">Dual Genes</button>
      <button class="tablinks" onclick="selFun(event, 'SANK')">Sankey Diagram</button>
      <button class="tablinks" onclick="selFun(event, 'STACBAR')">Stacked Barplot</button>
      <button class="tablinks" onclick="selFun(event, 'GD')">Gene Detected</button>
      <button class="tablinks" onclick="selFun(event, 'DEG')">DEG</button>
      <button class="tablinks" onclick="selFun(event, 'MARK')">Marker Genes</button>
      <button class="tablinks" onclick="selFun(event, 'CLI')">Command Line Interface</button>
      <hr size="1" noshade>
      <button class="tablinks" onclick="selFun(event, 'IMG')">Figure Option</button>
      <button class="tablinks" onclick="selFun(event, 'ABBR')">Comb. & Abbr.</button>
      <button class="tablinks" onclick="save()">Save Session</button>
      <button class="tablinks" onclick="inputClick()">Load Session</button>
      <button class="tablinks" onclick="ABBRall()">Check All Annotations</button>
      <button class="tablinks" onclick="selFun(event, 'BRUSH');">Brush</button>
      <button class="tablinks" onclick="sync();">Sync</button>
    </div>
  </div>
  <div id="separator" ></div>
  <div id="panel-right" >
    	  
    <input id="loadfile" type="file" onchange="load()" style="display:none"/>
    <div id="ABBR" class="tabcontent"> <!-- Abbreviation -->
      <h3>Create a combinatorial annotation</h3>
      <ol type="1">
        <li><button onclick="ABBRclear()">Uncheck All Annotations</button></li>
        <li><p>Select the annotations of interest on the left panel of main window, then <button onclick="ABBRcombine();ABBRall()">Create</button> a new annotation named "Custom_combine" includes the following categories: <p id="ABBRcombine"></p>
		</li>
      </ol>
    
      <hr>
      <h3>Create abbreviations of annotations</h3><!-- <button onclick="ABBRsave()">Save to local</button><input id="ABBRfile" type="file" value="Load from local" onchange="ABBRload()"/>-->
      <p>Choose annotations:</p>
      <p id="ABBRgrp"></p>
      <p id="ABBRtable"></p>
    </div><!--ABBR-->
    
    <div id="ADDG" class="tabcontent">
      <h3>Add genes of interest for plotting</h3>
      <p>Please input a list of genes (comma, semicolon, space or tab delimited):</p>
      <p><textarea id="ADDGinput" rows="5" cols="70"></textarea></p>
      <p><button onclick="ADDGadd()"><b>Add</b></button></p>
      <p>The following genes were added (click on a gene name to remove): <span id="ADDGsel"></span></p>
      <p>The following genes are not in the data: <span id="ADDGdel" style="color:red"></span></p>
      
      <h3>The user-defined gene sets:</h3>
      <p id="ADDGsets"></p>

      <h3>Add gene sets of interest manually</h3>
      <p>Gene set name: <input id="Geneset_Name0" type="text" size=10/>
        Gene names:<input id="Geneset_Genes0" type="text" size=35/>
        <button onclick="ADDGsetAdd(0)"><b>Add</b></button>
      </p>
      
      <h3>Add gene sets from databases</h3>
      
    	<div id="div_geneset1" class="div_geneset">
    		<div class="my-3 dropdown btn-group">
    			<input style="width: 30rem;" id="Geneset_Name1" name="Geneset_Name1" placeholder="Start typing to enter or select a geneset" type="text" class="form-control form-control-sm geneset_name" />
    			<a href="Javascript: void(0);" class="btn btn-sm btn-primary btn_browse_geneset">&#9906;</a>
    			<div id="Dropdown1" class="geneset_dropdown dropdown-menu"></div>
      		<button onclick="ADDGsetAdd(1)"><b>Add</b></button>
    		</div>
    		<input class="geneset_id" type="hidden" id="Geneset_ID1" name="Geneset_ID1" />
    		<textarea class="form-control my-3 geneset_genes" style="width: 30rem;" rows="5" id="Geneset_Genes1" name="Geneset_Genes1" placeholder="Here will list all genes in a geneset"></textarea>
    	</div>      
      
      <span id="ADDGsetError" style="color:red"></span><br>
      
    </div><!--ADDG-->
    
    <div id="GD" class=tabcontent>
      <div id="GDinput">
        <h3>Genes detected in selected groups of cells </h3>
        <p>In group 1 (<span id="GDcellN1" style="color:red;">0</span>) cells and group 2 (<span id="GDcellN2" style="color:red;">0</span>) cells</p>
        <p>Expression cutoff:
        <input id="GDcutoff" type="number" step="0.1" value="1"/></p>
      </div>
      <button class="GDbt" onclick="GDplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('GDresize')">Zoom In</button>
      <button onclick="ZoomOut('GDresize')">Zoom Out</button>
      <button class="GDpngBT" onclick="saveImg('GDimg')" disabled>Save</button>
      <button onclick="hideShow($('#GDinput'),$(this));">Hide Input</button>
      <pre id="GDnote" style="color:red;"></pre>
      <div id="GDspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="GDresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
        <p id="GDfig"></p></div>
    </div><!--GD-->
    
    <div id="MARK" class="tabcontent">
      <div id="MARKinput">
        <h3>Marker genes of annotation categories on selected cells</h3>
        <p>Choose cell set(s):
          <label><input class="MARKcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="MARKcellN1" style="color:red">0</span>)</label>
          <label><input class="MARKcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="MARKcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Select an annotation: 
        <select id="MARKgrp"></select></p>
        <p>By method: 
        <select id="MARKmethod">
          <option value="logreg">logreg</option>
          <option value="t-test">t-test</option>
          <option value="wilcoxon">wilcoxon</option>
          <option value="t-test_overestim_var">t-test_overestim_var</option>
        </select> with the number of genes per group: <input id="MARKn" type="number" step="1" value="10" min="3"/></p>
      </div>
      <button class="MARKbt" onclick="MARKplot()"><b>Find/Plot</b></button>
      <button onclick="ZoomIn('MARKresize')">Zoom In</button>
      <button onclick="ZoomOut('MARKresize')">Zoom Out</button>
      <button class="MARKpngBT" onclick="saveImg('MARKimg')" disabled>Save</button>
      <button onclick="hideShow($('#MARKinput'),$(this));">Hide Input</button>
      <pre id="MARKnote" style="color:red;"></pre>
      <div id="MARKspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      
      <table id="MARKtable" class="display"></table>
      <div id="MARKresize" style="height: 'auto'">
        <p id="MARKfig"></p>
      </div>
    </div><!--MARK-->
    
    <div id="SGV" class="tabcontent">
      <div id="SGVinput">
        <h3>Violin plot of a gene on selected cells</h3>
        <p>Choose cell set(s):
          <label><input class="SGVcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="SGVcellN1" style="color:red">0</span>)</label>
          <label><input class="SGVcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="SGVcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose a gene ('Add Genes' first): <select id="SGVgene"></select></p>
        <p>Expression cutoff: <input id="SGVcutoff" type="number" step="0.1" value="1"/></p>
        <p>Group by:
        <select id="SGVgrp"></select></p>
      </div>
      <button class="SGVbt" onclick="SGVplot('plot')"><b>Plot</b></button>
      <button class="SGVsaveBT" onclick="SGVplot('data')">Get Data</button>
      <button onclick="ZoomIn('SGVresize')">Zoom In</button>
      <button onclick="ZoomOut('SGVresize')">Zoom Out</button>
      <button class="SGVpngBT" onclick="saveImg('SGVimg')" disabled>Save</button>
      <button onclick="hideShow($('#SGVinput'),$(this));">Hide Input</button>
      <pre id="SGVnote" style="color:red;"></pre>
      <div id="SGVspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="SGVresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="SGVfig"></p></div>
    </div><!--SGV-->
    
    <div id="PGV" class="tabcontent">
      <div id="PGVinput">
        <h3>Stacked violin plot of genes on selected cells</h3>
        <p>Choose cell set(s):
          <label><input class="PGVcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="PGVcellN1" style="color:red">0</span>)</label>
          <label><input class="PGVcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="PGVcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.PGVgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="PGVgene"></p>
        <p>Expression cutoff: <input id="PGVcutoff" type="number" step="0.1" value="1"/></p>
        <p>Select user-defined gene sets:</p><p id="PGVgeneGrp"></p>
        <div id="PVGgeneGrpCollapsing">
          <p>Collapsing the above gene sets:
            <input type="radio" name="PGVgeneGrpColl" value="Yes" onclick="$('#PGVgeneGrpCollShow').show()">Yes
            <input type="radio" name="PGVgeneGrpColl" value="No" onclick="$('#PGVgeneGrpCollShow').hide()" checked>No
          </p>
          <div id='PGVgeneGrpCollShow'><p>By:
            <input type="radio" name="PGVgeneGrpCollCal" value="mean" checked>Average
            <input type="radio" name="PGVgeneGrpCollCal" value="median">Median</p>
          </div>
        </div>   
        <p>Group by:<select id="PGVgrp"></select> as
        <select id="PGVgrpXY">
          <option value="Rows">Rows</option>
          <option value="Columns">Columns</option>
        </select>
        </p>
        <p>Color: <input id="PGVcolor_r" type="checkbox">Reverse
        <select id="PGVcolor">
          <option value='Purples'>Purples</option>
          <option value='Blues'>Blues</option>
          <option value='Greens'>Greens</option>
          <option value='Oranges'>Oranges</option>
          <option value='Reds' selected>Reds</option>
          <option value='YlOrBr'>YlOrBr</option>
          <option value='YlOrRd'>YlOrRd</option>
          <option value='OrRd'>OrRd</option>
          <option value='PuRd'>PuRd</option>
          <option value='RdPu'>RdPu</option>
          <option value='BuPu'>BuPu</option>
          <option value='PuBu'>PuBu</option>
          <option value='PuBuGn'>PuBuGn</option>
          <option value='BuGn'>BuGn</option>
          <option value='GnBu'>GnBu</option>
          <option value='YlGnBu'>YlGnBu</option>
          <option value='YlGn'>YlGn</option>
          <option value='Spectral'>Spectral</option>
          <option value='PiYG'>PiYG</option>
          <option value='PRGn'>PRGn</option>
          <option value='BrBG'>BrBG</option>
          <option value='PuOr'>PuOr</option>
          <option value='RdGy'>RdGy</option>
          <option value='RdBu'>RdBu</option>
          <option value='RdYlBu'>RdYlBu</option>
          <option value='RdYlGn'>RdYlGn</option>
        </select> <button onclick="hideShow($('#PGVcolorhelp'),$(this),'Color Help');">Show Color Help</button></p>  
		<div id="PGVcolorhelp" style="display:none;">
          <table border=0>
            <tr valign="top"><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_sequential.png" width="300"></td><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_diverging.png" width="300"></td></tr>
          </table>
		</div>
      </div>

      <button class="PGVbt"  onclick="PGVplot('plot')"><b>Plot</b></button>
      <button class="PGVsaveBT" onclick="PGVplot('data')">Get Data</button>
      <button onclick="ZoomIn('PGVresize')">Zoom In</button>
      <button onclick="ZoomOut('PGVresize')">Zoom Out</button>
      <button class="PGVpngBT" onclick="saveImg('PGVimg')" disabled>Save</button>
      <button onclick="hideShow($('#PGVinput'),$(this));">Hide Input</button>
      <pre id="PGVnote" style="color:red;"></pre>
      <div id="PGVspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="PGVresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="PGVfig"></p></div>
    </div><!--PGV-->
    
    <div id="HEAT" class="tabcontent">
      <div id="HEATinput">
        <h3>Heatmap on selected cells and genes</h3>
        <p>Choose cell set(s):
          <label><input class="HEATcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="HEATcellN1" style="color:red">0</span>)</label>
          <label><input class="HEATcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="HEATcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.HEATgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="HEATgene"></p>
        <p>Cluster by:
        <select id="HEATorder"></select></p>
        <p>Choose annotation:</p>
        <p id="HEATgrp"></p>
        <p>Select plotting value:
        <select id="HEATnorm">
          <option value="X">Expression</option>
          <option value="zscore">Expression Z-score</option>
        </select></p>
      </div>
      <button class="HEATbt"  onclick="HEATplot()"><b>Plot</b></button>
      <button class="HEATsaveBT" onclick="HEATdata()" disabled>Get Data</button>
      <button onclick="ZoomIn('HEATresize')">Zoom In</button>
      <button onclick="ZoomOut('HEATresize')">Zoom Out</button>
      <button class="HEATpngBT" onclick="saveImg('HEATimg')" disabled>Save</button>
      <button onclick="hideShow($('#HEATinput'),$(this));">Hide Input</button>
      <pre id="HEATnote" style="color:red;"></pre>
      <div id="HEATspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="HEATresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="HEATfig"></p></div>
      <div id="HEATdata" style="visibility: hidden;"></div>
    
    </div><!--HEAT-->
    
    <div id="DOT" class="tabcontent">
      <div id="DOTinput">
        <h3>Dot plot shows per group, the fraction of cells expressing a gene (dot size) and the mean expression of the gene in those cell (color scale)</h3>
        <p>Choose cell set(s):
          <label><input class="DOTcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="DOTcellN1" style="color:red">0</span>)</label>
          <label><input class="DOTcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="DOTcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.DOTgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="DOTgene"></p>
        <p>Expression cutoff: <input id="DOTcutoff" type="number" step="0.1" value="1"/></p>
        <p>Expression is averaged only over cells expressing a given gene above the cutoff: <input type="radio" name="mean_only_expressed" value="Yes">Yes<input type="radio" name="mean_only_expressed" value="No" checked>No</p>
  
        <p>Select user-defined gene sets:</p><p id="DOTgeneGrp"></p>
        <div id="DOTgeneGrpCollapsing">
          <p>Collapsing the above gene sets:
            <input type="radio" name="DOTgeneGrpColl" value="Yes" onclick="$('#DOTgeneGrpCollShow').show()">Yes
            <input type="radio" name="DOTgeneGrpColl" value="No" onclick="$('#DOTgeneGrpCollShow').hide()" checked>No
          </p>
          <div id='DOTgeneGrpCollShow'><p>By:
            <input type="radio" name="DOTgeneGrpCollCal" value="mean" checked>Average
            <input type="radio" name="DOTgeneGrpCollCal" value="median">Median</p>
          </div>
        </div>
        
        <p>Group by: <select id="DOTgrp"></select></p>
        <p>Adjust the legend width: <input id="DOTlegendW" type="number" step="0.1" value="1.5"/></p>
        <p>Color: <input id="DOTcolor_r" type="checkbox">Reverse
        <select id="DOTcolor">
          <option value='Purples'>Purples</option>
          <option value='Blues'>Blues</option>
          <option value='Greens'>Greens</option>
          <option value='Oranges'>Oranges</option>
          <option value='Reds' selected>Reds</option>
          <option value='YlOrBr'>YlOrBr</option>
          <option value='YlOrRd'>YlOrRd</option>
          <option value='OrRd'>OrRd</option>
          <option value='PuRd'>PuRd</option>
          <option value='RdPu'>RdPu</option>
          <option value='BuPu'>BuPu</option>
          <option value='PuBu'>PuBu</option>
          <option value='PuBuGn'>PuBuGn</option>
          <option value='BuGn'>BuGn</option>
          <option value='GnBu'>GnBu</option>
          <option value='YlGnBu'>YlGnBu</option>
          <option value='YlGn'>YlGn</option>
          <option value='Spectral'>Spectral</option>
          <option value='PiYG'>PiYG</option>
          <option value='PRGn'>PRGn</option>
          <option value='BrBG'>BrBG</option>
          <option value='PuOr'>PuOr</option>
          <option value='RdGy'>RdGy</option>
          <option value='RdBu'>RdBu</option>
          <option value='RdYlBu'>RdYlBu</option>
          <option value='RdYlGn'>RdYlGn</option>
        </select> <button onclick="hideShow($('#DOTcolorhelp'),$(this),'Color Help');">Show Color Help</button></p>  
		<div id="DOTcolorhelp" style="display:none;">
          <table border=0>
            <tr valign="top"><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_sequential.png" width="300"></td><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_diverging.png" width="300"></td></tr>
          </table>
		</div>
        
        <pre id="DOTnote" style="color:red;">Note: Unscaled data is used as the visualiztion is hard to interpret for scaled or corrected matrices in which zero counts had been replaced by other values. Scaling control in the Figure Option does not apply to Dot plot</pre>
      </div>
      <button onclick="DOTplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('DOTresize')">Zoom In</button>
      <button onclick="ZoomOut('DOTresize')">Zoom Out</button>
      <button class="DOTpngBT" onclick="saveImg('DOTimg')" disabled>Save</button>
      <button onclick="hideShow($('#DOTinput'),$(this));">Hide Input</button>
      <div id="DOTspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="DOTresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="DOTfig"></p></div>
    
    </div><!--DOT-->
    
    <div id="EMBED" class="tabcontent">
      <div id="EMBEDinput">
        <h3>Embedding plots on selected genes and annotations</h3>
        <p>Choose cell set(s):
          <label><input class="EMBEDcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="EMBEDcellN1" style="color:red">0</span>)</label>
          <label><input class="EMBEDcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="EMBEDcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose annotation groups:</p>
        <p id="EMBEDgrp"></p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.EMBEDgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="EMBEDgene"></p>
        <p>Split by: 
        <select id="EMBEDsplit">
          <option value="NONE">NONE</option>
        </select>
        </p>
        <p>Select the embedding layout:
        <select id="EMBEDlayout">
        </select> the number of plots per row:
        <input id="EMBEDslot" type="number" value="4" min="2"/>
        </p>
      </div> 
      <button class="EMBEDbt"  onclick="EMBEDplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('EMBEDresize')">Zoom In</button>
      <button onclick="ZoomOut('EMBEDresize')">Zoom Out</button>
      <button class="EMBEDpngBT" onclick="saveImg('EMBEDimg')" disabled>Save</button>
      <button onclick="hideShow($('#EMBEDinput'),$(this));">Hide Input</button>
      <pre id="EMBEDnote" style="color:red;"></pre>
      <div id="EMBEDspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="EMBEDresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="EMBEDfig"></p></div>
    
    </div><!--EMBED-->
    
    <div  id="TRAK" class="tabcontent">
      <div id="TRAKinput">
        <h3>Track plot shows the gene expression represented by height</h3>
        <p>Choose cell set(s):
          <label><input class="TRAKcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="TRAKcellN1" style="color:red">0</span>)</label>
          <label><input class="TRAKcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="TRAKcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.TRAKgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="TRAKgene"></p>
        <p>Select user-defined gene sets: </p><p id="TRAKgeneGrp"></p>
        <div id="TRAKgeneGrpCollapsing">
          <p>Collapsing the above gene sets:
            <input type="radio" name="TRAKgeneGrpColl" value="Yes" onclick="$('#TRAKgeneGrpCollShow').show()">Yes
            <input type="radio" name="TRAKgeneGrpColl" value="No" onclick="$('#TRAKgeneGrpCollShow').hide()" checked>No
          </p>
          <div id='TRAKgeneGrpCollShow'><p>By:
            <input type="radio" name="TRAKgeneGrpCollCal" value="mean" checked>Average
            <input type="radio" name="TRAKgeneGrpCollCal" value="median">Median</p>
          </div>
        </div>

        <p>Group by: <select id="TRAKgrp"></select></p>
      </div>
      <button onclick="TRAKplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('TRAKresize')">Zoom In</button>
      <button onclick="ZoomOut('TRAKresize')">Zoom Out</button>
      <button class="TRAKpngBT" onclick="saveImg('TRAKimg')" disabled>Save</button>
      <button onclick="hideShow($('#TRAKinput'),$(this));">Hide Input</button>
      <pre id="TRAKnote" style="color:red;"></pre>
      <div id="TRAKspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="TRAKresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="TRAKfig"></p></div>
    
    </div><!--TRAK-->

    <div  id="DENS" class="tabcontent">
      <div id="DENSinput">
        <h3>Density plots of selected genes</h3>
        <p>Choose cell set(s):
          <label><input class="DENScell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="DENScellN1" style="color:red">0</span>)</label>
          <label><input class="DENScell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="DENScellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.DENSgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="DENSgene"></p>
        <p>Split by:
        <select id="DENSsplit"></select></p>
        <p>Color by:
        <select id="DENSgrp"></select></p>
        <p>Bandwidth: <input id="DENSprec" type="number" step="0.1" value="0.5" min="0.1"/> controls how tightly the curve is fit to the data, much like the bin size in a histogram.</p>
      </div>
      <button onclick="DENSplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('DENSresize')">Zoom In</button>
      <button onclick="ZoomOut('DENSresize')">Zoom Out</button>
      <button class="DENSpngBT" onclick="saveImg('DENSimg')" disabled>Save</button>
      <button onclick="hideShow($('#DENSinput'),$(this));">Hide Input</button>
      <pre id="DENSerror" style="color:red;"></pre>
      <div id="DENSspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="DENSresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="DENSfig"></p></div>
    
    </div><!--DENS-->

    <div  id="DENS2D" class="tabcontent">
      <div id="DENS2Dinput">
        <h3>2D density plot to show expression relationship of two genes</h3>
        <p>Choose cell set(s):
          <label><input class="DENS2Dcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="DENS2DcellN1" style="color:red">0</span>)</label>
          <label><input class="DENS2Dcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="DENS2DcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose two genes ('Add Genes' first):</p>
        <p id="DENS2Dgene"></p>
        <p>Expression cutoff: <input id="DENS2Dcutoff" type="number" step="0.1" value="1"/></p>
        <p>The minimal bandwidth: <input id="DENS2Dbandwidth" type="number" step="0.1" value="0.5"/></p>
      </div>
      <button class="DENS2Dbt"  onclick="DENS2Dplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('DENS2Dresize')">Zoom In</button>
      <button onclick="ZoomOut('DENS2Dresize')">Zoom Out</button>
      <button class="DENS2DpngBT" onclick="saveImg('DENS2Dimg')" disabled>Save</button>
      <button onclick="hideShow($('#DENS2Dinput'),$(this));">Hide Input</button>
      <pre id="DENS2Dnote" style="color:red;"></pre>
      <div id="DENS2Dspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="DENS2Dresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="DENS2Dfig"></p></div>
      
    </div><!--DENS2D-->
      
    <div  id="DUAL" class="tabcontent">
      <div id="DUALinput">
        <h3>Embedding plot of two genes on selected cells</h3>
        <p>Choose cell set(s):
          <label><input class="DUALcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="DUALcellN1" style="color:red">0</span>)</label>
          <label><input class="DUALcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="DUALcellN2" style="color:red">0</span>)</label>
        </p>
        <p id="DUALgene"></p>
        <p>Expression cutoff: <input id="DUALcutoff" type="number" step="0.1" value="1"/></p>
        <p>Select the embedding layout:
        <select id="DUALlayout">
        </select>
        </p>
      </div>
      <button class="DUALbt"  onclick="DUALplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('DUALresize')">Zoom In</button>
      <button onclick="ZoomOut('DUALresize')">Zoom Out</button>
      <button class="DUALpngBT" onclick="saveImg('DUALimg')" disabled>Save</button>
      <button onclick="hideShow($('#DUALinput'),$(this));">Hide Input</button>
      <pre id="DUALnote" style="color:red;"></pre>
      <div id="DUALspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="DUALresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="DUALfig"></p></div>
      
    </div><!--DUAL-->

    <div  id="SANK" class="tabcontent">
      <div id="SANKinput">
        <h3>Interactive sankey diagram of annotations and/or genes on selected cells</h3>
        <p>Choose cell set(s):
          <label><input class="SANKcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="SANKcellN1" style="color:red">0</span>)</label>
          <label><input class="SANKcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="SANKcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.SANKgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="SANKgene"></p>
        <!-- <p>Choose user-defined gene sets:</p><p id="SANKgeneGrp"></p> -->
        <p>Number of equally divided gene expression ranges: <input id="sankBin" type="number" step="1" value="5"/></p>
        <p>Choose annotation:</p><p id="SANKgrp"></p>
        <p>Change the order by dragging items up and down:
        <button id="SANKorderBTs" onclick="SANKorderShow()" style="padding:2px;border-radius:45%"><b>&#43;</b></button>
        <button id="SANKorderBTh" onclick="SANKorderHide()" style="padding:2px;border-radius:40%"><b>&#8722;</b></button>
        </p>
        <div id='SANKorderGrp'>
          <ul id="SANKdynList" class="sortable-list"></ul><!--   -->
        </div>
        <p>Please specify the diagram <u>width per category</u>: <input id="SANKw" type="number" step="10" value="250"/> and the <u>diagram height</u>: <input id="SANKh" type="number" step="10" value="700"/></p>
      </div>
      <button class="SANKbt"  onclick="SANKplot()"><b>Plot</b> </button>
      <button onclick="hideShow($('#SANKinput'),$(this));">Hide Input</button>
      <button class="SANKbt" style="float:right;" onclick="for (let item of document.getElementsByClassName('node-label-guide')) {item.setAttribute('style', 'display:none');}; var a = document.createElement('a'); a.href = 'data:application/octet-stream;base64,' + btoa(document.getElementsByClassName('main-svg')[0].outerHTML); a.download = 'Sankey.svg'; a.click();">Save as SVG</button>

      <pre id="SANKnote" style="color:red;"></pre>
      <div id="SANKspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="SANKresize" style="height: 'auto'; width: 'auto'; padding: 0.5em;"></div>
      
    </div><!--SANK-->
    
    <div id="DEG" class="tabcontent">
      <div id="DEGinput">
        <h3>Find differentially expressed genes</h3>
        <p>Group 1 (<span id="DEGcellN1" style="color:red;">0</span>) cells and group 2 (<span id="DEGcellN2" style="color:red;">0</span>) cells</p>
        <p>Differentially expressed genes will be detected between the two cell gorups defined above, unless two categories of the 'Custom_combine' annotation are selected, 
        which enables DE analysis between those two:</p>
        <p id="DEGgrp"></p>
        <p>Select the method: 
        <select id="DEGmethod">
          <option value='default'>Default</option>
          <option value='t-test'>Welchâ€™s t-test</option>
          <option value='rank'>Wilcoxon rank sum</option>
          <option value='wald'>Wald test</option>
        </select></p>
      </div>
      
      <button class="DEGbt" onclick="DEGfind()"><b>Find</b></button>
      <button class="DEGpngBT" onclick="saveImg('DEGimg')" disabled>Save</button>
      <button onclick="hideShow($('#DEGinput'),$(this));">Hide Input</button>
      <pre id="DEGnote" style="color:red;">Note: Unscaled data is used. Scaling control in the Figure Option does not apply to DEG.</pre>
      <label id='DEGinfo'></label>
      <div id="DEGspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
     
      <div id="DEGfinal">
        <p>Remove genes in volcano plot with the absolute log2(FC) larger than: <input id="DEGlogfcCut" type="number" step="1" value="10" min="0"/></p>
        <select id="DEGplotBy">
          <option value='color'>Color by</option>
          <option value='label'>Label by</option>
        </select>
        <input type="radio" name="DEGcolorBy" value="significance" onclick="$('#DEGplotBySig').show();$('#DEGplotByGinfo').hide();$('#DEGplotByGsel').hide();" checked>Significance
        <input type="radio" name="DEGcolorBy" value="geneInfo" onclick="$('#DEGplotBySig').hide();$('#DEGplotByGinfo').show();$('#DEGplotByGsel').hide();">Gene Information
        <input type="radio" name="DEGcolorBy" value="selGene" onclick="$('#DEGplotBySig').hide();$('#DEGplotByGinfo').hide();$('#DEGplotByGsel').show();">Gene Selection
        <div id="DEGplotBySig">
          FDR: <input id="DEGfdr" type="number" step="0.001" value="0.05" min="0" max="1"/>
          Asolute logFC: <input id="DEGlogfc" type="number" step="0.1" value="1" min="0"/>
        </div>
        <div id="DEGplotByGinfo">
          <select id="DEGplotByGinfoSel"></select>
          <input type="radio" name="DEGplotByGinfoRd" value=">">Larger than
          <input type="radio" name="DEGplotByGinfoRd" value="=" checked>Equal to
          <input type="radio" name="DEGplotByGinfoRd" value="<">smaller than
          <input id="DEGplotByGinfoVal" type="text"/>
        </div>
        <div id="DEGplotByGsel">
          <input type="checkbox" onclick="$('.DEGgenes').prop('checked', this.checked);">Uncheck / Check All
          <p id="DEGgene"></p>
        </div>
        <div>
          <button id="DEGfigReset">Reset Axes</button>
          <button onclick="DEGplot('DEG')">Replot</button>
          <p id="DEGfig" style="height: 500px; width:500px; padding: 0.5em; border: 1px solid #B6B6B6"></p>
        </div>
      </div>
      <table id="DEGtable" class="display"></table>
    </div><!--DEG-->
    
    <div id='STACBAR' class='tabcontent'>
      <div id="STACBARinput">
        <h3>Interactive stacked barplot of cell distribution of selected cells over two annotations and/or genes</h3>
        <p>Choose cell set(s):
          <label><input class="STACBARcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="STACBARcellN1" style="color:red">0</span>)</label>
          <label><input class="STACBARcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="STACBARcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose two genes ('Add Genes' first):</p>
        <p id="STACBARgene"></p>
        <div id='STACBARgeneSel'>
          <p>Number of equally divided gene expression ranges: <input id="STACBARbin" type="number" step="1" value="5"/></p>
        </div>
        <p>Annotations:</p><p id="STACBARgrp"></p>
        <p>Color by <select id="STACBARcol"></select></p>
		<p>Figure width: <input id="STACBARwidth" type="number" step="50" value="800"/> , height: <input id="STACBARheight" type="number" step="40" value="480"/></p>
        <p>x-Axis font size: <input id="STACBARxfontsize" type="number" step="2" value="16"/> , 
        label rotate: <input id="STACBARxlabelrotate" type="number" step="5" value="-60"/> ,
		label shift: <input id="STACBARxlabelshift" type="number" step="2" value="-24"/></p>
      </div>
      
      
      <button class="STACBARbt"  onclick="STACBARplot()"><b>Plot</b> </button>
      <button onclick="hideShow($('#STACBARinput'),$(this));">Hide Input</button>
      <button class="STARBARbt" style="float:right;" onclick="var a = document.createElement('a'); a.href = 'data:application/octet-stream;base64,' + btoa(document.getElementById('STACBARsvg').outerHTML); a.download = 'Stackedbar.svg'; a.click();">Save as SVG</button>

      <pre id="STACBARnote" style="color:red;"></pre>
      <div id="STACBARspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id='STACBARresize' style="height: 'auto'; width: 'auto'; padding: 0.5em;"></div>
    </div><!--STACBAR -->
    
    <div id="IMG" class="tabcontent">
      <div><p>Scale data to unit variance for plotting: 
        <input type="radio" name="IMGscale" value="Yes" onclick="IMGset()">Yes
        <input type="radio" name="IMGscale" value="No" onclick="IMGset()">No
        </p>
        <div id="IMGscaleShow">
          <Blockquote>
            <input id="IMGscaleZero" type="checkbox" onclick="IMGset()"/> Zero Center
            <input id="IMGclipValue" type="checkbox" onclick="IMGset()"/> Capped at a maximum value of <input id="IMGscaleMax" type="number" step="0.1" onchange="IMGset()"/> after scaling
          </Blockquote>
        </div>
      </div>
      <p>Figure DPI (Dots Per Inch): 
      <select id="IMGdpi" onchange="IMGset()">
        <option value=150>150</option>
        <option value=300>300</option>
        <option value=600>600</option>
      </select></p>
      <p>Figure format: 
      <select id="IMGimg" onchange="IMGset()">
        <option value='png'>PNG</option>
        <option value='svg'>SVG</option>
      </select></p>
      <p>Font size:
      <input id="IMGfontsize" type="number" step="1" value="11" min="5" onchange="IMGset()"/></p>  
      <p>Vector friendly:
        <input type="radio" name="IMGvector" value="Yes" onclick="IMGset()" checked>Yes
        <input type="radio" name="IMGvector" value="No" onclick="IMGset()">No
      </p>  
      <p>Transparent figure:
        <input type="radio" name="IMGtransparent" value="Yes" onclick="IMGset()">Yes
        <input type="radio" name="IMGtransparent" value="No" onclick="IMGset()" checked>No
      </p>  
      <p>Scanpy plot style:
        <input type="radio" name="IMGscanpybranch" value="master" onclick="IMGset()">master
        <input type="radio" name="IMGscanpybranch" value="split_show" onclick="IMGset()" checked>split_show
      </p>  
      <p>Colormap of UMAP/tSNE, 2D Density (top 4):
      <select id="IMGcolor" onchange="IMGset()">
        <option value='viridis'>viridis</option>
        <option value='plasma'>plasma</option>
        <option value='inferno'>inferno</option>
        <option value='magma'>magma</option>
        <option value='spring'>spring</option>
        <option value='summer'>summer</option>
        <option value='autumn'>Fall</option>
        <option value='winter'>winter</option>
      </select></p>  
      <table border=0>
        <tr><td>viridis</td><td rowspan=8><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_map.png" height="180px"></tr>
        <tr><td>plasma</td></tr>
        <tr><td>inferno</td></tr>
        <tr><td>magma</td></tr>
        <tr><td>spring</td></tr>
        <tr><td>summer</td></tr>
        <tr><td>Fall</td></tr>
        <tr><td>winter</td></tr>
      </table>

    </div><!--IMG-->
    
    <div id="BRUSH" class="tabcontent">
      <h3>Brushed ranges</h3>
      <div id="BRUSHset"></div>
    </div><!--BRUSH-->

    <div id="SYNC" class="tabcontent">
      <pre id="SYNCnote"></pre>
    </div><!--SYNC-->
	  
    <div id="CLI" class="tabcontent">
      <div id="CLIinput">
        <p>Choose cell set(s):
          <label><input class="CLIcell" type="checkbox" value="celllist1" checked/>Group 1 (<span id="CLIcellN1" style="color:red">0</span>)</label>
          <label><input class="CLIcell" type="checkbox" value="celllist2" checked/>Group 2 (<span id="CLIcellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.CLIgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="CLIgene"></p>
        <p>Choose annotation:</p><p id="CLIgrp"></p>
        <p>Choose layout:</p><p id="CLIlayout"></p>
        # Use lines with leading '#' to seperate code for readability as empty lines will create new Jupyter notebook cells<br>
        # adata: a python AnnData object created for you to use in CLI based on your selected cells and metadata <br>
        # strPath: the full file name without suffix. strPath+'.h5ad' is the data file of adata. <br>
        # R code cells should start with "%%R" as shown in Vignette 2. Add "%Rdevice svg" if SVG format is desired and numbers of elements in SVG plots are managable<br>
        <div id="CLIscript" style="height:380px;" ></div>
      </div> 
      <button class="CLIbt"  onclick="CLIplot()"><b>Run</b></button>
      <button class="CLIbt"  onclick="editor.getSession().setValue('')">Clear</button>
      <button class="CLIbt"  onclick="editor.getSession().setValue($('#CLIVignetteCode1').text().trim());">Vignette 1</button>
      <button class="CLIbt"  onclick="editor.getSession().setValue($('#CLIVignetteCode2').text().trim());">Vignette 2</button>
      <button onclick="hideShow($('#CLIinput'),$(this));">Hide Input</button>
      <button onclick="printNotebook();">Print Notebook</button>
      <button onclick="hideShow($('.input'),$(this),'Code');">Hide Code</button>
      <button onclick="hideShow($('.output_text'),$(this),'Output');">Hide Output</button>
      <div id="CLIsvg"></div>
      <pre id="CLInote" style="color:red;"></pre>
      <div id="CLIspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="CLIresize" style="height: 'auto'; width: 'auto'; padding: 0.5em;"></div>
    </div><!--CLI-->
  </div>
	  
</div>

<script type="text/javascript">
var heatCell = 10000;
var editor;
var VIPurl = '';

function printNotebook(){
    var html = $('#CLIresize').html();
    var w=window.open('','_blank');
    w.document.write('<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script\>\n');
    w.document.write('<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script\>\n');
    w.document.write('<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script\>\n');
    w.document.write('<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.css" integrity="sha512-pli9aKq758PMdsqjNA+Au4CJ7ZatLCCXinnlSfv023z4xmzl8s+Jbj2qNR7RI8DsxFp5e8OvbYGDACzKntZE9w==" crossorigin="anonymous" />\n');
    w.document.write
    w.document.write(html); 
    w.document.close();
}

function checkLoading(){
    if(!window.store || !window.store.getState().categoricalSelection || !window.store.getState().annoMatrix._cache.obs.__columnsAccessor){
      return false;
    }
    var cateN = 0;
    for (i=0;i<store.getState().annoMatrix._cache.obs.__columnsAccessor.length;i++) {
        if (store.getState().annoMatrix._cache.obs.__columnsAccessor[i].summarize().categorical) cateN++;
    }
    if(cateN!=Object.keys(window.store.getState().categoricalSelection).length) {
      return false;
    }
    return true;
}
function enableButton(){
    if(!checkLoading()){
      for(const one of document.getElementsByClassName('tablinks')){
        one.disabled=true;
      }
      setTimeout(enableButton, 100);
    }else{
      for(const one of document.getElementsByClassName('tablinks')){
        one.disabled=false;
      }
      if (window.store.getState().annoMatrix.schema.annotations.obsByName[">Description"] !== undefined) {
        if(window.store.getState().annoMatrix.schema.annotations.obsByName[">Description"].categories[0].split("|").length>1){
          var ginit = window.store.getState().annoMatrix.schema.annotations.obsByName[">Description"].categories[0].split("|")[1].split(";");
          window.store.dispatch({type: "set layout choice", layoutChoice: ginit[0].split("by")[1].trim()});
          window.store.dispatch({type: "color by categorical metadata", colorAccessor: ginit[1].split("by")[1].trim()});
          window.store.dispatch({type: "show centroid labels for category", showLabels: true});
        }
      }
      window.store.dispatch({type: "store current cell selection as differential set 1", data:Int32Array.from({length:window.store.getState().annoMatrix.nObs},(v,k)=>k)});
      init();
    }
    //var d = new Date();
    //console.log(d.getTime());
}
enableButton();

function dragElement( element, direction){
    var   md; // remember mouse down info
    const pLeft  = document.getElementById("panel-left");
    const pRight = document.getElementById("panel-right");
    
    element.onmousedown = onMouseDown;
    
    function onMouseDown( e )
    {
    	//console.log("mouse down: " + e.clientX);
    	md = {e,
    	      panelW:pLeft.parentElement.offsetWidth,
    	      lper:parseInt(pLeft.style.width)
    	};
      document.onmousemove = onMouseMove;
      document.onmouseup = () => { 
          document.onmousemove = document.onmouseup = null;
      }
      console.log(md);
    }
    
    function onMouseMove( e )
    {
      var moveR = Math.ceil(100*(e.clientX - md.e.x)/md.panelW);
    	if (direction === "H" ) // Horizontal
    	{
    	    // prevent negative-sized elements
    	    moveR = (moveR < -md.lper) ? 0 : (moveR > (99-md.lper)) ? 0 : moveR;

          element.style.left = md.lper+moveR+"%";
          pLeft.style.width = md.lper+moveR+"%";
          pRight.style.width = (100-(md.lper+moveR)) +"%";
    	}
    }
}
dragElement( document.getElementById("separator"), "H" );

function selFun(evt,sel){
  var i, tabC, tabL;
  tabC = document.getElementsByClassName("tabcontent");
  for (i = 0; i < tabC.length; i++) {
    tabC[i].style.display = "none";
  }
  tabL = document.getElementsByClassName("tablinks");
  for (i = 0; i < tabL.length; i++) {
    tabL[i].className = tabL[i].className.replace(" active", "");
  }
  document.getElementById(sel).style.display = "block";
  evt.currentTarget.className += " active";
  sync();
}
function ZoomIn(strDiv){
  var r = $("#"+strDiv).width() / $("#"+strDiv).parent().width() * 100;
  r += 5;
  $("#"+strDiv).width(r+'%');//Math.min(95,r)
}
function ZoomOut(strDiv){
  var r = $("#"+strDiv).width() / $("#"+strDiv).parent().width() * 100;
  r -= 5;
  $("#"+strDiv).width(Math.max(10,r)+'%');
}
function showImg(eID,data){
  if (['SGVfig','HEATfig'].includes(eID)) {
    document.getElementById(eID).innerHTML = '<img id="'+eID.replace('fig','img')+'" src="data:image/'+window.figOpt['img'].replace('svg','svg+xml')+';base64,'+data+'" width="100%" height="100%"/>';
  } else {
    document.getElementById(eID).innerHTML = '<img id="'+eID.replace('fig','img')+'" src="data:image/'+window.figOpt['img'].replace('svg','svg+xml')+';base64,'+data+'" width="100%" height="auto"/>';
  } 
  
}
function saveImg(pngID){
  var gh = document.getElementById(pngID).src;
  var a = document.createElement('a');
  a.href = gh;
  a.download = pngID+"."+window.figOpt['img'];
  a.target="_blank";
  a.click();
}
function cellNupdate(){
  var n1=0,n2=0;
  if(!!window.store.getState().differential.celllist1){
    n1 = window.store.getState().differential.celllist1.length;
  }
  if(!!window.store.getState().differential.celllist2){
    n2 = window.store.getState().differential.celllist2.length;
  }
  $('#MARKcellN1').text(n1);
  $('#SGVcellN1').text(n1);
  $('#PGVcellN1').text(n1);
  $('#HEATcellN1').text(n1);
  $('#DOTcellN1').text(n1);
  $('#DEGcellN1').text(n1);
  $('#GDcellN1').text(n1);
  $('#EMBEDcellN1').text(n1);
  $('#TRAKcellN1').text(n1);
  $('#DUALcellN1').text(n1);
  $('#DENScellN1').text(n1);
  $('#DENS2DcellN1').text(n1);
  $('#SANKcellN1').text(n1);
  $('#STACBARcellN1').text(n1);
  $('#CLIcellN1').text(n1);

  $('#MARKcellN2').text(n2);
  $('#SGVcellN2').text(n2);
  $('#PGVcellN2').text(n2);
  $('#HEATcellN2').text(n2);
  $('#DEGcellN2').text(n2);
  $('#DOTcellN2').text(n2);
  $('#GDcellN2').text(n2);
  $('#EMBEDcellN2').text(n2);
  $('#TRAKcellN2').text(n2);
  $('#DUALcellN2').text(n2);
  $('#DENScellN2').text(n2);
  $('#DENS2DcellN2').text(n2);
  $('#SANKcellN2').text(n2);
  $('#STACBARcellN2').text(n2);
  $('#CLIcellN2').text(n2);
}

//Page Initial functions
function createOpt(val){
  var opt = document.createElement("option");
  opt.text = val;
  opt.value = val;
  return(opt)
}
function init(){
  if(!('bioInit' in window)){
    //----------------
    $("#panel-left").prop('style')['width'] = "20%";
    $("#separator").prop('style')['left'] = "20%";
    $("#panel-right").prop('style')['width'] = "80%";
    //--------------------------------------------------------------------------
    //ABBRinit
    var grps = Object.keys(window.store.getState().categoricalSelection).sort();
    var grpL = window.store.getState().annoMatrix.schema.annotations.obsByName;
    var htmlCheck= "";
    for(i=0;i<grps.length;i++){
      var one = grps[i];
      if (one.startsWith(">")) continue;
      htmlCheck += '<label><input class="ABBRgrps" type="checkbox" value="'+grps[i]+'" onclick="ABBRlist()"/> '+grps[i]+' ('+grpL[grps[i]].categories.length+')</label>';
    }  
    $('#ABBRgrp').html(htmlCheck);
    
    var bioAbbr = {}
    for(i=0;i<grps.length;i++){
      var one = {}
      for(var j=0;j<grpL[grps[i]].categories.length;j++){
        one[grpL[grps[i]].categories[j]] = grpL[grps[i]].categories[j].toString();
      }
      bioAbbr[grps[i]] = one;
    }
    window.bioAbbr=bioAbbr;
    window.bioCombine={};
    window.bioCombineCell={};
    window.biogen=[];
    window.bioGeneGrp={};
    window.figOpt={scale:'No',scaleZero:'No',clipValue:'Yes',scaleMax:20,dpi:150,img:'png',fontsize:11,vectorFriendly:'Yes',transparent:'No',scanpybranch:'split_show',colorMap:'viridis'};
    if (grpL[">Description"] !== undefined && !grpL[">Description"].categories[0].split(";")[1].includes("scale")) {
      window.figOpt['scale'] = "Yes";
    }

    //--------------------------------------------------------------------------
    //add dropdown annotation categories
    var sgv = document.getElementById("SGVgrp");
    var pgv = document.getElementById("PGVgrp");
    var heat = document.getElementById("HEATorder");
    var dot = document.getElementById("DOTgrp");
    var trak = document.getElementById("TRAKgrp");
    var mark = document.getElementById("MARKgrp");
    var densS = document.getElementById("DENSsplit");
    var densG = document.getElementById("DENSgrp");
    var embed = document.getElementById("EMBEDsplit");

    for(const one of grps){
      if (one.startsWith(">")) continue; 
      sgv.add(createOpt(one));
      pgv.add(createOpt(one));
      heat.add(createOpt(one));
      dot.add(createOpt(one));
      trak.add(createOpt(one));
      mark.add(createOpt(one));
      densS.add(createOpt(one));
      densG.add(createOpt(one));
      embed.add(createOpt(one));
    }
    densS.add(createOpt('None'));
    densG.add(createOpt('None'));
    //--------------------------------------------------------------------------
    //add layout to EMBED, DUAL, and CLI checkbox
    var embedLay = document.getElementById("EMBEDlayout");
    var dualLay = document.getElementById("DUALlayout");
    htmlCheck="";
    for(const ix of Object.keys(window.store.getState().annoMatrix.schema.layout.obs)){
      var one = window.store.getState().annoMatrix.schema.layout.obs[ix].name
      embedLay.add(createOpt(one));
      dualLay.add(createOpt(one));
      htmlCheck += '<label><input class="eIDlayouts" type="checkbox" value="'+one+'"/> '+one+' </label>';
    }
    $('#CLIlayout').html(htmlCheck.replace(/eID/g,'CLI'));
    //--------------------------------------------------------------------------
    htmlCheck= "";
    for(const one of grps){
      if (one.startsWith(">")) continue; 
      htmlCheck += '<label id="eIDgrpsCK'+one+'"><input class="eIDgrps" type="checkbox" value="'+one+'"/> '+one+' ('+grpL[one].categories.length+')</label>';
    }  
    //HEAT, EMBED, SANK, STACBAR, CLI
    $('#HEATgrp').html(htmlCheck.replace(/eID/g,'HEAT'));
    $('#EMBEDgrp').html(htmlCheck.replace(/eID/g,'EMBED'));
    $('#CLIgrp').html(htmlCheck.replace(/eID/g,'CLI'));
    $('#SANKgrp').html(htmlCheck.replace(/eID/g,'SANK').replace(/value/g,'onchange="SANKorderUpdate()" value'));
    $('#STACBARgrp').html(htmlCheck.replace(/eID/g,'STACBAR').replace(/value/g,'onchange="STACBARref()" value'));
    
    
    //SANK
    $("#SANKorderGrp").hide();
    document.getElementById("SANKorderBTh").style.display="none";
    
    $('.sortable-list').sortable({
        connectWith: '.sortable-list',
        placeholder: 'sortable-list-dropholder'
        //console.log("init SANK")
    });
    $('#SGVcutoff').prop('min',-5);
    $('#PGVcutoff').prop('min',-5);  
    $('#GDcutoff').prop('min',-5);
    $('#DUALcutoff').prop('min',-5);
    $('#DOTcutoff').prop('min',-5);  

    //add text on select brush on histogram
    var html = "<table style='border: 1px solid gray; padding: 5px;'><tr><th style='border: 1px solid gray; padding: 5px;'>Name</th><th style='border: 1px solid gray; padding: 5px;'>Low end value</th><th style='border: 1px solid gray; padding: 5px;'>High end value</th><tr>";
    $("div[id^=histogram]").each(function(){
      var aID = $(this).prop('id').replace("histogram_","");
      html += "<tr><td style='border: 1px solid gray; padding: 5px;'>" +aID+"</td>";
      html += "<td style='border: 1px solid gray; padding: 5px;'><input type='number' id='input_"+aID+"_left' step=0.01 style='width: 12ch;border: 0' readonly></td>";
      html += "<td style='border: 1px solid gray; padding: 5px;'><input type='number' id='input_"+aID+"_right' step=0.01 style='width: 12ch;border: 0' readonly></td></tr>";
    });
    
    $("#BRUSHset").html(html);
    
    
    $("div[id^=histogram]").each(function(){initBrush($(this).prop('id'));});
    const unsubscribe = window.store.subscribe(() => {
      if (window.store.getState()["@@undoable/filterState"].prevAction) {
        var action = window.store.getState()["@@undoable/filterState"].prevAction;
        if (action.type.includes("user defined gene success") || action.type.includes("store current cell selection as differential set")) {
          sync();
        }
        if (action.type.includes("continuous metadata histogram")) {
//          $('#SYNCnote').text(JSON.stringify(window.store.getState().continuousSelection));
          $("div[id^=histogram]").each(function(){showBrush($(this).prop('id'));});
        }
      }
    });
    
    window.bioInit=true;

    editor = ace.edit("CLIscript");
    editor.setTheme("ace/theme/iplastic");
    editor.getSession().setMode("ace/mode/python");
    editor.getSession().setUseWrapMode(true);
    editor.setFontSize("14px");
    editor.getSession().setValue($("#CLIVignetteCode1").html().trim());


	document.getElementsByClassName('tablinks')[0].click();

    var observer = new ResizeObserver(function(entries) {
      entries.forEach(function(entry) {
        editor.resize();
      });
    });
    observer.observe(document.getElementById('CLIscript'));
    console.log("VIP (Visualization in Plugin) is ready on client side.");

    var tryurls = [window.location.href.replace(/\/+$/,'')+"/VIP",window.location.origin+"/VIP","/VIP","VIP"];
    var results = [];
    tryurls.reduce(function(prev, cur, index, arr) {
      return prev.then(function(data) {
        return $.ajax({
          type: "POST",
          url: cur,
          data:JSON.stringify({'method':"HELLO"}),
          contentType: 'application/json;charset=UTF-8',
          success: function(res, status, req){
            console.log(res + " url: "+this.url);
            results.push(this.url);
          },
          error: function(request,status,error){
//            console.log(request.status+': handshake failed, keep trying ...');
          }
        }).then(function(data) {
//          console.log("URL " + index);
        }).catch(function(jqXhr) {
        });
      })
    }, $().promise()).done(function() {
      // last ajax call done
      // all results are in the results array
      console.log(results);
      VIPurl = results[0];
    });
  }
  return true;
}
function initBrush(aID){
  for(const i of ["left","right"]){
    var one = d3.select("[id='"+aID+"'] g.brush").append("text").attr("id",aID+"_"+i).attr("y",$('[id="'+aID+'"] rect.selection').prop("y").baseVal.value).attr("fill","#000");
    if(i=="left") one.attr("text-anchor", "end");
  }
}
function showBrush(aID){
  var val = window.store.getState().continuousSelection[aID.replace('histogram','obsAnno')];
  if(val===undefined){
    val = ["",""];
  }else{
    var valMax = Math.max(...val);
    var decimals = (valMax<10) ? 2 : (valMax<100) ? 1:0;
    val = [val[0].toFixed(decimals),val[1].toFixed(decimals)];
  }
  var pos = ['left','right']
  for(i in [0,1]){
    var one = d3.select("[id='"+aID+"_"+pos[i]+"']").attr("x",$('[id="'+aID+'"] rect.selection').prop("x").baseVal.value).text(val[i]);
    if(i==1) one.attr("x",$('[id="'+aID+'"] rect.selection').prop("x").baseVal.value+$('[id="'+aID+'"] rect.selection').prop("width").baseVal.value);
    $('[id="'+aID.replace('histogram','input')+'_'+pos[i]+'"]').val(val[i]);
  }
}
//Page sync functions
function sync(){
  cellNupdate();
  ABBRref();
  ADDGref();
  GDref();
  SGVref();
  PGVref();
  HEATref();
  DOTref();
  TRAKref();
  DUALref();
  DEGref();
  EMBEDref();
  IMGref();
  DENSref();
  DENS2Dref();
  SANKref();
  STACBARref();
  CLIref();
}
function ABBRref(){
}
function ADDGref(){
  var html =  "";
  for(const one of window.biogen){
    html += '<button class="link" style="cursor: crosshair;" id="ADDG_'+one+'" onclick=\'ADDGdel("'+one+'")\'>'+one+'</button>';
  }
  $("#ADDGsel").html(html);
  //$("#ADDGsel").text(window.biogen.join(", "));
  
  var html= "";
  var grpKeys = Object.keys(window.bioGeneGrp);
  for(const one of grpKeys){
    html += "<p><button onclick=\"ADDGsetDel('"+one+"')\"><b>Delete</b></button> <b>"+one+"</b>: " + (""+window.bioGeneGrp[one]).replaceAll(",",", ")+"</p>";
  }
  $("#ADDGsets").html(html);
}
function GDref(){
}
function addCustomCombine(element) {
  var x = document.getElementById(element);
//  var grps = Object.keys(window.store.getState().categoricalSelection).filter(function(e) { return !(e.startsWith('>') || e.startsWith('|')) });
  $("#"+element+" option[value='Custom_combine']").remove();
  var grpKeys = Object.keys(window.bioCombine);
  if(grpKeys.length>0){
    x.add(createOpt("Custom_combine"));
  }
//  if(grpKeys.length>0 && grps.length==x.length){
//    x.add(createOpt("Custom_combine"));
//  }else if(grpKeys.length==0){// && grps.length<x.length
//    $("#"+element+" option[value='Custom_combine']").remove();
    //x.remove(grps.length);
//  }
}
function addCustomCheck(aID){
  $('#'+aID+'grpsCKcustom').remove();
  var grpKeys = Object.keys(window.bioCombine);
  if(grpKeys.length>0){
    var htmlCheck = $('#'+aID+'grp').html();
    htmlCheck += '<label id="'+aID+'grpsCKcustom"><input class="'+aID+'grps" type="checkbox" value="Custom_combine"/> Custom_combine ('+Object.keys(window.bioCombineCell).length+')</label>';
  }
  $('#'+aID+'grp').html(htmlCheck);
}
function MARKref(){
  addCustomCombine("MARKgrp");
}
function SGVref(){
  var genes = window.store.getState().controls.userDefinedGenes;
  genes = genes.concat(window.biogen);
  genes = [...new Set(genes)];
  optG = [];
  var i,x = document.getElementById("SGVgene");
  //remove options which are not in the gene list
  $("#SGVgene option").each(function(){
    optG.push($(this).val());
    if(!genes.includes($(this).val())){
      x.remove($(this).index())
    }
  });
  //add options which are missing from gene list
  if(genes.length>0){
    for(const g of genes){
      if(!optG.includes(g)){
        x.add(createOpt(g));
      }
    }
  }
  addCustomCombine("SGVgrp");
  if($('#SGVgene').val()) $('#SGVnote').text('');
}
function showGeneSet(aID){
  var gSets = Object.keys(window.bioGeneGrp);
  var optG = [], NcheckG=[];
  //remove checkbox which are not in the gene list
  $("."+aID+"sets").each(function(){
    optG.push($(this).val());
    if(!gSets.includes($(this).val())){
      $('#'+aID+'ck'+$(this).val()).remove();
    }else if($(this).prop("checked") == false){
      NcheckG.push($(this).val());
    }
  })
  //add options which are missing from gene list
  if(gSets.length>0){
    var htmlCheck = $('#'+aID+'geneGrp').html();
    for(const g of gSets){
      if(!optG.includes(g)){
        htmlCheck += '<label id="'+aID+'ck'+g+'"><input class="'+aID+'sets" type="checkbox" value="'+g+'" checked/>'+g+'</label>';// '+g+'
      }
    }
    $('#'+aID+'geneGrp').html(htmlCheck);
  }
  $("."+aID+"sets").each(function(){
    if(NcheckG.includes($(this).val())){
      $(this).prop("checked",false);
    }
  })
  
}
function showGene(aID, addInfo='', ckDefault='checked'){
  //console.log(aID+":"+ckDefault+":"+addInfo);
  var genes = window.store.getState().controls.userDefinedGenes;
  genes = genes.concat(window.biogen);
  genes = [...new Set(genes)];
  var optG = [], NcheckG=[],checkG=[];
  //remove checkbox which are not in the gene list
  $("."+aID+"genes").each(function(){
    optG.push($(this).val());
    if(!genes.includes($(this).val())){
      $('#'+aID+'ck'+$(this).val()).remove();
    }else if($(this).prop("checked") == false){
      NcheckG.push($(this).val());
    }else if($(this).prop("checked")){
      checkG.push($(this).val());
    }
  })
  //add options which are missing from gene list
  if(genes.length>0){
    var htmlCheck = $('#'+aID+'gene').html();
    for(const g of genes){
      if(!optG.includes(g)){
        htmlCheck += '<label id="'+aID+'ck'+g+'"><input class="'+aID+'genes" type="checkbox" value="'+g+'" '+ckDefault+' '+addInfo+'/>'+g+'</label>';// '+g+'
      }
    }
    $('#'+aID+'gene').html(htmlCheck);
  }
  $("."+aID+"genes").each(function(){
    if(NcheckG.includes($(this).val())){
      $(this).prop("checked",false);
    }else if(checkG.includes($(this).val())){
      $(this).prop("checked",true);
    }
  })
}
function showGeneGrpCollSet(aID){
  if(Object.keys(window.bioGeneGrp).length>0){
    $('#'+aID+'geneGrpCollapsing').show();
  }else{
    $('#'+aID+'geneGrpCollapsing').hide();
  }
  if($("input[name='"+aID+"geneGrpColl']:checked").val() == 'Yes'){
    $('#'+aID+'geneGrpCollShow').show();
  }else{
    $('#'+aID+'geneGrpCollShow').hide();
  }  
}
function PGVref(){
  showGene('PGV');
  showGeneSet('PGV');
  addCustomCombine("PGVgrp");
  radioCheckTriger("PGVgeneGrpColl");
}
function HEATref(){
  showGene('HEAT');

  var nCells = 0;
  var x = document.getElementsByClassName('HEATcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked && !!window.store.getState().differential[x[i].value]){
      nCells += window.store.getState().differential[x[i].value].length;
    }
  }
  var gL = Object.keys(window.store.getState().categoricalSelection).length-1; //not considering the last one ">Description"
  addCustomCombine("HEATorder");
  $("#HEATorder option[value='Expression']").remove();
  x = document.getElementById("HEATorder");
  if(nCells<heatCell){
    x.add(createOpt("Expression"));
  }
  addCustomCheck("HEAT");
}
function DOTref(){
  showGene('DOT');
  showGeneSet('DOT');
  addCustomCombine("DOTgrp");
  showGeneGrpCollSet("DOT");
}
function TRAKref(){
  showGene('TRAK');
  showGeneSet('TRAK');
  addCustomCombine("TRAKgrp");
  showGeneGrpCollSet("TRAK");
}
function DENSref(){
  var genes = window.store.getState().controls.userDefinedGenes;
  genes = genes.concat(window.biogen);
  genes = [...new Set(genes)];
  var optG = [], NcheckG=[];
  //remove checkbox which are not in the gene list
  $(".DENSgenes").each(function(){
    optG.push($(this).val());
    if(!genes.includes($(this).val())){
      $('#DENSck'+$(this).val()).remove();
    }else if($(this).prop("checked") == false){
      NcheckG.push($(this).val());
    }
  })
  //add options which are missing from gene list
  if(genes.length>0){
    var htmlCheck = $('#DENSgene').html();
    for(const g of genes){
      if(!optG.includes(g)){
        htmlCheck += '<label id="DENSck'+g+'"><input class="DENSgenes" type="checkbox" value="'+g+'" checked/>'+g+'</label>';//
      }
    }
    $('#DENSgene').html(htmlCheck);
  }
  $(".DENSgenes").each(function(){
    if(NcheckG.includes($(this).val())){
      $(this).prop("checked",false);
    }
  })
  addCustomCombine("DENSgrp");
  addCustomCombine("DENSsplit");
}
function DENS2Dref(){
  var genes = window.store.getState().controls.userDefinedGenes;
  genes = genes.concat(window.biogen);
  genes = [...new Set(genes)];
  var optG = [], checkG=[];
  //remove checkbox which are not in the gene list
  $(".DENS2Dgenes").each(function(){
    optG.push($(this).val());
    if(!genes.includes($(this).val())){
      $('#DENS2Dck'+$(this).val()).remove();
    }else if($(this).prop("checked")){
      checkG.push($(this).val());
    }
  })
  //add options which are missing from gene list
  if(genes.length>0){
    var htmlCheck = $('#DENS2Dgene').html();
    for(const g of genes){
      if(!optG.includes(g)){
        htmlCheck += '<label id="DENS2Dck'+g+'"><input class="DENS2Dgenes" type="checkbox" value="'+g+'" onclick="DENS2Dref()"/>'+g+'</label>';// '+g+'
      }
    }
    $('#DENS2Dgene').html(htmlCheck);
  }
  var iEnabled=0;
  $(".DENS2Dgenes").each(function(){
    if(checkG.includes($(this).val())){
      $(this).prop("checked",true);
      iEnabled++;
    }
  })
  if(iEnabled>1){
    $(".DENS2Dgenes").each(function(){
      if($(this).prop("checked")==false){
        $(this).prop("disabled",true);
      }
    })
  }else{
    $('.DENS2Dgenes').prop("disabled",false);
  }
  
}
function EMBEDref(){
  showGene('EMBED');
}
function DUALref(){
  var genes = window.store.getState().controls.userDefinedGenes;
  genes = genes.concat(window.biogen);
  genes = [...new Set(genes)];
  var optG = [], checkG=[];
  //remove checkbox which are not in the gene list
  $(".DUALgenes").each(function(){
    optG.push($(this).val());
    if(!genes.includes($(this).val())){
      $('#DUALck'+$(this).val()).remove();
    }else if($(this).prop("checked")){
      checkG.push($(this).val());
    }
  })
  //add options which are missing from gene list
  if(genes.length>0){
    var htmlCheck = $('#DUALgene').html();
    for(const g of genes){
      if(!optG.includes(g)){
        htmlCheck += '<label id="DUALck'+g+'"><input class="DUALgenes" type="checkbox" value="'+g+'" onclick="DUALref()"/>'+g+'</label>';// '+g+'
      }
    }
    $('#DUALgene').html(htmlCheck);
  }
  var iEnabled=0;
  $(".DUALgenes").each(function(){
    if(checkG.includes($(this).val())){
      $(this).prop("checked",true);
      iEnabled++;
    }
  })
  if(iEnabled>1){
    $(".DUALgenes:checkbox:not(:checked)").prop("disabled",true);
  }else{
    $('.DUALgenes').prop("disabled",false);
  }
  
}
function SANKref(){
  showGene('SANK','onchange="SANKorderUpdate()"');
  //showGeneSet('SANK');
  SANKorderUpdate();
}
function SANKorderUpdate(){
  var selSank = [];
  $(".SANKgrps").each(function(){
    if($(this).prop("checked")){
      selSank.push($(this).val());
  }})
  selSank = selSank.concat(getGenes('SANK'));
  selSank=selSank.map(function(one){return 'SANKli'+one;})

  for(const one of $("#SANKdynList").sortable('toArray')){
    if(!selSank.includes(one)){
      $(document.getElementById(one)).remove();
    }
  }
  selSank = [...new Set($("#SANKdynList").sortable('toArray').concat(selSank))];

  var htmlCheck = "";
  for(const one of selSank){
    htmlCheck += "<li id='"+one+"'><label>"+one.replace("SANKli","")+"</label></li>";
  }
  $("#SANKdynList").html(htmlCheck);
}
function SANKorderShow(){
  $("#SANKorderGrp").show();
	document.getElementById("SANKorderBTs").style.display="none";
  document.getElementById("SANKorderBTh").style.display="inline";
}
function SANKorderHide(){
  $("#SANKorderGrp").hide();
	document.getElementById("SANKorderBTh").style.display="none";
  document.getElementById("SANKorderBTs").style.display="inline";
}
function STACBARref(){
  showGene('STACBAR','onchange="STACBARref()"','');
  var iEnabled=0;
  $("#STACBARcol").empty();
  var colSel = document.getElementById("STACBARcol");
  $(".STACBARgenes").each(function(){
    if($(this).prop("checked")){
      iEnabled++;
      colSel.add(createOpt($(this).val()));
    }
  })
  $("#STACBARgeneSel").hide();
  if(iEnabled>0){
    $("#STACBARgeneSel").show();
  }
  $(".STACBARgrps").each(function(){
    if($(this).prop("checked")){
      iEnabled++;
      colSel.add(createOpt($(this).val()));
    }
  })
  if(iEnabled>1){
    $(".STACBARgenes:checkbox:not(:checked)").prop("disabled",true);
    $(".STACBARgrps:checkbox:not(:checked)").prop("disabled",true);
  }else{
    $(".STACBARgenes").prop("disabled",false);
    $(".STACBARgrps").prop("disabled",false);
  }

}
function radioCheckTriger(eName){
  $("input[name='"+eName+"']:checked").trigger("click");
}
function DEGref(){
  showGene('DEG',"","");
  
  var cName=[],cVal=Object.keys(window.bioCombineCell);
  for(const one of Object.values(window.bioCombineCell)){
    cName.push(one.cName);
  }
  
  var optG = [], checkG=[];
  //remove checkbox which are not in the annotation list
  $(".DEGgrps").each(function(){
    optG.push($(this).val());
    if(!cVal.includes($(this).val())){
      $('#DEGck'+$(this).val()).remove();
    }else if($(this).prop("checked")){
      checkG.push($(this).val());
    }
  })
  //add options which are missing from annotation list
  if(cVal.length>0){
    var htmlCheck = $('#DEGgrp').html();
    for(var i=0;i<cVal.length;i++){
      if(!optG.includes(cVal[i])){
        htmlCheck += '<label id="DEGck'+cVal[i]+'"><input class="DEGgrps" type="checkbox" value="'+cVal[i]+'" onclick="DEGref()"/>'+cName[i]+'</label>';
      }
    }
    $('#DEGgrp').html(htmlCheck);
  }

  //check the originally checked
  var iEnabled=0;
  $(".DEGgrps").each(function(){
    if(checkG.includes($(this).val())){
      $(this).prop("checked",true);
      iEnabled++;
    }
  })
  //disable when 2 are checked or enable when less than 2 is checked
  if(iEnabled>1){
    $(".DEGgrps").each(function(){
      if($(this).prop("checked")==false){
        $(this).prop("disabled",true);
      }
    })
  }else{
    $('.DEGgrps').prop("disabled",false);
  }
  radioCheckTriger("DEGcolorBy");
  if($("#DEGfig").html().length==0) $("#DEGfinal").hide();
}
function IMGref(){
  $('#IMGscaleZero').prop('checked',(window.figOpt['scaleZero']=="Yes"));
  $('#IMGclipValue').prop('checked',(window.figOpt['clipValue']=="Yes"));
  $('#IMGscaleMax').val(window.figOpt['scaleMax']);
  if(window.figOpt['scale'] == "Yes"){
    $('#IMGscaleShow').show();
    $("input[name='IMGscale'][value='Yes']").prop('checked', true);
  }else{
    $('#IMGscaleShow').hide();
    $("input[name='IMGscale'][value='No']").prop('checked', true);
  }

  $('#IMGdpi').val(window.figOpt['dpi']);
  $('#IMGimg').val(window.figOpt['img']);
  $('#IMGfontsize').val(window.figOpt['fontsize']);
  $('#IMGvector').val(window.figOpt['vectorFriendly']);
  $('#IMGtransparent').val(window.figOpt['transparent']);
  $('#IMGscanpybranch').val(window.figOpt['scanpybranch']);
  $('#IMGcolor').val(window.figOpt['colorMap']);
}
function IMGset(){
  window.figOpt['scale'] = $("input[name='IMGscale']:checked").val();
  if(window.figOpt['scale'] == "Yes"){
    $('#IMGscaleShow').show();
    window.figOpt['scaleZero'] = $('#IMGscaleZero').prop('checked')?"Yes":"No";
    window.figOpt['clipValue'] = $('#IMGclipValue').prop('checked')?"Yes":"No";
    window.figOpt['scaleMax'] = $('#IMGscaleMax').val();
  }else{
    $('#IMGscaleShow').hide();
  }

  window.figOpt['dpi'] = $('#IMGdpi').val();
  window.figOpt['img'] = $('#IMGimg').val();
  window.figOpt['fontsize'] = Math.max($('#IMGfontsize').val(),$('#IMGfontsize').prop('min'));
  window.figOpt['vectorFriendly'] = $("input[name='IMGvector']:checked").val();
  window.figOpt['transparent'] = $("input[name='IMGtransparent']:checked").val();
  window.figOpt['scanpybranch'] = $("input[name='IMGscanpybranch']:checked").val();
  window.figOpt['colorMap'] = $('#IMGcolor').val();
  //IMGref();
}
function CLIref(){
  showGene('CLI');
  $(".CLIgrps:checkbox[value='cell_type']").prop("checked","true");
  $(".CLIlayouts:checkbox[value='umap_harmony']").prop("checked","true");
}

function SYNCtest(){
  
}
// process annotation
function ABBRlist(){
  var grps = {},nrows=0,i;
  var x = document.getElementsByClassName('ABBRgrps');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      grps[x[i].value]=window.store.getState().annoMatrix.schema.annotations.obsByName[x[i].value].categories.map(String);
      nrows = Math.max(nrows,grps[x[i].value].length);
    }
  }
  var grpKey = Object.keys(grps);
  var htmlTable = "<table style='width:100%;border-collapse: collapse;'><tr>";
  for(i=0;i<grpKey.length;i++){
    htmlTable += "<th colspan='2' style='background-color: #4CAF50;color: white;'>"+grpKey[i]+"</th>"
  }
  htmlTable += "</tr><tr>";
  for(i=0;i<nrows;i++){
    for(var j=0;j<grpKey.length;j++){
      if(j>0){
        htmlTable += "<td align='right' style='border-left:2px solid #888'>"
      }else{
        htmlTable += "<td align='right'>"
      }
      if(i<grps[grpKey[j]].length){//grpKey[j]+"','"+grps[grpKey[j]][i]
        htmlTable += grps[grpKey[j]][i]+'</td><td><input type="text" id="'+grpKey[j]+"|"+grps[grpKey[j]][i]+'" value="' +window.bioAbbr[grpKey[j]][grps[grpKey[j]][i]]+'" onchange=ABBRupdate(["'+grpKey[j].replace(/ /g,";;")+'","'+grps[grpKey[j]][i].replace(/ /g,";;")+'"])></td>';
      }else{
        htmlTable += "</td><td></td>";
      }
    }
    htmlTable += "</tr><tr>"
  }
  htmlTable += "</tr></table>"
  $('#ABBRtable').html(htmlTable);
}
function ABBRupdate(eID){
  for(var i=0;i<eID.length;i++){
    eID[i] = eID[i].replace(/;;/g," ");
  }
  window.bioAbbr[eID[0]][eID[1]] = $('#'+eID.join("\\|").replace(/ /g,"\\ ")).val();
}
function ABBRclear(){
  var grpKeys = Object.keys(window.store.getState().categoricalSelection);
  for(var i=0;i<grpKeys.length;i++) {
    var selectionMap = new Map();
    var categories = window.store.getState().annoMatrix.schema.annotations.obsByName[grpKeys[i]].categories;
    for (var j=0;j<categories.length; j++) {
      selectionMap.set(categories[j], false);
    }
    window.store.dispatch({type: "categorical metadata filter none of these", metadataField: grpKeys[i], labelSelectionState: selectionMap});
  }
  return;
}
function ABBRall(){
  var grpKeys = Object.keys(window.store.getState().categoricalSelection);
  for(var i=0;i<grpKeys.length;i++) {
    var selectionMap = new Map();
    var categories = window.store.getState().annoMatrix.schema.annotations.obsByName[grpKeys[i]].categories;
    for (var j=0;j<categories.length; j++) {
      selectionMap.set(categories[j], true);
    }
    window.store.dispatch({type: "categorical metadata filter all of these", metadataField: grpKeys[i], labelSelectionState: selectionMap});
  }
  return;
}
//const indexOfAll = (arr, val) => arr.reduce((acc, el, i) => (el === val ? [...acc, i] : acc), []);
function ABBRcombineCell(){
  var annKeys = Object.keys(window.store.getState().categoricalSelection);
  var obsMap = window.store.getState().annoMatrix._cache.obs.colIndex.index;
  var grpKeys = Object.keys(window.bioCombine);
  var cName = [], cVal = [];//, selC=[]
  for(const key of grpKeys){
    var k = annKeys.indexOf(key);
    if(cName.length==0){
      for(const v of window.bioCombine[key]){
        cName.push(window.bioAbbr[key][v]);
        cVal.push(k+"_"+window.store.getState().annoMatrix.schema.annotations.obsByName[key].categories.indexOf(v));
        //selC.push(indexOfAll(window.store.getState().annoMatrix._cache.obs.__columns[obsMap.get(key)],v));
      }
    }else{
      var newGrp = [], newVal = [];//, newC = []
      for(var i=0;i<cName.length;i++){
        window.bioCombine[key].forEach(function(v){
          newGrp.push(cName[i]+":"+window.bioAbbr[key][v]);
          newVal.push(cVal[i]+"__"+k+"_"+window.store.getState().annoMatrix.schema.annotations.obsByName[key].categories.indexOf(v));
          //var oneSel = indexOfAll(window.store.getState().annoMatrix._cache.obs.__columns[obsMap.get(key)],v);
          //newC.push(selC[i].filter(v => oneSel.includes(v)));
        });
      }
      cName = newGrp;
      cVal = newVal;
      //selC = newC;
    }
  }
  var combCell = {};
  for(var i=0;i<cName.length;i++){
    //cName[i] += " ("+selC[i].length+")";
    //var cells = new Int32Array(selC[i].length);
    //cells.set(selC[i]);
    combCell[cVal[i]] = {cName:cName[i]};//,selC:cells
  }
  return combCell;
}
function ABBRcombine(){
  window.bioCombine={};
  var combGrp = {};
  var grpKeys = Object.keys(window.store.getState().categoricalSelection);
  for(var i=0;i<grpKeys.length;i++) {
    if(window.store.getState().categoricalSelection[grpKeys[i]].size == 0){
      combGrp[grpKeys[i]] = window.store.getState().annoMatrix.schema.annotations.obsByName[grpKeys[i]].categories;
    }else{
      var one = [];
      for (let [key, value] of window.store.getState().categoricalSelection[grpKeys[i]]) {
        if(value) {
          one.push(key);
        }
      }
      if(one.length>0){
        combGrp[grpKeys[i]] = one;
      }
    }
  }
  window.bioCombine=combGrp;
  window.bioCombineCell = ABBRcombineCell();
  var combVal = [];
  for(const one of Object.values(window.bioCombineCell)){
    combVal.push(one.cName);
  }
  
  $('#ABBRcombine').html("<span style='color:red;'><strong>"+combVal.join('; ')+"</strong></span>");
}
function ABBRcombineUpdate(){
  var combGrp = window.bioCombine;
  var combKey = Object.keys(combGrp);
  var allGrp = store.getState().annoMatrix.schema.annotations.obsByName;
  for(var i=0;i<combKey.length;i++){
    //console.log(allGrp[combKey[i]].categories);
    var selectionMap = new Map();
    for(var j=0;j<allGrp[combKey[i]].categories.length;j++){
      if(combGrp[combKey[i]].includes(allGrp[combKey[i]].categories[j])){
        selectionMap.set(allGrp[combKey[i]].categories[j], true);
        //console.log([combKey[i],allGrp[combKey[i]].categories[j]]);
      } else {
        selectionMap.set(allGrp[combKey[i]].categories[j], false);
      }
    }
    window.store.dispatch({type: "categorical metadata filter select", metadataField: combKey[i], labelSelectionState: selectionMap});
  }
  return;
}
// process genes
function ADDGadd(){
  var addG = $("#ADDGinput").val().trim().split(/\s+|\s*,\s*|\s*;\s*/);
  var genes =  window.biogen;
  var notG = [];
  var totalG = window.store.getState().annoMatrix._cache.var.__columns[0];
  var totalGlower = totalG.map(function(s){return s.toLowerCase();});
  
  for(var i=0;i<addG.length;i++){
    var one = addG[i].trim();
    if (addG[i].trim().length<1)
      continue;
    var ix = totalGlower.indexOf(one.toLowerCase());
    if(ix !== -1){
      var oneG = totalG[ix];
      if(!genes.includes(oneG)){
        genes.push(oneG);
      }
    }else{
      notG.push(one);
    }
  }
  $("#ADDGdel").text(notG.join(", "));
  $("#ADDGinput").val("");
  ADDGref();
}
function ADDGdel(gName){
  const ix = window.biogen.indexOf(gName);
  window.biogen.splice(ix,1);
  $("#ADDG_"+gName).remove();
}
function ADDGsetAdd(idx){//Geneset_Name Geneset_Genes
  $("#ADDGsetError").text("");
  var grpName = $("#Geneset_Name"+idx).val().trim();
  if(grpName.length<1){
    $("#ADDGsetError").text("Please provide a gene set name with at least one character!");
    return;
  }
  var grpKeys = Object.keys(window.bioGeneGrp);
  if(grpKeys.includes(grpName)){
    $("#ADDGsetError").text("The gene set name exists!");
    return;
  }
  var definedG = [];
  for(var i=0;i<grpKeys.length;i++){
    definedG = definedG.concat(window.bioGeneGrp[grpKeys[i]]);
  }
  
  var gNames = $("#Geneset_Genes"+idx).val().trim().split(/\s+|\s*,\s*|\s*;\s*/);
  var notG = [],geneNames=[];
  var totalG = window.store.getState().annoMatrix._cache.var.__columns[0];
  var totalGlower = totalG.map(function(s){return s.toLowerCase();});
  for(var i=0;i<gNames.length;i++){
    var one = gNames[i].trim();
    var ix = totalGlower.indexOf(one.toLowerCase());
    if(ix !== -1){
      var oneG = totalG[ix];
      if (one.length<1) continue;
      if(!definedG.includes(oneG)){
        geneNames.push(oneG);
      }else{
        notG.push(one);
      }
    }else{
      notG.push(one);
    }
  }
  var eMsg= "";
  if(notG.length>0){
    eMsg +="The following genes cannot be found in the data or defined in previous gene sets: "+notG;
    $("#ADDGsetError").text(eMsg);
  }
  if(geneNames.length==0){
    $("#ADDGsetError").text(eMsg+". No valid gene names is provided");
    return;
  }
  $("#Geneset_Name"+idx).val("");
  $("#Geneset_Genes"+idx).val("");
  window.bioGeneGrp[grpName] = geneNames;

  geneNames = geneNames.concat(window.biogen);
  geneNames = [...new Set(geneNames)];
  window.biogen = geneNames;
  ADDGref();
}
function ADDGsetDel(grpName){
  delete window.bioGeneGrp[grpName];
  ADDGref();
}
function hideShow(section, button, label) {
  section.toggle();
  if (label == null) label = 'Input'
  button.text(button.text()=='Hide '+label?'Show '+label:'Hide '+label);
}
// plot function 
function GDplot(){
  if(!window.store.getState().differential.celllist1 && !window.store.getState().differential.celllist2){
    $('#GDnote').text("Please select cells into group 1 or/and group 2");
    return;
  }
  var cells = {};
  if(window.store.getState().differential.celllist1){
    cells['group1'] = window.store.getState().differential.celllist1;
  }
  if(window.store.getState().differential.celllist2){
    cells['group2'] = window.store.getState().differential.celllist2;
  }
  if(+$('#GDcutoff').prop('min') > +$("#GDcutoff").val()){
    $('#GDnote').text("The minimal expression level is: "+$('#GDcutoff').prop('min'));
    return;
  }

  var D = {'method':'GD',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'cutoff':$("#GDcutoff").val(),
            'figOpt':window.figOpt};
  $(".GDbt").prop('disabled',true);
  $(".GDpngBT").prop('disabled',true);
  document.getElementById("GDspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      $(".GDbt").prop("disabled",false);
      document.getElementById("GDspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#GDnote').text(res);
        document.getElementById("GDfig").innerHTML = '<img id="GDpng" src="data:image/png;base64," width=100% height="auto"/>';
        return;
      }
      showImg('GDfig',res);
      $(".GDpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#GDnote').text(request.status+request.responseText);
      document.getElementById("GDfig").innerHTML = '<img id="GDpng" src="data:image/png;base64," width=100% height="auto"/>';
      $(".GDbt").prop("disabled",false);
      document.getElementById("GDspin").style.visibility="hidden";
    }
  });

}
function MARKplot(){
  var cells=[], x = document.getElementsByClassName('MARKcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#MARKnote').text("Please use selection tools to put cells into groups");
    return;
  }
  
  if($("#MARKn").val()<3){
    $('#MARKnote').text("Minimal number of markers per group is 3");
    return;
  }
  var grp = [$("#MARKgrp option:selected").val()];
  var combGrp = [];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
  }
  
  
  $(".MARKbt").prop("disabled",true);
  $(".MARKpngBT").prop('disabled',true);
  document.getElementById("MARKspin").style.visibility="visible";
  var D = {'method':'MARK',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':[],
            'grp':grp,
            'markMethod':$("#MARKmethod option:selected").val(),
            'geneN':$("#MARKn").val(),
            'abb':window.bioAbbr,
            'combine':combGrp,
            'figOpt':window.figOpt};
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      document.getElementById("MARKspin").style.visibility="hidden";
      $(".MARKbt").prop("disabled",false);
      if(res.startsWith('ERROR')){
        $('#MARKnote').text(res);
        $('#MARKtable').empty();
        document.getElementById("MARKfig").innerHTML = '';
        return;
      }

      resD = JSON.parse(res);
      var tD = resD[0];
      var keyD = tD.shift();
      var titleD = [];
      for(const one of keyD){
        titleD.push({'title':one})
      }
      $('#MARKtable').empty();
      var table = $('#MARKtable').DataTable({
        destroy: true,
        dom:'Blfrtip',
        lengthMenu:[[10,50,100,200,-1],[10,50,100,200,"All"]],
        order:[[1,"dsc"]],
        buttons:['csv'],
        data:tD,
        columns:titleD
      });

      $('#MARKtable tbody').on('click',"tr",function(){
        var data = table.row(this).data();
        //alert('Your clicked on '+data[0]+" row");
        var genes =  window.biogen;
        if(genes.includes(data[0])){
          genes.splice(genes.indexOf(data[0]),1);
        }else{
          genes.push(data[0]);
        }
        $('#MARKnote').text("Selected genes: "+genes.join(", "));
      });
      showImg('MARKfig',resD[1]);
      $(".MARKpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#MARKnote').text(request.status+request.responseText);
      $('#MARKtable').empty();
      document.getElementById("MARKfig").innerHTML = '';
      $(".MARKbt").prop("disabled",false);
      document.getElementById("MARKspin").style.visibility="hidden";
    }
  })

}
function SGVplot(acc){
  $('#SGVnote').text("");
  if($("#SGVgene option:selected").val()=="NA"){
    $('#SGVnote').text("Please select a gene to plot");
    return;
  }
  if($("#SGVgrp option:selected").val()=="NA"){
    $('#SGVnote').text("Please select a category to plot");
    return;
  }
  var cells=[], x = document.getElementsByClassName('SGVcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if($('#SGVgene').val() == null) {
    $('#SGVnote').text('Please add genes either from the VIP menu or the main window');
    return;
  }
  if(cells.length==0){
    $('#SGVnote').text("Please use selection tools to put cells into groups");
    return;
  }
  if(+$("#SGVcutoff").val() < +$("#SGVcutoff").prop('min')){
    $('#SGVnote').text("The minimal expression level is "+$("#SGVcutoff").prop('min'));
    return;
  }
  //$('#SGVnote').text("Plotting on "+cells.length+" selected cells");
  var grp = [$("#SGVgrp").val()];
  var combGrp = [];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
  }
  
  var D = {'method':'SGV',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':[$("#SGVgene").val()],
            'grp':grp,
            'abb':window.bioAbbr,
            'cutoff':$("#SGVcutoff").val(),
            'combine':combGrp,
            'figOpt':window.figOpt};
  if(acc=='data') D['method']='VIOdata';
  $(".SGVbt").prop("disabled",true);
  $(".SGVsaveBT").prop('disabled',true);
  $(".SGVpngBT").prop('disabled',true);
  document.getElementById("SGVspin").style.visibility="visible";
  
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".SGVbt").prop("disabled",false);
      $(".SGVsaveBT").prop('disabled',false);
      document.getElementById("SGVspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#SGVnote').text(res);
        document.getElementById("SGVfig").innerHTML = '';
        return;
      }
      if(acc=='data'){
        var hiddenE = document.createElement('a');
        hiddenE.href="data:attachment/text,"+encodeURIComponent(res);
        hiddenE.target='_blank';
        hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.violinData.csv';
        hiddenE.click();
      }else{
        showImg('SGVfig',res);
      }
      //document.getElementById("SGVfig").innerHTML = '<img id="SGVpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".SGVpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#SGVnote').text(request.status+request.responseText);
      document.getElementById("SGVfig").innerHTML = '<img id="SGVpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".SGVbt").prop("disabled",false);
      document.getElementById("SGVspin").style.visibility="hidden";
    }
  })
}
function getGeneSet(aID){
  var grpKeys=[];
  $("."+aID+"sets").each(function(){
    if($(this).prop("checked")){
      grpKeys.push($(this).val());
    }
  });
  var selSets={};
  for(const one of grpKeys){
    selSets[one]=window.bioGeneGrp[one];
  }
  return(selSets)
}
function getGenes(aID){
  var genes = [];
  $("."+aID+"genes").each(function(){
    if($(this).prop("checked")){
      genes.push($(this).val());
    }
  });
  return(genes);
}
function PGVplot(acc){
  $('#PGVnote').text("");
  var genes =getGenes('PGV');
  var geneSets = getGeneSet('PGV');

  if(genes.length==0&&Object.keys(geneSets).length==0){
    $('#PGVnote').text("Please select at least a gene or a gene set to plot");
    return;
  }
  
  if($("#PGVgrp option:selected").val()=="NA"){
    $('#PGVnote').text("Please select a category to plot");
    return;
  }
  var cells=[], x = document.getElementsByClassName('PGVcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#PGVnote').text("Please use selection tools to put cells into groups");
    return;
  }
  if(+$("#PGVcutoff").val()< +$("#PGVcutoff").prop('min')){
    $('#PGVnote').text("The minimal expression level is "+$("#PGVcutoff").prop('min'));
    return;
  }

  var grp = [$("#PGVgrp").val()];
  var combGrp = [];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
  }
  var geneGrpColl = $("input[name='PGVgeneGrpColl']:checked").val()
  if( geneGrpColl== 'Yes'){
    geneGrpColl = $("input[name='PGVgeneGrpCollCal']:checked").val()
  }
  var color = $("#PGVcolor").val();
  if ($("#PGVcolor_r")[0].checked) color += '_r';
  var D = {'method':'PGV',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'geneGrp':geneSets,
            'geneGrpColl':geneGrpColl,
            'grp':grp,
            'color':color,
            'by':$("#PGVgrpXY").val(),
            'abb':window.bioAbbr,
            'cutoff':$("#PGVcutoff").val(),
            'combine':combGrp,
            'figOpt':window.figOpt};
  if(acc=='data') D['method']='VIOdata';
  $(".PGVbt").prop('disabled',true);
  $(".PGVsaveBT").prop('disabled',true);
  $(".PGVpngBT").prop('disabled',true);
  document.getElementById("PGVspin").style.visibility="visible";

  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".PGVbt").prop("disabled",false);
      $(".PGVsaveBT").prop("disabled",false);
      document.getElementById("PGVspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#PGVnote').text(res);
        document.getElementById("PGVfig").innerHTML = '';
        return;
      }
      if(acc=='data'){
        var hiddenE = document.createElement('a');
        hiddenE.href="data:attachment/text,"+encodeURIComponent(res);
        hiddenE.target='_blank';
        hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.violinData.csv';
        hiddenE.click();
      }else{
        showImg('PGVfig',res);
      }
      //document.getElementById("PGVfig").innerHTML = '<img id="PGVpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".PGVpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#PGVnote').text(request.status+request.responseText);
      document.getElementById("PGVfig").innerHTML = '<img id="PGVpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".PGVbt").prop("disabled",false);
      document.getElementById("PGVspin").style.visibility="hidden";

    }
  })
}
function violinData(){
  
}
function HEATcheck(){
  $('#HEATnote').text("");
  var cells=[], x = document.getElementsByClassName('HEATcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }

  if(cells.length==0){
    $('#HEATnote').text("Please use selection tools to put cells into groups");
    return;
  }

  var genes = getGenes('HEAT');
  if(genes.length<2){
    $('#HEATnote').text("It needs at least two genes");
    return;
  }
  
  var grps = [];
  x = document.getElementsByClassName('HEATgrps');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      grps.push(x[i].value);
    }
  }
  
  var orderGrp = $("#HEATorder").val();
  if(orderGrp=="Expression" && cells.length>heatCell){
    $('#HEATnote').text("Cluster by expression is not available for the number of selected cells larger than "+Math.round(heatCell/1000)+"k.");
    return;
  }
  if(orderGrp!="Expression" && !grps.includes(orderGrp)){
    grps.push(orderGrp);
  }
  
  var grp=[],combGrp = [];
  if(orderGrp=="Custom_combine" || grps.includes("Custom_combine")){
    grp = grps.filter(function(e) { return e !== 'Custom_combine' });
    grps = [...new Set(grp.concat(Object.keys(window.bioCombine)))];
    combGrp = window.bioCombine;
  }
  
  if(grps.length==0){
    $('#HEATnote').text("Please select groups to plot");
    return;
  }
  document.getElementById("HEATspin").style.visibility="visible";
  $(".HEATbt").prop('disabled',true);
  $(".HEATpngBT").prop('disabled',true);
  $(".HEATsaveBT").prop('disabled',true);
  var D = {'method':'HEAT',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'order':orderGrp,
            'norm':$("#HEATnorm option:selected").val(),
            'grp':grps,
            'abb':window.bioAbbr,
            'combine':combGrp,
            'addGrp':grp,
            'figOpt':window.figOpt};
  return D;
}
function HEATplot(){
  D = HEATcheck();
  if(D==undefined) return;
  D['method'] = 'HEATplot';
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".HEATbt").prop("disabled",false);
      document.getElementById("HEATspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#HEATnote').text(res);
        document.getElementById("HEATfig").innerHTML = '';
        return;
      }

      showImg('HEATfig',res);//resD[0]
      //document.getElementById("HEATfig").innerHTML = '<img id="HEATpng" src="data:image/png;base64,'+resD[0]+'" width=100% height="auto"/>';// height="450" width="450"
      $(".HEATpngBT").prop('disabled',false);
      $(".HEATsaveBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#HEATnote').text(request.status+request.responseText);
      document.getElementById("HEATfig").innerHTML = '<img id="HEATpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".HEATbt").prop("disabled",false);
      document.getElementById("HEATspin").style.visibility="hidden";

    }
  });
}
function HEATdata(){
  D = HEATcheck();
  if(D==undefined) return;
  D['method'] = 'HEATdata';
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".HEATbt").prop("disabled",false);
      document.getElementById("HEATspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#HEATnote').text(res);
        return;
      }

      var hiddenE = document.createElement('a');
      hiddenE.href="data:attachment/text,"+encodeURIComponent(res);
      hiddenE.target='_blank';
      hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.heatmapData.csv';
      hiddenE.click();

      $(".HEATpngBT").prop('disabled',false);
      $(".HEATsaveBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#HEATnote').text(request.status+request.responseText);
      document.getElementById("HEATfig").innerHTML = '<img id="HEATpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".HEATbt").prop("disabled",false);
      document.getElementById("HEATspin").style.visibility="hidden";

    }
  });
}
function DEGselCells(){
  var selAnno = {};
  var selNames = [];
  //var selC = {};
  var grpKey = Object.keys(window.store.getState().categoricalSelection);
  $(".DEGgrps").each(function(){
    if($(this).prop("checked")){
      selNames.push(window.bioCombineCell[$(this).val()].cName);
      //selC[window.bioCombineCell[$(this).val()].cName.split(' ')[0]] = window.bioCombineCell[$(this).val()].selC;
      //window.store.getState().annoMatrix.schema.annotations.obsByName[key].categories.indexOf(v)
      $(this).val().split("__").forEach(function(one){
        [k,v]=one.split("_");
        if(selAnno[grpKey[k]]==undefined){selAnno[grpKey[k]]=[]}
        if(!selAnno[grpKey[k]].includes(window.store.getState().annoMatrix.schema.annotations.obsByName[grpKey[k]].categories[v])){
          selAnno[grpKey[k]].push(window.store.getState().annoMatrix.schema.annotations.obsByName[grpKey[k]].categories[v]);
        }
      });
    }
  })
  return [selAnno,selNames];
  //return selC;
}
function operator(a,b,op){
  switch(op){
    case ">": return a>b;
    case "=": return a==b;
    case "<": return a<b;
  }
}
function getSelOpt(eID){
  var selOpt=[];
  if($("input[name='"+eID+"colorBy']:checked").val()=='significance'){
    selOpt.push({column:'absLogfc',opr:'>',val:+$("#"+eID+"logfc").val()});
    selOpt.push({column:'qval',opr:'<',val:+$("#"+eID+"fdr").val()});
  }else if($("input[name='"+eID+"colorBy']:checked").val()=='geneInfo'){
    selOpt.push({column:$("#"+eID+"plotByGinfoSel").val(),
                  opr:$("input[name='"+eID+"plotByGinfoRd']:checked").val(),
                  val:$("#"+eID+"plotByGinfoVal").val()});
  }else if($("input[name='"+eID+"colorBy']:checked").val()=='selGene'){
    selOpt.push({column:'gene',opr:'include',val:getGenes('DEG')});
  }else{
    $('#'+eID+'note').text("Unknown selection!");
  }
  return selOpt; 
}
function updateDEGdata(rawD,selOpts){
  return d3.csvParse(rawD,function(d){
    for (var key in d) {
     d[key] = isNaN(+d[key]) ? d[key]:(+d[key]);
     d['absLogfc'] = Math.abs(d.log2fc)
     //for color-----------
     selCol = selOpts['color'];
     d['color'] = "unselected";
     if(selCol[0].opr=="include"){
       if(selCol[0].val.includes(d[selCol[0].column])) d['color'] = "selected";
     }else if(isNaN(+selCol[0].val)){
       d['color'] = d[selCol[0].column]
     }else{
       var flag=true;
       selCol.forEach(function(ref){
         flag = flag && operator(d[ref.column],ref.val,ref.opr);
       })
       if(flag) d['color'] = "selected";
     }
     //for label -----------
    selLab = selOpts['label'];
     d['label'] = true;
     if(selLab[0].opr=="include"){
       if(!selLab[0].val.includes(d[selLab[0].column])) d['label'] = false;
     }else if(isNaN(+selLab[0].val)){
       if(d['label'] != d[selLab[0].column]) d['label'] = false;
     }else{
       selLab.forEach(function(ref){
         d['label'] = d['label'] && operator(d[ref.column],ref.val,ref.opr);
       })
     }
    }
    var selD = {};
    ['gene','color','log2fc','qval','label'].forEach(function(i){
      selD[i] = d[i];
    })
    return selD;
  });
}
function DEGplot(eID){//window.DEGraw,window.DEGsel 'DEG'
  $("#"+eID+"fig").html("");
  window[eID+'sel'][$("#"+eID+"plotBy").val()] = getSelOpt(eID);
  var DEGdata=updateDEGdata(window[eID+'raw'],window[eID+'sel']);

  var vol = volcanoPlot()
      .width($("#"+eID+"fig").width())
      .height($("#"+eID+"fig").height())
      .xAxisLabel('log<tspan baseline-shift="sub">2</tspan>Fold-change')
      .yAxisLabel('-log<tspan baseline-shift="sub">10</tspan>False Discovery Rate')
      .vLines({a:-$("#"+eID+"logfc").val(),b:+$("#"+eID+"logfc").val()})
      .hLines({['qvalue='+$("#"+eID+"fdr").val()]:$("#"+eID+"fdr").val()})
      .sampleID("gene")
      .xColumn("log2fc")
      .yColumn("qval")
      .colorColumn('color')
      .labelColumn('label')
      .maxY(50)
      .tooltips(['gene','log2fc','qval'])
      .resetBtnID(eID+"figReset");


  d3.select('#'+eID+'fig')
    .data([DEGdata.filter(function(d){return Math.abs(d.log2fc)< +$("#"+eID+"logfcCut").val();})])
    .call(vol);
  
}
function DEGfind(){
  $('#DEGnote').text("");
  var D = {'method':'DEG',
           'DEmethod':$("#DEGmethod").val(),
           'figOpt':window.figOpt}
  const [combGrp,combName] = DEGselCells();
  var cells = {};
  if(combName.length!=2){
    if(!window.store.getState().differential.celllist1 || !window.store.getState().differential.celllist2){
      $('#DEGnote').text("Please select cells into both group 1 and group 2");
      return;
    }else{
      cells['group1'] = window.store.getState().differential.celllist1;
      cells['group2'] = window.store.getState().differential.celllist2;
    }
    D['cells'] = cells;
    D['cellN'] = window.store.getState().annoMatrix.nObs;
    D['comGrp'] = Object.keys(cells);
  }else{
    D['combine']=combGrp;
    D['abb']=window.bioAbbr;
    D['grp'] = Object.keys(combGrp);
    D['cells'] = new Int32Array(Array(window.store.getState().annoMatrix.nObs).keys());
    D['comGrp'] = combName.sort();
  }
  $('#DEGinfo').text("DEG: "+D['comGrp'].join(" vs "));

  document.getElementById("DEGspin").style.visibility="visible";
  $(".DEGbt").prop('disabled',true);
  $(".DEGpngBT").prop('disabled',true);
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      document.getElementById("DEGspin").style.visibility="hidden";
      $(".DEGbt").prop("disabled",false);
      $(".DEGpngBT").prop('disabled',false);
      $('#DEGtable').empty();
      
      if(res.startsWith('ERROR')){
        $('#DEGnote').text(res);
        return;
      }
      //resD = JSON.parse(res);
      //showImg('DEGfig',resD[1]);
      // the default color and label
      var selOpt=[];
      selOpt.push({column:'absLogfc',opr:'>',val:+$("#DEGlogfc").val()});
      selOpt.push({column:'qval',opr:'<',val:+$("#DEGfdr").val()});
      window.DEGsel = {};
      $("#DEGplotBy option").each(function(){
        window.DEGsel[$(this).val()]=selOpt;
      });
      // gene information
      res.split("\n")[0].split(',').slice(5).forEach(function(d){
        $('#DEGplotByGinfoSel').append('<option value="'+d+'">'+d+'</option>')
      });
      // raw DEG table
      window.DEGraw = res;
      DEGplot('DEG');
      $("#DEGfinal").show();
      // process DEG table
      var resD = d3.csvParse(res,function(data){
                              return [data.gene,
                                      (data.log2fc*1).toFixed(2),
                                      (1*data.pval).toExponential(2),
                                      (1*data.qval).toExponential(2)];
                            });
      var table = $('#DEGtable').DataTable({
        destroy: true,
        dom:'Blfrtip',
        lengthMenu:[[10,50,100,200,-1],[10,50,100,200,"All"]],
        order:[[3,"asc"]],
        buttons:['csv'],
        data:resD,
        columns:[
          {title: "Gene"},
          {title: "Log Fold Change"},
          {title: "p-value"},
          {title: "padj"}
        ]
      });
      $('#DEGtable tbody').on('click',"tr",function(){
        var data = table.row(this).data();
        //alert('Your clicked on '+data[0]+" row");
        var genes =  window.biogen;
        if(genes.includes(data[0])){
          genes.splice(genes.indexOf(data[0]),1);
        }else{
          genes.push(data[0]);
        }
        $('#DEGnote').text("Selected genes: "+genes.join(", "));
      });
    },
    error: function(request,status,error){
      $('#DEGnote').text(request.status+request.responseText);
      $('#DEGtable').empty();
      document.getElementById("DEGspin").style.visibility="hidden";
      $(".DEGbt").prop("disabled",false);
    }
  })
  
}
function DOTplot(){
  $('#DOTnote').text("");
  var genes =getGenes('DOT');
  var geneSets = getGeneSet('DOT');

  if(genes.length==0&&Object.keys(geneSets).length==0){
    $('#DOTnote').text("Please select at least a gene or a gene set to plot");
    return;
  }

  var cells=[], x = document.getElementsByClassName('DOTcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#DOTnote').text("Please use selection tools to put cells into groups");
    return;
  }

  if(+$('#DOTcutoff').prop('min') > +$("#DOTcutoff").val()){
    $('#DOTnote').text("The minimal expression level is: "+$('#DOTcutoff').prop('min'));
    return;
  }
  
  var grp = [$("#DOTgrp").val()];
  var combGrp = [];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
  }
  var geneGrpColl = $("input[name='DOTgeneGrpColl']:checked").val()
  if( geneGrpColl== 'Yes'){
    geneGrpColl = $("input[name='DOTgeneGrpCollCal']:checked").val()
  }

  var color = $("#DOTcolor").val();
  if ($("#DOTcolor_r")[0].checked) color += '_r';
  
  var D = {'method':'DOT',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'abb':window.bioAbbr,
            'combine':combGrp,
            'geneGrp':geneSets,
            'geneGrpColl':geneGrpColl,
            'cutoff':$("#DOTcutoff").val(),
            'mean_only_expressed':$("input[name='mean_only_expressed']:checked").val(),
            'legendW':$('#DOTlegendW').val(),
            'color':color,
            'figOpt':window.figOpt};
  $(".DOTbt").prop('disabled',true);
  document.getElementById("DOTspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".DOTbt").prop("disabled",false);
      document.getElementById("DOTspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#DOTnote').text(res);
        document.getElementById("DOTfig").innerHTML = '';
        return;
      }
      
      showImg('DOTfig',res)
      //document.getElementById("DOTfig").innerHTML = '<img id="DOTpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".DOTpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#DOTnote').text(request.status+request.responseText);
      document.getElementById("DOTfig").innerHTML = '<img id="DOTpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".DOTbt").prop("disabled",false);
      document.getElementById("DOTspin").style.visibility="hidden";
    }
  })
}
function TRAKplot(){
  $('#TRAKnote').text("");
  var genes =getGenes('TRAK');
  var geneSets = getGeneSet('TRAK');
  if(genes.length==0&&Object.keys(geneSets).length==0){
    $('#TRAKnote').text("Please select at least a gene or a gene set to plot");
    return;
  }

  var cells=[], x = document.getElementsByClassName('TRAKcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#TRAKnote').text("Please use selection tools to put cells into groups");
    return;
  }
  
  var grp = [$("#TRAKgrp").val()];
  var combGrp = [];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
  }
  var geneGrpColl = $("input[name='TRAKgeneGrpColl']:checked").val()
  if( geneGrpColl== 'Yes'){
    geneGrpColl = $("input[name='TRAKgeneGrpCollCal']:checked").val()
  }

  var D = {'method':'TRAK',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'abb':window.bioAbbr,
            'combine':combGrp,
            'geneGrp':geneSets,
            'geneGrpColl':geneGrpColl,
            'figOpt':window.figOpt};
  $(".TRAKbt").prop('disabled',true);
  document.getElementById("TRAKspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".TRAKbt").prop("disabled",false);
      document.getElementById("TRAKspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#TRAKnote').text(res);
        document.getElementById("TRAKfig").innerHTML = '';
        return;
      }

      showImg('TRAKfig',res);
      //document.getElementById("TRAKfig").innerHTML = '<img id="TRAKpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".TRAKpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#TRAKnote').text(request.status+request.responseText);
      document.getElementById("TRAKfig").innerHTML = '<img id="TRAKpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".TRAKbt").prop("disabled",false);
      document.getElementById("TRAKspin").style.visibility="hidden";
    }
  })
  
}
function DENSplot(){
  $('#DENSerror').text("");
  var cells=[], x = document.getElementsByClassName('DENScell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#DENSerror').text("Please use selection tools to put cells into groups");
    return;
  }
  
  var genes = getGenes('DENS');
  if(genes.length<1){
    $('#DENSerror').text("It needs at least a gene");
    return;
  }
  if(+$("#DENSprec").val() < +$("#DENSprec").prop('min')){
    $('#DENSerror').text("The minimal bandwidth is "+$("#DENSprec").prop('min'));
    return;
  }
  
  if($("#DENSsplit").val()==$("#DENSgrp").val()){
    $('#DENSerror').text("Please choose different annotations for Split and Color!");
    return;
  }
  var category = [$("#DENSsplit").val(),$("#DENSgrp").val()];
  var combGrp = [];
  var grp= [];
  if(category.includes("Custom_combine")){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
  }
  if($("#DENSsplit").val()!="Custom_combine" && $("#DENSsplit").val()!='None'){
    grp.push($("#DENSsplit").val())
  }
  if($("#DENSgrp").val()!="Custom_combine" && $("#DENSgrp").val()!='None'){
    grp.push($("#DENSgrp").val())
  }

  var bw = $("#DENSprec").val();
  var D = {'method':'DENS',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'bw':bw,
            'category':category,
            'abb':window.bioAbbr,
            'combine':combGrp,
            'figOpt':window.figOpt};
  $(".DENSbt").prop('disabled',true);
  document.getElementById("DENSspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".DENSbt").prop("disabled",false);
      document.getElementById("DENSspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#DENSerror').text(res);
        document.getElementById("DENSfig").innerHTML = '';
        return;
      }

      showImg('DENSfig',res);
      //document.getElementById("TRAKfig").innerHTML = '<img id="TRAKpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".DENSpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#DENSerror').text(request.status+request.responseText);
      document.getElementById("DENSfig").innerHTML = '<img id="DENSpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".DENSbt").prop("disabled",false);
      document.getElementById("DENSspin").style.visibility="hidden";
    }
  })
            
}
function DENS2Dplot(){
  $('#DENS2Dnote').text("");
  var cells=[], x = document.getElementsByClassName('DENS2Dcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#DENS2Dnote').text("Please use selection tools to put cells into groups");
    return;
  }
  
  var genes = [];
  $(".DENS2Dgenes").each(function(){
    if($(this).prop("checked")){
      genes.push($(this).val());
    }
  })
  if(genes.length!=2){
    $('#DENS2Dnote').text("Please select two genes to plot");
    return;
  }

  document.getElementById("DENS2Dspin").style.visibility="visible";
  $(".DENS2Dbt").prop('disabled',true);
  $(".DENS2DpngBT").prop('disabled',true);

  var D = {'method':"DENS2D",
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':[],
            //'layout':$("#DUALlayout").val(),
            'cutoff':$("#DENS2Dcutoff").val(),
            'bandwidth':$("#DENS2Dbandwidth").val(),
            'figOpt':window.figOpt};
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".DENS2Dbt").prop("disabled",false);
      document.getElementById("DENS2Dspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#DENS2Dnote').text(res);
        document.getElementById("DENS2Dfig").innerHTML = '';
        return;
      }

      showImg('DENS2Dfig',res);
      $(".DENS2DpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#DENS2Dnote').text(request.status+request.responseText);
      document.getElementById("DENS2Dfig").innerHTML = '';// height="450" width="450"
      $(".DENS2Dbt").prop("disabled",false);
      document.getElementById("DENS2Dspin").style.visibility="hidden";
    }
  })
  
}
function EMBEDplot(){
  $('#EMBEDnote').text("");
  var cells=[], x = document.getElementsByClassName('EMBEDcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#EMBEDnote').text("Please use selection tools to put cells into groups");
    return;
  }
  
  var genes = getGenes('EMBED');
  if(genes.length==0){
    $('#EMBEDnote').text("Please select a gene to plot");
    return;
  }
  
  var grps = [];
  x = document.getElementsByClassName('EMBEDgrps');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      grps.push(x[i].value);
    }
  }
  if(grps.length==0){
    $('#EMBEDnote').text("Please select groups to plot");
    return;
  }
  if($('#EMBEDslot').val()<2){
    $('#EMBEDnote').text("The minimal number of plots per row is 2");
    return;
  }
  
  
  document.getElementById("EMBEDspin").style.visibility="visible";
  $(".EMBEDbt").prop('disabled',true);
  $(".EMBEDpngBT").prop('disabled',true);

  var D = {'method':"EMBED",
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grps,
            'ncol':$('#EMBEDslot').val(),
            'layout':$("#EMBEDlayout").val(),
            'figOpt':window.figOpt};
  if($("#EMBEDsplit").val()!="NONE"){
    D['splitGrp'] = $("#EMBEDsplit").val();
    grps.push($("#EMBEDsplit").val());
    D['grp'] = [...new Set(grps)];
  }
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".EMBEDbt").prop("disabled",false);
      document.getElementById("EMBEDspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#EMBEDnote').text(res);
        document.getElementById("EMBEDfig").innerHTML = '';
        return;
      }

      showImg('EMBEDfig',res);
      //document.getElementById("EMBEDfig").innerHTML = '<img id="EMBEDpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".EMBEDpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#EMBEDnote').text(request.status+request.responseText);
      document.getElementById("EMBEDfig").innerHTML = '';// height="450" width="450"
      $(".EMBEDbt").prop("disabled",false);
      document.getElementById("EMBEDspin").style.visibility="hidden";
    }
  })

}
function DUALplot(){
  $('#DUALnote').text("");
  var cells=[], x = document.getElementsByClassName('DUALcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#DUALnote').text("Please use selection tools to put cells into groups");
    return;
  }
  
  var genes = [];
  $(".DUALgenes").each(function(){
    if($(this).prop("checked")){
      genes.push($(this).val());
    }
  })
  if(genes.length!=2){
    $('#DUALnote').text("Please select two genes to plot");
    return;
  }
  if(+$('#DUALcutoff').prop('min') > +$("#DUALcutoff").val()){
    $('#DUALnote').text("The minimal expression level is: "+$('#DUALcutoff').prop('min'));
    return;
  }
  
  document.getElementById("DUALspin").style.visibility="visible";
  $(".DUALbt").prop('disabled',true);
  $(".DUALpngBT").prop('disabled',true);

  var D = {'method':"DUAL",
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':[],
            'layout':$("#DUALlayout").val(),
            'cutoff':$("#DUALcutoff").val(),
            'figOpt':window.figOpt};
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".DUALbt").prop("disabled",false);
      document.getElementById("DUALspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#DUALnote').text(res);
        document.getElementById("DUALfig").innerHTML = '';
        return;
      }

      showImg('DUALfig',res);
      //document.getElementById("DUALfig").innerHTML = '<img id="DUALpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".DUALpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#DUALnote').text(request.status+request.responseText);
      document.getElementById("DUALfig").innerHTML = '';// height="450" width="450"
      $(".DUALbt").prop("disabled",false);
      document.getElementById("DUALspin").style.visibility="hidden";
    }
  })
}
function SANKcheck(){
  $('#SANKnote').text("");
  var cells=[], x = document.getElementsByClassName('SANKcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#SANKnote').text("Please use selection tools to put cells into groups");
    return;
  }
  var genes =getGenes('SANK');
  //var geneSets = getGeneSet('SANK');
  var grp = [];
  $(".SANKgrps").each(function(){
  if($(this).prop("checked")){
    grp.push($(this).val());
  }})
  
  var selSank = $("#SANKdynList").sortable('toArray');
  selSank=selSank.map(function(one){return one.replace("SANKli","");})
  if (selSank.length < 2) {
    $('#SANKnote').text("Please select at least two features (gene or annotation) to plot");
    return;
  }
  var D = {'method':'SANK',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            //'geneGrp':geneSets,
            'grp':grp,
            'sankOrder':selSank,
            'sankBin':$("#sankBin").val(),
            'imgW':$("#SANKw").val(),
            'imgH':$("#SANKh").val(),
            'abb':window.bioAbbr,
            'figOpt':window.figOpt};
  return(D);  
}
function SANKplot(){
  D = SANKcheck();
  if (!D) return;
  $(".SANKbt").prop('disabled',true);
  document.getElementById("SANKspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".SANKbt").prop("disabled",false);
      document.getElementById("SANKspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#SANKnote').text(res);
        $('#SANKresize').text("");
        return;
      }
      $('#SANKresize').html(res);

    },
    error: function(request,status,error){
      $('#SANKnote').text(request.status+request.responseText);
      $(".SANKbt").prop("disabled",false);
      document.getElementById("SANKspin").style.visibility="hidden";

    }
  })
}
function SANKimgSave(){
  var D = SANKcheck();
  if (!D) return;
  D['imgSave'] = 'svg';
  $(".SANKbt").prop('disabled',true);
  document.getElementById("SANKspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".SANKbt").prop("disabled",false);
      document.getElementById("SANKspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#SANKnote').text(res);
        $('#SANKresize').text("");
        return;
      }
      var a = document.createElement('a');
      a.href = "data:image/svg+xml;base64,"+res;//
      a.download = "Sankey.svg";
      a.target="_blank";
      a.click();
    },
    error: function(request,status,error){
      $('#SANKnote').text(request.status+request.responseText);
      $(".SANKbt").prop("disabled",false);
      document.getElementById("SANKspin").style.visibility="hidden";

    }
  })
  
}
function STACBARplot(){
  $('#STACBARnote').text("");
  var cells=[], x = document.getElementsByClassName('STACBARcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#STACBARnote').text("Please use selection tools to put cells into groups");
    return;
  }
  var genes =getGenes('STACBAR');
  //var geneSets = getGeneSet('SANK');
  var grp = [];
  $(".STACBARgrps:checkbox:checked").each(function(){
    grp.push($(this).val());
  })
  if((genes.length+grp.length)<2){
    $('#STACBARnote').text("Please select at least two features (genes and/or annotations) to plot");
    return;
  }

  var D = {'method':'STACBAR',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'colorBy':$("#STACBARcol").val(),
            'Nbin':$("#STACBARbin").val(),
            'abb':window.bioAbbr,
            'figOpt':window.figOpt};

  $(".STACBARbt").prop('disabled',true);
  document.getElementById("STACBARspin").style.visibility="visible";
  $("#STACBARresize").html("");
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".STACBARbt").prop("disabled",false);
      document.getElementById("STACBARspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#STACBARnote').text(res);
        $('#STACBARresize').text("");
        return;
      }
      res = JSON.parse(res);
console.log(res);
      stackedBar('STACBARresize',res);
    },
    error: function(request,status,error){
      $('#STACBARnote').text(request.status+request.responseText);
      $(".STACBARbt").prop("disabled",false);
      document.getElementById("STACBARspin").style.visibility="hidden";

    }
  })
  
}
function CLIplot(){
  $('#CLInote').text("");
  $('#CLIsvg').html("");
  var cells=[], x = document.getElementsByClassName('CLIcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#CLInote').text("Please use selection tools to put cells into groups");
    return;
  }
  
  var genes = getGenes('CLI');
  if(genes.length==0){
    if(!confirm("No genes is sepecified. Are you sure you want to continue with all genes (might be very slow or out of memory)?")){
      return;
    }
  }
  
  var grps = [];
  $(".CLIgrps:checkbox:checked").each(function(){
    grps.push($(this).val());
  })
  if(grps.length==0){
    $('#CLInote').text("Please select at least one annotation");
    return;
  }

  var layouts = [];
  $(".CLIlayouts:checkbox:checked").each(function(){
    layouts.push($(this).val());
  })  
  if(layouts.length==0){
    $('#CLInote').text("Please select at least one layout");
    return;
  }
  
  document.getElementById("CLIspin").style.visibility="visible";
  $(".CLIbt").prop('disabled',true);
  
  var D = {'method':"CLI",
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grps,
            'layout':layouts,
            'script':editor.getSession().getValue()};
   $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".CLIbt").prop("disabled",false);
      document.getElementById("CLIspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#CLInote').text(res);
        $("#CLIresize").html("");
        return;
      }
      var filteredRes = '<style type="text/css"> pre { display: block; padding: 8.5px; margin: 0 0 9px; font-size: 12px; line-height: 1.42857143; word-break: break-all; word-wrap: break-word; color: #333333; background-color: #f5f5f5; border: 1px solid #ccc; border-radius: 2px; }';
      var readIt = false;
      res.split(/(\r\n|\n|\r)/g).forEach(function(line) {
        if (line.includes('div.inner_cell') || line.includes('<body>')) readIt = true;
        if (readIt) {
//          line = line.replace('output_subarea','');
          filteredRes += line;
        }
        if (line.includes('<\/body>')) readIt = false;
      });
      $("#CLIresize").html(filteredRes);

      var num = 0;
      for (let item of document.getElementsByClassName('output_svg')) {
        makeIdsUnique(item.getElementsByTagName('svg')[0], false);
        var onclickCode = "var a = document.createElement('a'); a.href = 'data:application/octet-stream;base64,' + btoa(unescape(encodeURIComponent(document.getElementsByClassName('output_svg')["+num+"].getElementsByTagName('svg')[0].outerHTML))); a.download = 'CLI-"+num+".svg'; a.click();"
        item.innerHTML += '<p><button class="CLIbt" onclick="' + onclickCode + '">Save as SVG</button></p>';
        num ++;
      }
    },
    error: function(request,status,error){
      $('#CLInote').text(request.status+request.responseText);
      $("#CLIresize").html("");
      $(".CLIbt").prop("disabled",false);
      document.getElementById("CLIspin").style.visibility="hidden";
    }
  })
}
//save and load
function checkSave(eID){
  var notCK = [];
  for(const one of $("."+eID+":checkbox:not(:checked)")){
    notCK.push(one.value);
  }
  return notCK;
}
function checkLoad(eID,notCK){
  for(const one of $("."+eID)){
    if(notCK.includes(one.value)){
      one.checked=false;
    }else{
      one.checked=true;
    }
  }
}
function ABBRsave(){
  return [window.bioAbbr,window.bioCombine,checkSave('ABBRgrps')];
}
function ABBRload(content){
  window.bioAbbr= content[0];
  window.bioCombine = content[1];
  checkLoad("ABBRgrps",content[2]);
  
  ABBRlist();
  ABBRclear();
  ABBRcombineUpdate();
  ABBRcombine();
}
function ADDsave(){
  return [window.biogen,window.bioGeneGrp];
}
function ADDload(content){
  window.biogen = content[0];
  window.bioGeneGrp = content[1];
  sync();
}
function IMGsave(){
  return(window.figOpt);
}
function IMGload(content){
  Object.keys(content).forEach(function(k) {
    window.figOpt[k] = content[k];
  });
}

function GDsave(){
  var cells = {'celllist1':[],'celllist2':[]};
  if(window.store.getState().differential.celllist1){
    cells['celllist1'] = window.store.getState().differential.celllist1;
  }
  if(window.store.getState().differential.celllist2){
    cells['celllist2'] = window.store.getState().differential.celllist2;
  }
  
  if($('#GDcutoff').prop('min')>$("#GDcutoff").val()){
    $('#GDnote').text("The minimal expression level is: "+$('#GDcutoff').prop('min'));
    return;
  }
  return [cells,Math.max($("#GDcutoff").val(),$('#GDcutoff').prop('min'))];
}
function GDload(content){
  cells = content[0];
  var newCells={};
  for(const one in cells){
    var arr=[];
    for(var p in Object.getOwnPropertyNames(cells[one])){
      arr[p]=cells[one][p];
    }
    newCells[one] = new Int32Array(arr);
  }

  for(const one of Object.keys(newCells)){
    if(newCells[one].length>0){
      window.store.getState().differential[one] = newCells[one];
    }else{
      window.store.getState().differential[one] = null;
    }
  }
  $("#GDcutoff").val(content[1]);
  cellNupdate();
}
function MARKsave(){
  return [checkSave('MARKcell'),$("#MARKgrp").val(),$("#MARKmethod").val(),$("#MARKn").val()];
}
function MARKload(content){
  checkLoad('MARKcell',content[0]);
  $("#MARKgrp").val(content[1])
  $("#MARKmethod").val(content[2])
  $("#MARKn").val(content[3])
}
function SGVsave(){
  return [checkSave('SGVcell'),$("#SGVgene").val(),$("#SGVgrp").val(),
          Math.max($("#SGVcutoff").val(),$('#SGVcutoff').prop('min'))];
}
function SGVload(content){
  checkLoad('SGVcell',content[0]);
  $("#SGVgene").val(content[1]);
  $("#SGVgrp").val(content[2]);
  $("#SGVcutoff").val(content[3]);
}
function PGVsave(){
  return [checkSave('PGVcell'),checkSave("PGVgenes"),checkSave('PGVgeneGrp'),
          $("#PGVgrp").val(),$("#PGVgrpXY").val(),
          Math.max($("#PGVcutoff").val(),$('#PGVcutoff').prop('min'))];
}
function PGVload(content){
  checkLoad('PGVcell',content[0]);
  checkLoad('PGVgenes',content[1]);
  checkLoad('PGVgeneGrp',content[2])
  $("#PGVgrp").val(content[3]);
  $("#PGVgrpXY").val(content[4]);
  $("#PGVcutoff").val(content[5]);
}
function HEATsave(){
  return [checkSave('HEATcell'),checkSave("HEATgenes"),$("#HEATorder").val(),checkSave("HEATgrps")];
}
function HEATload(content){
  checkLoad('HEATcell',content[0]);
  checkLoad("HEATgenes",content[1]);
  $("#HEATorder").val(content[2]);
  checkLoad("HEATgrps",content[3]);
}
function EMBEDsave(){
  return [checkSave('EMBEDcell'),checkSave("EMBEDgenes"),checkSave("EMBEDgrps"),
          $("#EMBEDlayout").val(),$('#EMBEDslot').val(),$("#EMBEDsplit").val()];
}
function EMBEDload(content){
  checkLoad('EMBEDcell',content[0]);
  checkLoad("EMBEDgenes",content[1]);
  checkLoad("EMBEDgrps",content[2]);
  $("#EMBEDlayout").val(content[3]);
  $('#EMBEDslot').val(content[4]);
  $('#EMBEDsplit').val(content[5]);
}
function DOTsave(){
  return [checkSave('DOTcell'),checkSave("DOTgenes"),checkSave('DOTgeneGrp'),
          $("#DOTgrp").val(),Math.max($("#DOTcutoff").val(),$('#DOTcutoff').prop('min'))];
}
function DOTload(content){
  checkLoad('DOTcell',content[0]);
  checkLoad("DOTgenes",content[1]);
  checkLoad('DOTgeneGrp',content[2])
  $("#DOTgrp").val(content[3]);
  $("#DOTcutoff").val(content[4]);
}
function TRAKsave(){
  return [checkSave('TRAKcell'),checkSave("TRAKgenes"),$("#TRAKgrp").val(),checkSave("TRAKgeneGrp")];
}
function TRAKload(content){
  checkLoad('TRAKcell',content[0]);
  checkLoad("TRAKgenes",content[1]);
  $("#TRAKgrp").val(content[2]);
  checkLoad("TRAKgeneGrp",content[3]);
}
function DENSsave(){
  return [checkSave('DENScell'),checkSave("DENSgenes"),
          $("#DENSsplit").val(),$("#DENSgrp").val(),
          $("#DENScutoff").val()]
}
function DENSload(content){
  checkLoad('DENScell',content[0]);
  checkLoad('DENSgenes',content[1]);
  $("#DENSsplit").val(content[2]);
  $("#DENSgrp").val(content[3]);
  $("#DENScutoff").val(content[4]);
}
function DUALsave(){
  return [checkSave('DUALcell'),checkSave("DUALgenes"),Math.max($("#DUALcutoff").val(),$('#DUALcutoff').prop('min')),$("#DUALlayout").val()];
}
function DUALload(content){
  checkLoad('DUALcell',content[0]);
  checkLoad("DUALgenes",content[1]);
  $("#DUALcutoff").val(content[2]);
  $("#DUALlayout").val(content[3]);
}
function DEGsave(){
  return [checkSave('DEGgrp'),$("#DEGtop").val(),$("#DEGmethod").val()]
}
function DEGload(content){
  checkLoad('DEGgrp',content[0])
  $("#DEGtop").val(content[1]);
  $("#DEGmethod").val(content[2]);
}
function save(){
  sync();

  var content = ['cellxgene.'+window.store.getState().config.displayNames.dataset];
  content.push(ABBRsave());
  content.push(ADDsave());
  content.push(IMGsave());
  content.push(GDsave());
  content.push(MARKsave());
  content.push(SGVsave());
  content.push(PGVsave());
  content.push(HEATsave());
  content.push(EMBEDsave());
  content.push(DOTsave());
  content.push(TRAKsave());
  content.push(DENSsave());
  content.push(DUALsave());
  content.push(DEGsave());
  
  
  var hiddenE = document.createElement('a');
  hiddenE.href="data:attachment/text,"+encodeURIComponent(JSON.stringify(content));
  hiddenE.target='_blank';
  hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.info.txt';
  hiddenE.click();
}
function inputClick(){
  sync();
  $("#loadfile").trigger('click');
}
function load(){
  var reader = new FileReader();
  reader.onload=function(ev){
    var content = JSON.parse(decodeURIComponent(ev.target.result));
    if(content[0]!='cellxgene.'+window.store.getState().config.displayNames.dataset){
      if(!confirm("Saved information is NOT matching with the current data! Continue?")){
        return;
      }
    }
    var i=1;
    ABBRload(content[i++]);
    ADDload(content[i++]);
    IMGload(content[i++]);
    GDload(content[i++]);
    MARKload(content[i++]);
    SGVload(content[i++]);
    PGVload(content[i++]);
    HEATload(content[i++]);
    EMBEDload(content[i++]);
    DOTload(content[i++]);
    TRAKload(content[i++]);
    DENSload(content[i++]);
    DUALload(content[i++]);
    DEGload(content[i++]);
    ABBRall();
//    alert("Saved information was loaded successfully!\nYou might start with 'Check All Features' in 'Comb. & Abbr.' tab!");
  };
  reader.readAsText($("#loadfile")[0].files[0]);
  
}
// ignore global definition for stackedBar legend circle
function GetStyle(CLASSname)
{
    var styleSheets = document.styleSheets;
    var styleSheetsLength = styleSheets.length;
    for(var i = 0; i < styleSheetsLength; i++){
        if (styleSheets[i].rules ) { var classes = styleSheets[i].rules; }
        else {  
            try {  if(!styleSheets[i].cssRules) {continue;} }
            //Note that SecurityError exception is specific to Firefox.
            catch(e) { if(e.name == 'SecurityError') { console.log("SecurityError. Cant readd: "+ styleSheets[i].href);  continue; }}
            var classes = styleSheets[i].cssRules ;
        }       
        for (var x = 0; x < classes.length; x++) {
            if (classes[x].selectorText == CLASSname) {
                return classes[x];
            }       
        }       
    }
}
GetStyle('circle').style = {};
</script>

<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/DataTables/datatables.min.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/ace/ace.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/script.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/select.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/endAll.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/plot-transform.js"></script>
<script type="text/javascript" src="https://baohongz.github.io/figureComposer/dist/svg-inject.js"></script>
<div id="CLIVignetteCode1" style="display:none;">
import warnings
warnings.filterwarnings('ignore')
import scanpy as sc
from matplotlib import pyplot as plt
%config InlineBackend.figure_formats = ['svg']
#
sc.pp.scale(adata, zero_center=False)
sc.tl.rank_genes_groups(adata,groupby='cell_type',groups=['OPC','Astrocytes'],method='t-test')
marker_genes=['LHFPL3', 'PCDH15', 'DSCAM', 'TNR', 'NLGN1','AQP4', 'VCAN','CUX2']
geneGrpPos = [(0,2), (3,4), (5,7)]
geneGrpName = ['set1', 'set2', 'set3']
fig,axs = plt.subplots(1,3,figsize=(15,5))
sc.pl.rank_genes_groups_violin(adata,groups='OPC',n_genes=8,ax=axs[0],show=False,strip=False)
sc.pl.stacked_violin(adata,marker_genes,groupby='cell_type',var_group_positions=geneGrpPos,var_group_labels=geneGrpName,row_palette='muted',rotation=90,yticklabels=True,ax=axs[1],show=False)
sc.pl.matrixplot(adata,marker_genes,groupby='cell_type',var_group_positions=geneGrpPos,var_group_labels=geneGrpName,ax=axs[2],show=False)
fig.tight_layout()
</div>

<div id="CLIVignetteCode2" style="display:none;">
from pyarrow import feather
import pandas as pd
X = adata.to_df()
meta = pd.concat([adata.obs,adata.obsm.to_df()],axis=1)
feather.write_feather(X,strPath+'.feather')
feather.write_feather(meta,strPath+".obs.feather")

%%R
# Seurat version > 3.0.2 has issues with Jupyter notebook.
#!!! Please add your installation location of Seurat v3.0.2 to the code below,
library(Seurat,lib.loc=c('/share/Seurat_3.0.2/','/usr/lib64/R/library/Seurat_3.0.2/','/usr/share/R/library','/usr/lib64/R/library'))
library(patchwork)
library(ggplot2)
D <- t(arrow::read_feather(paste(strPath,".feather",sep="")))
meta <- as.data.frame(arrow::read_feather(paste(strPath,".obs.feather",sep="")))
colnames(D) <- meta[,"name_0"]
X <- CreateSeuratObject(counts=D)
X@meta.data <- cbind(X@meta.data,meta)

%Rdevice png

%%R -r 600 -w 7200 -h 3600
X[["percent.mt"]] <- PercentageFeatureSet(X, pattern = "^MT-")
VlnPlot(X, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(X, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(X, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

%%R -r 300 -w 3000 -h 1600
X <- subset(X, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
X <- FindVariableFeatures(X, selection.method = "vst", nfeatures = 2000,verbose=F)
top10 <- head(VariableFeatures(X,verbose=F), 10)
plot1 <- VariableFeaturePlot(X)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2

%Rdevice svg

%%R -w 8 -h 4
all.genes <- rownames(X)
X <- ScaleData(X, features = all.genes,verbose=F)
X <- RunPCA(X, features = VariableFeatures(object = X))
print(X[["pca"]], dims = 1:5, nfeatures = 5)
DimPlot(X, reduction = "pca")

%Rdevice png

%%R -r 150 -w 1000 -h 1000
DimHeatmap(X, dims = 1, cells = 500, balanced = TRUE)

%%R -r 150 -w 2000 -h 2400
DimHeatmap(X, dims = 1:15, cells = 500, balanced = TRUE)

%Rdevice svg

%%R -w 6 -h 8
sink(stdout(),type="message")
suppressMessages(library(dplyr))
X <- FindNeighbors(X, dims = 1:10, verbose=FALSE)
X <- FindClusters(X, resolution = 0.5)
X.markers <- FindAllMarkers(X, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
top10 <- X.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_logFC)
DoHeatmap(X, features = top10$gene, label=FALSE) + theme(text = element_text(size = 8))
</div>
</body>
</html>
