<html>
  <head>
    <!-- check the browser -->
    <script type="text/javascript">
      var ua= navigator.userAgent, tem,
      M= ua.match(/(opera|chrome|safari|firefox|msie|trident(?=\/))\/?\s*(\d+)/i) || [];
      if(M[1]!="Chrome"){
        alert("Chrome is tested browser!")
      }else{
        if(M[2]<85){
          alert("Chrome v85 or higher is required!")
        }
      }
    </script>
    <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css">
    <link rel="stylesheet" type="text/css" href="https://interactivereport.github.io/cellxgene_VIP/static/gridstack/gridstack.min.css">
    <link rel="stylesheet" type="text/css" href="https://interactivereport.github.io/cellxgene_VIP/static/gridstack/gridstack-extra.min.css">
    <script src="https://interactivereport.github.io/cellxgene_VIP/static/gridstack/gridstack-h5.js"></script>
    <!-- for datatables -->
    <link rel="stylesheet" type="text/css" href="https://interactivereport.github.io/cellxgene_VIP/static/DataTables/datatables.min.css">
    <script src="https://interactivereport.github.io/cellxgene_VIP/static/jquery-ui.min.js"></script>
    <!-- for d3 volcano -->
    <script src="https://interactivereport.github.io/cellxgene_VIP/static/d3plot/volcanoPlot.js"></script>
    <link rel="stylesheet" href="https://interactivereport.github.io/cellxgene_VIP/static/d3plot/volcanoPlot.css">
    <!-- for gene sets -->
	  <script src="https://bxaf.net/genesets/bootstrap-3/js/bootstrap.min.js"></script>
  	<script>
		  var GENESET_ACTION_URL = "https://bxaf.net/genesets/genesets3.php";
    </script>
	  <link href="https://bxaf.net/genesets/genesets3.css" rel="stylesheet">
	  <script src="https://bxaf.net/genesets/genesets3.js"></script>
    <!-- for save & load session -->
    <script src="https://interactivereport.github.io/cellxgene_VIP/static/saveVIP.js"></script>
    <!-- <script src="static/saveVIP.js"></script> -->
    <!-- for testing the VIP -->
    <script src="https://interactivereport.github.io/cellxgene_VIP/static/testVIP.js"></script>
    <!--  <script src="static/testVIP.js"></script>-->
    <!--  end -->
    <style>
      .dropup,
      .dropdown {
        position: relative;
      }
      .dropdown-toggle:focus {
        outline: 0;
      }
      .dropdown-menu {
        position: absolute;
        top: 100%;
        left: 0;
        z-index: 1000;
        display: none;
        float: left;
        min-width: 160px;
        padding: 5px 0;
        margin: 2px 0 0;
        font-size: 14px;
        text-align: left;
        list-style: none;
        background-color: #ffffff;
        -webkit-background-clip: padding-box;
        background-clip: padding-box;
        border: 1px solid #cccccc;
        border: 1px solid rgba(0, 0, 0, 0.15);
        border-radius: 4px;
        -webkit-box-shadow: 0 6px 12px rgba(0, 0, 0, 0.175);
        box-shadow: 0 6px 12px rgba(0, 0, 0, 0.175);
      }
      .dropdown-menu.pull-right {
        right: 0;
        left: auto;
      }
      .dropdown-menu .divider {
        height: 1px;
        margin: 9px 0;
        overflow: hidden;
        background-color: #e5e5e5;
      }
      .dropdown-menu > li > a {
        display: block;
        padding: 3px 20px;
        clear: both;
        font-weight: 400;
        line-height: 1.42857143;
        color: #333333;
        white-space: nowrap;
      }
      .dropdown-menu > li > a:hover,
      .dropdown-menu > li > a:focus {
        color: #262626;
        text-decoration: none;
        background-color: #f5f5f5;
      }
      .dropdown-menu > .active > a,
      .dropdown-menu > .active > a:hover,
      .dropdown-menu > .active > a:focus {
        color: #ffffff;
        text-decoration: none;
        background-color: #337ab7;
        outline: 0;
      }
      .dropdown-menu > .disabled > a,
      .dropdown-menu > .disabled > a:hover,
      .dropdown-menu > .disabled > a:focus {
        color: #777777;
      }
      .dropdown-menu > .disabled > a:hover,
      .dropdown-menu > .disabled > a:focus {
        text-decoration: none;
        cursor: not-allowed;
        background-color: transparent;
        background-image: none;
        filter: progid:DXImageTransform.Microsoft.gradient(enabled = false);
      }
      .open > .dropdown-menu {
        display: block;
      }
      .open > a {
        outline: 0;
      }
      .dropdown-menu-right {
        right: 0;
        left: auto;
      }
      .dropdown-menu-left {
        right: auto;
        left: 0;
      }
      .dropdown-header {
        display: block;
        padding: 3px 20px;
        font-size: 12px;
        line-height: 1.42857143;
        color: #777777;
        white-space: nowrap;
      }
      .dropdown-backdrop {
        position: fixed;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: 990;
      }
      .pull-right > .dropdown-menu {
        right: 0;
        left: auto;
      }
      .dropup .caret,
      .navbar-fixed-bottom .dropdown .caret {
        content: "";
        border-top: 0;
        border-bottom: 4px dashed;
        border-bottom: 4px solid \9;
      }
      .dropup .dropdown-menu,
      .navbar-fixed-bottom .dropdown .dropdown-menu {
        top: auto;
        bottom: 100%;
        margin-bottom: 2px;
      }
      .modal-open {
        overflow: hidden;
      }
      .modal {
        position: fixed;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: 1050;
        display: none;
        overflow: hidden;
        -webkit-overflow-scrolling: touch;
        outline: 0;
      }
      .modal.fade .modal-dialog {
        -webkit-transform: translate(0, -25%);
        -ms-transform: translate(0, -25%);
        -o-transform: translate(0, -25%);
        transform: translate(0, -25%);
        -webkit-transition: -webkit-transform 0.3s ease-out;
        -o-transition: -o-transform 0.3s ease-out;
        transition: transform 0.3s ease-out;
      }
      .modal.in .modal-dialog {
        -webkit-transform: translate(0, 0);
        -ms-transform: translate(0, 0);
        -o-transform: translate(0, 0);
        transform: translate(0, 0);
      }
      .modal-open .modal {
        overflow-x: hidden;
        overflow-y: auto;
      }
      .modal-dialog {
        position: relative;
        width: auto;
        margin: 10px;
      }
      .modal-content {
        position: relative;
        background-color: #ffffff;
        -webkit-background-clip: padding-box;
        background-clip: padding-box;
        border: 1px solid #999999;
        border: 1px solid rgba(0, 0, 0, 0.2);
        border-radius: 6px;
        -webkit-box-shadow: 0 3px 9px rgba(0, 0, 0, 0.5);
        box-shadow: 0 3px 9px rgba(0, 0, 0, 0.5);
        outline: 0;
      }
      .modal-backdrop {
        position: fixed;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
        z-index: 1040;
        background-color: #000000;
      }
      .modal-backdrop.fade {
        filter: alpha(opacity=0);
        opacity: 0;
      }
      .modal-backdrop.in {
        filter: alpha(opacity=50);
        opacity: 0.5;
      }
      .modal-header {
        padding: 15px;
        border-bottom: 1px solid #e5e5e5;
      }
      .modal-header .close {
        margin-top: -2px;
      }
      .modal-title {
        margin: 0;
        line-height: 1.42857143;
      }
      .modal-body {
        position: relative;
        padding: 15px;
      }
      .modal-footer {
        padding: 15px;
        text-align: right;
        border-top: 1px solid #e5e5e5;
      }
      .modal-footer .btn + .btn {
        margin-bottom: 0;
        margin-left: 5px;
      }
      .modal-footer .btn-group .btn + .btn {
        margin-left: -1px;
      }
      .modal-footer .btn-block + .btn-block {
        margin-left: 0;
      }
      .modal-scrollbar-measure {
        position: absolute;
        top: -9999px;
        width: 50px;
        height: 50px;
        overflow: scroll;
      }

      button.link{background:none;border:none;}
      #sortable { list-style-type: none; margin: 0; padding: 0; width: 60%; }
      #sortable li { margin: 0 3px 3px 3px; padding: 0.4em; padding-left: 1.5em; font-size: 1.4em; height: 18px; cursor:move; }
      #sortable li span { position: absolute; margin-left: -1.3em; }
      #sortable li.fixed{cursor:default; color:#959595; opacity:0.5;}

      .splitter {
          display: flex;
          width: 100%;
      }

      #separator {
          cursor: col-resize;
          min-width: 6px;
          background-color: #aaa;
          background: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAhCAQAAABOpSL+AAAAIklEQVR4AWMwbb/PdR+JZDD9f1/oPhI5sgVGBSruc9xHIgGdSQqqQJGkRgAAAABJRU5ErkJggg==') center center no-repeat #F1F1F1;
      /* prevent browser's built-in drag from interfering */
          -moz-user-select: none;
          -ms-user-select: none;
          user-select: none;
      }

      #panel-left {
          background-color: #dde;
          width: 20%;
          white-space: nowrap;
      }

      #panel-right {
          background-color: #fff;
          width: 80%;
      }
      #separator, #panel-left, #panel-right {
          vertical-align: top;
      }
      #STACBARplot {
          margin: 40px;
          position: relative;
      }
      .axis path,
      .axis line {
        fill: none;
        stroke: #000;
        shape-rendering: crispEdges;
      }
      .fly-in {
        font-size: 8px;

      }
      .axis .tick line {
        stroke: #ccc;
        /*opacity: 0.5;*/
        pointer-events: none;
      }
      /*.axis .minor  line{*/
          /*stroke: red;*/
      /*}*/
      .highlight {
        font-weight: bold ;
      }
      svg {
        overflow: visible;
      }

      .block {
        display: block;
        width: 100%;
        border: none;
        background-color: #B6B6B6;
        color: white;
        padding: 4px 28px;
        font-size: 18px;
        cursor: pointer;
        text-align: center;
      }
      .block:hover {
        background-color: #ffff33;
        color: black;
      }

      .colorButton{
        color: #008cba;
      }
      .tab {
        float: left;
        width: 100%;
        height: 100%;

        border: 1px solid #ccc;
        background-color: #f1f1f1;
      }
      .tab button {
        display: block;
        width: 100%;
        text-align: left;

        background-color: inherit;
        border: none;
        outline: none;
        cursor: pointer;
        padding: 5px 10px;
        transition: 0.3s;
      }
      .tab button:hover {
        background-color: #ddd;
      }
      .tab button.active {
        background-color: #ccc;
      }
      .tabcontent {
        float: left;
        border-left: none;
        width:100%;

        display: none;
        padding: 0px 10px;
      }

      .topTab {
        overflow: hidden;
        border: 1px solid #ccc;
        background-color: #f1f1f1;
      }
      .topTab button{
        text-align: top;
        text-align: left;

        background-color: inherit;
        float: left;
        border: none;
        outline: none;
        cursor: pointer;
        padding: 5px 5px;
        transition: 0.3s;
      }
      .topTab button:hover {
        background-color: #ddd;
      }
      .topTab button.active {
        background-color: #ccc;
      }
      .DEGcontent {
        float: left;
        border-left: none;
        width:100%;
        display: none;
        padding: 0px 10px;
      }

      h3 {
        display: block;
        font-size: 1.17em;
        margin-top: 1em;
        margin-bottom: 1em;
        margin-left: 0;
        margin-right: 0;
        font-weight: bold;
      }
      .loading {
          width: 50%;
          height: auto;
          position: relative;
      }
      .loading-wheel {
          width: 70px;
          height: 70px;
          margin-top: -60px;
          margin-left: -60px;

          position: absolute;
          top: 50%;
          left: 50%;

          border-width: 30px;
          border-radius: 50%;
          -webkit-animation: spin 2s linear infinite;
      }
      .style-2 .loading-wheel {
          border-style: double;
          border-color: #99f transparent;
      }
      @-webkit-keyframes spin {
          0% {
              -webkit-transform: rotate(0);
          }
          100% {
              -webkit-transform: rotate(-360deg);
          }
      }

      tr:nth-child(even) {background-color: #F1F1F1;}
      input[type=number]{
        width: 60px;
      }
      input,textarea,select{ font-size: 14px; }
      .ace_editor p,.ace_editor div,.ace_editor input,.ace_editor label,.ace_editor span,.ace_editor text {font-family: 'Monaco', 'Menlo', 'Droid Sans Mono', 'Courier New', monospace !important;}
      pre span {font-size: 12px; font-family: 'Arial', monospace !important;}
      pre { display: block; padding: 0; margin: 0; font-size: 12px; white-space: pre-wrap; background-color: #f5f5f5; border: 1px solid #ccc; }
      table {
        border-collapse: collapse;
      }
      th {
        background-color: #D4D8DE;
        color: black;
      }
      .switch {
        position: relative;
        display: inline-block;
        width: 40px;
        height: 22px;
      }

      .switch input {
        opacity: 0;
        width: 0;
        height: 0;
      }

      .slider {
        position: absolute;
        cursor: pointer;
        top: 0;
        left: 0;
        right: 0;
        bottom: 0;
        background-color: #ccc;
        -webkit-transition: .4s;
        transition: .4s;
      }

      .slider:before {
        position: absolute;
        content: "";
        height: 18px;
        width: 18px;
        left: 2px;
        bottom: 2px;
        background-color: white;
        -webkit-transition: .4s;
        transition: .4s;
      }

      input:checked + .slider {
        background-color: #2196F3;
      }

      input:focus + .slider {
        box-shadow: 0 0 1px #2196F3;
      }

      input:checked + .slider:before {
        -webkit-transform: translateX(18px);
        -ms-transform: translateX(18px);
        transform: translateX(18px);
      }
      .grid-stack {
        background: #F1F1F1;
      }
      .grid-stack-item-content {
        color: #2c3e50;
        text-align: center;
        background-color: #CCCCCC;
      }
      .colorInput{
          -webkit-appearance: none;
          -moz-appearance: none;
          appearance: none;
          width: 20px;
          height: 22px;
          background-color: transparent;
          border: none;
          cursor: pointer;
      }
      figure.zoom{
          & img:hover{
            opacity: 0;
          }
          img {
              transition: opacity .5s;
              display: block;
              width: 100%;
          }
          background-position: 50% 50%;
          position: relative;
          overflow: hidden;
          width: 100%;
      }
    </style>
  </head>
<body>
<div class="splitter">
  <div id="panel-left">
    <div class="tab">
      <button class="tablinks" onclick="selFun(event, 'ADDG')"><b>Add Genes</b></button>
      <button class="tablinks" onclick="selFun(event, 'SGV');defaultTopSel('SGV','SGVsingle')">Violin</button>
      <button class="tablinks" onclick="selFun(event, 'HEAT')">Heatmap</button>
      <button class="tablinks" onclick="selFun(event, 'EMBED')">Embedding Plot</button>
      <button class="tablinks" onclick="selFun(event, 'DOT')">Dot Plot</button>
      <button class="tablinks" onclick="selFun(event, 'TRAK')">Track Plot</button>
      <button class="tablinks" onclick="selFun(event, 'DENS')">Density Plot</button>
      <button class="tablinks" onclick="selFun(event, 'DENS2D')">Density Scatter</button>
      <button class="tablinks" onclick="selFun(event, 'DUAL')">Dual Genes</button>
      <button class="tablinks" onclick="selFun(event, 'SANK')">Sankey Diagram</button>
      <button class="tablinks" onclick="selFun(event, 'STACBAR')">Stacked Barplot</button>
      <button class="tablinks" onclick="selFun(event, 'GSP')">Gene Specificity</button>
      <button class="tablinks" onclick="selFun(event, 'GD')">Gene Detected</button>
      <button class="tablinks" onclick="selFun(event, 'DEG');defaultTopSel('DEGtop','DEGtopBT1')">DEG</button>
      <button id="preDEGpanelBT" class="tablinks" onclick="selFun(event, 'preDEGpanel');defaultTopSel('DEG','DEGlinksPlot')" style="display: none;">Pre-Computed DEG</button>
      <button class="tablinks" onclick="selFun(event, 'MARK')">Marker Genes</button>
      <button id="browserBT" class="tablinks" onclick="selFun(event, 'BROW')" style="display: none;">Multiome</button>
      <button id='cosMxBT' class="tablinks" onclick="selFun(event, 'COSMX');" style="display: none;">CosMx</button>
      <button id='visiumBT' class="tablinks" onclick="selFun(event, 'VISIUM');" style="display: none;">Visium</button>
      <button class="tablinks" onclick="selFun(event, 'SPATIAL');defaultTopSel('SPATIAL','SPATIALimage')">Spatial Transcriptomics</button>
      <button class="tablinks" onclick="selFun(event, 'CLI')">Command Line Interface</button>
      <button id="metaCellBT" class="tablinks colorButton" onclick="metaSel()" style="display: none;">Show MetaCell</button>
      <div id="metaCellBTspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <hr size="1" noshade>
      <button class="tablinks" onclick="selFun(event, 'IMG')">Global Setting</button>
      <button class="tablinks" onclick="selFun(event, 'ABBR')">Comb. & Abbr.</button>
      <button class="tablinks" onclick="saveLocal()">Save Session</button>
      <button class="tablinks" onclick="selFun(event, 'LOAD')">Load Session</button>
      <button class="tablinks" onclick="ABBRall()">Check All Annotations</button>
      <button class="tablinks" onclick="selFun(event, 'BRUSH');">Brush</button>
      <button class="tablinks" onclick="sync();">Sync</button>
      <button id="VIPtestBT" class="tablinks" onclick="selFun(event, 'tVIP')" style="display: none;">Test</button>
    </div>
  </div>
  <div id="separator" ></div>
  <div id="panel-right" >
    <div id="ABBR" class="tabcontent"> <!-- Abbreviation -->
      <h3>Create a combinatorial annotation <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-comb.-abbr." target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
      <p>First, check below categorical annotations to be combined. Second, select items in each annotation. Then, click <button onclick="ABBRcombineNew()">Combine</button>. The max combinatorial number <input type="number" id="ABBRcomBmax" min=2 value="200"></p>
      <p id="ABBRcomBgrp"></p>
      <p id="ABBRcomBnote"></p>
      <p id="ABBRcomBtable"></p>
      <!--
      <ol type="1">
        <li><button onclick="ABBRclear()">Uncheck All Annotations</button></li>
        <li><p>Select the annotations of interest on the left panel of main window, then
          <button onclick="ABBRcombine();ABBRall()">Create</button> a new annotation named "Custom_combine" includes the
          following categories:  <p id="ABBRcombine"></p>
		    </li>
      </ol>
      -->
      <div id="ABBRcombineList" style="display:none;">
        <p>It will become categorical annotation named "Custom_combine". Change the order by dragging items up or down:</p>
        <ul id="ABBRdynList" class="sortable-list"></ul><!--   -->
      </div>


      <hr>
      <h3>Create abbreviations of annotations</h3><!-- <button onclick="ABBRsave()">Save to local</button><input id="ABBRfile" type="file" value="Load from local" onchange="ABBRload()"/>-->
      <p>Choose annotation(s):</p>
      <p id="ABBRgrp"></p>
      <p id="ABBRtable"></p>
    </div><!--ABBR-->

    <div id="ADDG" class="tabcontent">
      <h3>Add genes of interest for plotting <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-add-genes-gene-sets" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
      <p>Please input a list of genes separated by comma, semicolon, space or tab:</p>
      <p><textarea id="ADDGinput" rows="5" cols="70"></textarea></p>
      <p><button onclick="ADDGadd()"><b>Add</b></button></p>
      <p>The following genes were added (click on a gene name to remove): <span id="ADDGsel"></span></p>
      <p>The following genes are not in the data: <span id="ADDGdel" style="color:red"></span></p>

      <h3>User-defined gene sets:</h3>
      <p id="ADDGsets"></p>

      <h4>Add gene sets of interest manually</h4>
      <p>Gene set name: <input id="Geneset_Name0" type="text" size=10/>
        Gene names: <input id="Geneset_Genes0" type="text" size=35/>
        <button onclick="ADDGsetAdd(0)"><b>Add</b></button>
      </p>
      
      <h4>Load gene sets from a text file</h4>
      <p>geneset1: ACTB, Gapdh<br>geneset2: BTK, CD4<br>
        <input type='file' id='Geneset_file' onchange=ADDGgsetFile(this) /></p>

      <h4>Add gene sets from databases</h4>
      <div id="div_geneset1" class="div_geneset">
  		<div class="my-3 dropdown input-group">
  			<input id="Geneset_Name1" name="Geneset_Name1" placeholder="enter geneset name" type="text" class="form-control geneset_name dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"/>
  			<button class="input-group-addon btn_browse_geneset"><b>Search</b></button>
  			<div id="Dropdown1" class="geneset_dropdown dropdown-menu" aria-labelledby="Geneset_Name1"></div>
  		</div>
  		<input class="geneset_id" type="hidden" id="Geneset_ID1" name="Geneset_ID1" />
  		<textarea class="form-control my-3 geneset_genes" style="width: 30rem; margin-top: 1rem;" rows="5" id="Geneset_Genes1" name="Geneset_Genes1" placeholder="Here will list all genes in a geneset"></textarea>
  		<p><button onclick="ADDGsetAdd(1)"><b>Add</b></button></p>
  	</div>

      <span id="ADDGsetError" style="color:red"></span><br>

    </div><!--ADDG-->

    <div id="GD" class=tabcontent>
      <div id="GDinput">
        <h3>Genes detected in cell selections <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-gene-detected" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>In selection 1 (<span class="cellN1" style="color:red;">0</span>) cells and selection 2 (<span class="cellN2" style="color:red;">0</span>) cells</p>
        <p>Expression cutoff:
        <input id="GDcutoff" type="number" step="0.1" value="1"/></p>
      </div>
      <button class="GDbt" onclick="GDplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('GDresize')">Zoom In</button>
      <button onclick="ZoomOut('GDresize')">Zoom Out</button>
      <button class="GDpngBT" onclick="saveImg('GDimg')" disabled>Save</button>
      <button onclick="hideShow($('#GDinput'),$(this));">Hide Input</button>
      <pre id="GDnote" style="color:red;"></pre>
      <div id="GDspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="GDresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
        <p id="GDfig"></p></div>
    </div><!--GD-->

    <div id="MARK" class="tabcontent">
      <div id="MARKinput">
        <h3>Marker genes of annotation categories from selected cells <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-marker-genes" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="MARKcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="MARKcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Select an annotation:
        <select id="MARKgrp"></select></p>
        <p>By method:
        <select id="MARKmethod">
          <option value="logreg">logreg</option>
          <option value="t-test">t-test</option>
          <option value="wilcoxon">wilcoxon</option>
          <option value="t-test_overestim_var">t-test_overestim_var</option>
        </select> with the number of genes per group: <input id="MARKn" type="number" step="1" value="10" min="3"/></p>
      </div>
      <button class="MARKbt" onclick="MARKplot()"><b>Find/Plot</b></button>
      <button onclick="ZoomIn('MARKresize')">Zoom In</button>
      <button onclick="ZoomOut('MARKresize')">Zoom Out</button>
      <button class="MARKpngBT" onclick="saveImg('MARKimg')" disabled>Save</button>
      <button onclick="hideShow($('#MARKinput'),$(this));">Hide Input</button>
      <pre id="MARKnote" style="color:red;"></pre>
      <div id="MARKspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>

      <table id="MARKtable" class="display"></table>
      <div id="MARKresize" style="height: 'auto'">
        <p id="MARKfig"></p>
      </div>
    </div><!--MARK-->

    <div id="SGV" class="tabcontent">
      <div class="topTab">
        <button id="SGVsingle" class="SGVlinks" onclick="selTop(event, 'SGVsingleV','SGV')">Single Gene</button>
        <button id="SGVcompare" class="SGVlinks" onclick="selTop(event, 'SGVcompareV','SGV')">Single Gene/Multi-factors</button>
        <button id="SGVcompare" class="SGVlinks" onclick="selTop(event, 'PGV','SGV')">Multi-genes</button>
        <button id="SGVcompare" class="SGVlinks" onclick="selTop(event, 'PGVcompareV','SGV')">Multi-genes/Multi-factors</button>
      </div>

      <div id="SGVsingleV" class="SGVcontent">
        <h3>Violin plot of a gene from selected cells <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-violin-plot" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <div id="SGVinput">
          <p>Choose cell selection(s):
            <label><input class="SGVcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
            <label><input class="SGVcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
          </p>
          <p>Choose a gene ('Add Genes' first): <select id="SGVgene"></select></p>
          <p>Expression cutoff for cell filtering: <input id="SGVcutoff" type="number" step="0.1" value="1" min="0"/></p>
          <p>Group by:
          <select id="SGVgrp"></select></p>
        </div>
        <button class="SGVbt" onclick="SGVplot('plot')"><b>Plot</b></button>
        <button class="SGVsaveBT" onclick="SGVplot('data')">Get Data</button>
        <button onclick="ZoomIn('SGVresize')">Zoom In</button>
        <button onclick="ZoomOut('SGVresize')">Zoom Out</button>
        <button class="SGVpngBT" onclick="saveImg('SGVimg')" disabled>Save</button>
        <button onclick="hideShow($('#SGVinput'),$(this));">Hide Input</button>
        <pre id="SGVnote" style="color:red;"></pre>
        <div id="SGVspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
        <div id="SGVresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="SGVfig"></p></div>
      </div> <!-- SGVsingleV -->

      <div id="SGVcompareV" class="SGVcontent">
        <h3>Plot sub-grouped violin <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <div id="SGVcompareinput">
          <p>Choose cell selection(s):
            <label><input class="SGVcomparecell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
            <label><input class="SGVcomparecell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
          </p>
          <p>Choose a gene ('Add Genes' first): <select id="SGVcomparegene"></select></p>
          <p>Expression cutoff (for cell filtering): <input id="SGVcompareCellcutoff" type="number" step="0.01" value="0.01" min="0"/></p>
          <p>Expression cutoff (for percentage calculation): <input id="SGVcomparecutoff" type="number" step="0.1" value="1" min="0"/></p>
          <p>Group by:
          <select id="SGVcompareorder"></select></p>
          <p>Sub-group by:
          <select id="SGVcomparegrp"></select></p>
        </div>
        <button class="SGVcomparebt" onclick="SGVcompareplot()"><b>Plot</b></button> <!-- 'plot' -->
        <!-- <button class="SGVsaveBT" onclick="SGVplot('data')">Get Data</button> -->
        <button onclick="ZoomIn('SGVcompareresize')">Zoom In</button>
        <button onclick="ZoomOut('SGVcompareresize')">Zoom Out</button>
        <button class="SGVcomparepngBT" onclick="saveImg('SGVcompareimg')" disabled>Save</button>
        <button onclick="hideShow($('#SGVcompareinput'),$(this));">Hide Input</button>
        <pre id="SGVcomparenote" style="color:red;"></pre>
        <div id="SGVcomparespin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
        <div id="SGVcompareresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="SGVcomparefig"></p></div>

      </div> <!-- SGVcompareV -->

      <div id="PGV" class="SGVcontent">
        <h3>Stacked violin plot of genes from selected cells <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-stacked-violin" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <div id="PGVinput">
          <p>Choose cell selection(s):
            <label><input class="PGVcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
            <label><input class="PGVcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
          </p>
          <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.PGVgenes').prop('checked', this.checked);">Uncheck / Check All</p>
          <p id="PGVgene"></p>
          <p>Expression cutoff for cell filtering: <input id="PGVcutoff" type="number" step="0.1" value="1"/></p>
          <p>Select user-defined gene sets: <input type="checkbox" checked onclick="$('.PGVsets').prop('checked', this.checked);">Uncheck / Check All</p><p id="PGVgeneGrp"></p>
          <div id="PVGgeneGrpCollapsing">
            <p>Collapsing the above gene sets:
              <input type="radio" name="PGVgeneGrpColl" value="Yes" onclick="$('#PGVgeneGrpCollShow').show()">Yes
              <input type="radio" name="PGVgeneGrpColl" value="No" onclick="$('#PGVgeneGrpCollShow').hide()" checked>No
            </p>
            <div id='PGVgeneGrpCollShow'><p>By:
              <input type="radio" name="PGVgeneGrpCollCal" value="mean" checked>Average
              <input type="radio" name="PGVgeneGrpCollCal" value="median">Median</p>
            </div>
          </div>
          <p>Group by:<select id="PGVgrp"></select> as<select id="PGVgrpXY">
            <option value="Rows">Rows</option>
            <option value="Columns">Columns</option>
          </select></p>
          <p>Color:<select id="PGVcolor">
            <option value='Purples'>Purples</option>
            <option value='Blues'>Blues</option>
            <option value='Greens'>Greens</option>
            <option value='Oranges'>Oranges</option>
            <option value='Reds' selected>Reds</option>
            <option value='YlOrBr'>YlOrBr</option>
            <option value='YlOrRd'>YlOrRd</option>
            <option value='OrRd'>OrRd</option>
            <option value='PuRd'>PuRd</option>
            <option value='RdPu'>RdPu</option>
            <option value='BuPu'>BuPu</option>
            <option value='PuBu'>PuBu</option>
            <option value='PuBuGn'>PuBuGn</option>
            <option value='BuGn'>BuGn</option>
            <option value='GnBu'>GnBu</option>
            <option value='YlGnBu'>YlGnBu</option>
            <option value='YlGn'>YlGn</option>
            <option value='Spectral'>Spectral</option>
            <option value='PiYG'>PiYG</option>
            <option value='PRGn'>PRGn</option>
            <option value='BrBG'>BrBG</option>
            <option value='PuOr'>PuOr</option>
            <option value='RdGy'>RdGy</option>
            <option value='RdBu'>RdBu</option>
            <option value='RdYlBu'>RdYlBu</option>
            <option value='RdYlGn'>RdYlGn</option>
          </select> <input id="PGVcolor_r" type="checkbox">Reverse <button onclick="hideShow($('#PGVcolorhelp'),$(this),'Color Help');">Show Color Help</button></p>
          <div id="PGVcolorhelp" style="display:none;">
                <table border=0>
                  <tr valign="top"><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_sequential.png" width="300"></td><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_diverging.png" width="300"></td></tr>
                </table>
          </div>
        </div>

        <button class="PGVbt"  onclick="PGVplot('plot')"><b>Plot</b></button>
        <button class="PGVsaveBT" onclick="PGVplot('data')">Get Data</button>
        <button onclick="ZoomIn('PGVresize')">Zoom In</button>
        <button onclick="ZoomOut('PGVresize')">Zoom Out</button>
        <button class="PGVpngBT" onclick="saveImg('PGVimg')" disabled>Save</button>
        <button onclick="hideShow($('#PGVinput'),$(this));">Hide Input</button>
        <pre id="PGVnote" style="color:red;"></pre>
        <div id="PGVspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
        <div id="PGVresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="PGVfig"></p></div>
      </div><!--PGV-->

      <div id="PGVcompareV" class="SGVcontent">
        <h3>Violins of selected genes splitted on two group annotations <small><a href="https://github.com/TheHumphreysLab/plot1cell">plot1cell</a></small></h3>
        <div id="PGVcompareinput">
          <p>Choose cell selection(s):
            <label><input class="PGVcomparecell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
            <label><input class="PGVcomparecell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label></p>
          <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.PGVcomparegenes').prop('checked', this.checked);">Uncheck / Check All</p>
          <p id="PGVcomparegene"></p>
          <p>Expression cutoff for cell filtering: <input id="PGVcomparecutoff" type="number" step="0.1" value="0.5"/></p>
          <p>Group by (row panels):<select id="PGVcompareRow"></select></p>
          <p>Sub-group by (within a subpanel):<select id="PGVcompareGrp"></select></p>
          <p>Width: <input id="PGVcomparewidth" type="number" step="0.5" value="6" min="1" max="24"/>
             Height: <input id="PGVcompareheight" type="number" step="0.5" value="8" min="1" max="24"/>
          </p>
        </div>

        <button class="PGVcomparebt"  onclick="PGVcomparebt()"><b>Plot</b></button>
        <button onclick="ZoomIn('PGVcompareresize')">Zoom In</button>
        <button onclick="ZoomOut('PGVcompareresize')">Zoom Out</button>
        <button class="PGVcomparepngBT" onclick="saveImg('PGVcompareimg')" disabled>Save</button>
        <button onclick="hideShow($('#PGVcompareinput'),$(this));">Hide Input</button>
        <pre id="PGVcomparenote" style="color:red;"></pre>
        <div id="PGVcomparespin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
        <div id="PGVcompareresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="PGVcomparefig"></p></div>

      </div>

    </div><!--SGV panel-->

    <div id="HEAT" class="tabcontent">
      <div id="HEATinput">
        <h3>Heatmap of selected cells and genes <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-heatmap" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="HEATcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="HEATcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input id='HEATgenesAll' type="checkbox" checked onclick="$('.HEATgenes').prop('checked', this.checked);">Uncheck / Check All <span id="HEATgeneAnno"><input id="HEATgeneAnnoCK" type="checkbox" checked>Using protein atlas annotation</span></p>
        <p id="HEATgene"></p>
        <p>Or upload a gene annotation csv file (first column header is 'gene'): <input type='file' id='HEATgeneFile' onchange="HEATgeneFileUpdate()" /></p>
        <p>Choose categorical annotation(s):</p>
        <p id="HEATgrp"></p>
        <p>Choose continuous annotation(s):</p>
        <p id="HEATmeta"></p>
        <p>Select plotting value:
        <select id="HEATnorm">
          <option value="X">Expression</option>
          <option value="zscore">Expression Z-score</option>
        </select></p>
        <p>Color:
        <select id="HEATcolor">
          <option value='Purples'>Purples</option>
          <option value='Blues'>Blues</option>
          <option value='Greens'>Greens</option>
          <option value='Oranges'>Oranges</option>
          <option value='Reds' selected>Reds</option>
          <option value='YlOrBr'>YlOrBr</option>
          <option value='YlOrRd'>YlOrRd</option>
          <option value='OrRd'>OrRd</option>
          <option value='PuRd'>PuRd</option>
          <option value='RdPu'>RdPu</option>
          <option value='BuPu'>BuPu</option>
          <option value='PuBu'>PuBu</option>
          <option value='PuBuGn'>PuBuGn</option>
          <option value='BuGn'>BuGn</option>
          <option value='GnBu'>GnBu</option>
          <option value='YlGnBu'>YlGnBu</option>
          <option value='YlGn'>YlGn</option>
          <option value='Spectral'>Spectral</option>
          <option value='PiYG'>PiYG</option>
          <option value='PRGn'>PRGn</option>
          <option value='BrBG'>BrBG</option>
          <option value='PuOr'>PuOr</option>
          <option value='RdGy'>RdGy</option>
          <option value='RdBu'>RdBu</option>
          <option value='RdYlBu'>RdYlBu</option>
          <option value='RdYlGn'>RdYlGn</option>
        </select> <input id="HEATcolor_r" type="checkbox">Reverse <button onclick="hideShow($('#HEATcolorhelp'),$(this),'Color Help');">Show Color Help</button></p>
    		<div id="HEATcolorhelp" style="display:none;">
              <table border=0>
                <tr valign="top"><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_sequential.png" width="300"></td><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_diverging.png" width="300"></td></tr>
              </table>
    		</div>
    		<p>Select plotting methods:
    		  <select id="HEATplotMethod" onchange="if($(this).val()=='cHeatmap'){$('#HEATcomplexSetting').show();$('#HEATseabornSetting').hide()}else{$('#HEATcomplexSetting').hide();$('#HEATseabornSetting').show()}">
    		    <option value="cHeatmap">ComplexHeatmap (R)</option>
    		    <option value="sns">Seaborn (python)</option>
    		  </select></p>
        <div id="HEATcomplexSetting">
          <p>Gene as
            <input type="radio" name="HEATswapAxes" value="F" checked onchange="$('#HEATcomplexSettingSwap').hide()">Column
            <input type="radio" name="HEATswapAxes" value="T" onchange="$('#HEATcomplexSettingSwap').show()">Row
            <span id="HEATcomplexSettingSwap" style="display:none"> || Max legend row: <input id="HEATcomplexSettingLegendRow" type="number" step="1" value="5" min="1" max="20"/></span>
          </p>
          <p>Width: <input id="HEATwidth" type="number" step="0.5" value="6" min="1" max="24"/>
             Height: <input id="HEATheight" type="number" step="0.5" value="8" min="1" max="24"/>
          </p>
          Annotation fontsize adjust: <input id="HEATcomplexSettingFontadj" type="number" step="1" value="-4" min="-10" max="10"/>
          <p>
          </p>
          <p>Cluster by:
            <select id="HEATcomplexSortby" onchange="if($(this).val()=='Annotation'){$('#HEATorderGrp').show()}else{$('#HEATorderGrp').hide()}">
              <option value="Annotation">Annotation</option>
              <option value="Expression">Expression</option>
            </select>
          </p>
          <div id='HEATorderGrp'>
            <p>Change the annotation order by dragging items up or down, cluster annotations can also be removed by clicking <i class='fa fa-trash-o'></i>:</p>
            <ul id="HEATdynList" class="sortable-list"></ul><!--   -->
          </div>

        </div>
        <div id="HEATseabornSetting" style="display:none">
          <p>Cluster by:
            <select id="HEATorder"></select></p>
        </div>
      </div>
      <button class="HEATbt"  onclick="HEATplot()"><b>Plot</b></button>
      <button class="HEATsaveBT" onclick="HEATdata()" disabled>Get Data</button>
      <button onclick="ZoomIn('HEATresize')">Zoom In</button>
      <button onclick="ZoomOut('HEATresize')">Zoom Out</button>
      <button class="HEATpngBT" onclick="saveImg('HEATimg')" disabled>Save</button>
      <button onclick="hideShow($('#HEATinput'),$(this));">Hide Input</button>
      <pre id="HEATnote" style="color:red;"></pre>
      <div id="HEATspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="HEATresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="HEATfig"></p></div>
      <div id="HEATdata" style="visibility: hidden;"></div>

    </div><!--HEAT-->

    <div id="GSP" class="tabcontent">
      <h3>Gene specificity for groups in an annotation</h3>
      <div id="GSPinput">
        <p>Choose cell selection(s):
          <label><input class="GSPcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="GSPcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes (If no gene select, all gene will be used.): <input id='GSPgenesAll' type="checkbox" checked onclick="$('.GSPgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="GSPgene"></p>
        Select an annotation for specificity:
        <select id="GSPgrp"></select></p>
      </div>
      <button class="GSPbt" onclick="GSPplot()"><b>Plot</b></button>
      <!-- <button class="SGVsaveBT" onclick="SGVplot('data')">Get Data</button> -->
      <button onclick="ZoomIn('GSPresize')">Zoom In</button>
      <button onclick="ZoomOut('GSPresize')">Zoom Out</button>
      <button class="GSPpngBT" onclick="saveImg('GSPimg')" disabled>Save</button>
      <button onclick="hideShow($('#GSPinput'),$(this));">Hide Input</button>
      <pre id="GSPnote" style="color:red;"></pre>
      <div id="GSPspin" class="loading style-2" style="display:none"><div class="loading-wheel"></div></div>
      <!-- <table id="GSPtable" class="display"></table> -->
      <div id="GSPtableDIV"></div>
      <div id="GSPresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="GSPfig"></p></div>
    </div><!-- Gene specificity -->

    <div id="DOT" class="tabcontent">
      <div id="DOTinput">
        <h3>Dot plot shows per group, the fraction of cells expressing a gene (dot size) and the mean expression of the gene in those cell (color scale) <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-dot-plot" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="DOTcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="DOTcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.DOTgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="DOTgene"></p>
        <p>Expression cutoff for cell fraction (dot size): <input id="DOTcutoff" type="number" step="0.1" value="1"/></p>
        <p>Expression is averaged only over cells expressing a given gene above the cutoff: <input type="radio" name="mean_only_expressed" value="Yes">Yes<input type="radio" name="mean_only_expressed" value="No" checked>No</p>

        <p>Select user-defined gene sets: <input type="checkbox" checked onclick="$('.DOTsets').prop('checked', this.checked);">Uncheck / Check All</p><p id="DOTgeneGrp"></p>
        <div id="DOTgeneGrpCollapsing">
          <p>Collapsing the above gene sets:
            <input type="radio" name="DOTgeneGrpColl" value="Yes" onclick="$('#DOTgeneGrpCollShow').show()">Yes
            <input type="radio" name="DOTgeneGrpColl" value="No" onclick="$('#DOTgeneGrpCollShow').hide()" checked>No
          </p>
          <div id='DOTgeneGrpCollShow'><p>By:
            <input type="radio" name="DOTgeneGrpCollCal" value="mean" checked>Average
            <input type="radio" name="DOTgeneGrpCollCal" value="median">Median</p>
          </div>
        </div>

        <p>Group by: <select id="DOTgrp"></select></p>
        <p>Adjust the legend width: <input id="DOTlegendW" type="number" step="0.1" value="1.5"/></p>
        <p>Color:
        <select id="DOTcolor">
          <option value='Purples'>Purples</option>
          <option value='Blues'>Blues</option>
          <option value='Greens'>Greens</option>
          <option value='Oranges'>Oranges</option>
          <option value='Reds' selected>Reds</option>
          <option value='YlOrBr'>YlOrBr</option>
          <option value='YlOrRd'>YlOrRd</option>
          <option value='OrRd'>OrRd</option>
          <option value='PuRd'>PuRd</option>
          <option value='RdPu'>RdPu</option>
          <option value='BuPu'>BuPu</option>
          <option value='PuBu'>PuBu</option>
          <option value='PuBuGn'>PuBuGn</option>
          <option value='BuGn'>BuGn</option>
          <option value='GnBu'>GnBu</option>
          <option value='YlGnBu'>YlGnBu</option>
          <option value='YlGn'>YlGn</option>
          <option value='Spectral'>Spectral</option>
          <option value='PiYG'>PiYG</option>
          <option value='PRGn'>PRGn</option>
          <option value='BrBG'>BrBG</option>
          <option value='PuOr'>PuOr</option>
          <option value='RdGy'>RdGy</option>
          <option value='RdBu'>RdBu</option>
          <option value='RdYlBu'>RdYlBu</option>
          <option value='RdYlGn'>RdYlGn</option>
        </select>  <input id="DOTcolor_r" type="checkbox">Reverse <button onclick="hideShow($('#DOTcolorhelp'),$(this),'Color Help');">Show Color Help</button></p>
    		<div id="DOTcolorhelp" style="display:none;">
              <table border=0>
                <tr valign="top"><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_sequential.png" width="300"></td><td><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_brewer_diverging.png" width="300"></td></tr>
              </table>
    		</div>

        <pre id="DOTnote" style="color:red;">Note: Unscaled data is used as the visualiztion is hard to interpret for scaled or corrected matrices in which zero counts had been replaced by other values. Scaling option in Global Setting does not apply to Dot plot</pre>
      </div>
      <button class="DOTbt" onclick="DOTplot()"><b>Plot</b></button>
      <button class="DOTbt" onclick="ZoomIn('DOTresize')">Zoom In</button>
      <button class="DOTbt" onclick="ZoomOut('DOTresize')">Zoom Out</button>
      <button class="DOTpngBT" onclick="saveImg('DOTimg')" disabled>Save</button>
      <button onclick="hideShow($('#DOTinput'),$(this));">Hide Input</button>
      <div id="DOTspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="DOTresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="DOTfig"></p></div>

    </div><!--DOT-->

    <div id="EMBED" class="tabcontent">
      <div id="EMBEDinput">
        <h3>Embedding plots of selected genes and/or annotations <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-umaptsne" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="EMBEDcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="EMBEDcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose categorical annotation(s):</p>
        <p id="EMBEDgrp"></p>
        <p>Choose continuous annotation(s):</p>
        <p id="EMBEDmeta"></p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.EMBEDgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="EMBEDgene"></p>
        <p>Split by:
        <select id="EMBEDsplit">
          <option value="NONE">NONE</option>
        </select>
        </p>
        <p>Select the embedding layout:
        <select id="EMBEDlayout">
        </select> the number of plots per row:
        <input id="EMBEDslot" type="number" value="4" min="2"/>
        </p>
      </div>
      <button class="EMBEDbt" onclick="EMBEDplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('EMBEDresize')">Zoom In</button>
      <button onclick="ZoomOut('EMBEDresize')">Zoom Out</button>
      <button class="EMBEDpngBT" onclick="saveImg('EMBEDimg')" disabled>Save</button>
      <button onclick="hideShow($('#EMBEDinput'),$(this));">Hide Input</button>
      <pre id="EMBEDnote" style="color:red;"></pre>
      <div id="EMBEDspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="EMBEDresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="EMBEDfig"></p></div>

    </div><!--EMBED-->

    <div  id="TRAK" class="tabcontent">
      <div id="TRAKinput">
        <h3>Track plot represents the gene expression by height <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-track-plot" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="TRAKcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="TRAKcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.TRAKgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="TRAKgene"></p>
        <p>Select user-defined gene sets: <input type="checkbox" checked onclick="$('.TRAKsets').prop('checked', this.checked);">Uncheck / Check All</p><p id="TRAKgeneGrp"></p>
        <div id="TRAKgeneGrpCollapsing">
          <p>Collapsing the above gene sets:
            <input type="radio" name="TRAKgeneGrpColl" value="Yes" onclick="$('#TRAKgeneGrpCollShow').show()">Yes
            <input type="radio" name="TRAKgeneGrpColl" value="No" onclick="$('#TRAKgeneGrpCollShow').hide()" checked>No
          </p>
          <div id='TRAKgeneGrpCollShow'><p>By:
            <input type="radio" name="TRAKgeneGrpCollCal" value="mean" checked>Average
            <input type="radio" name="TRAKgeneGrpCollCal" value="median">Median</p>
          </div>
        </div>

        <p>Group by: <select id="TRAKgrp"></select></p>
      </div>
      <button class="TRAKbt" onclick="TRAKplot()"><b>Plot</b></button>
      <button class="TRAKbt" onclick="ZoomIn('TRAKresize')">Zoom In</button>
      <button class="TRAKbt" onclick="ZoomOut('TRAKresize')">Zoom Out</button>
      <button class="TRAKpngBT" onclick="saveImg('TRAKimg')" disabled>Save</button>
      <button onclick="hideShow($('#TRAKinput'),$(this));">Hide Input</button>
      <pre id="TRAKnote" style="color:red;"></pre>
      <div id="TRAKspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="TRAKresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="TRAKfig"></p></div>

    </div><!--TRAK-->

    <div  id="DENS" class="tabcontent">
      <div id="DENSinput">
        <h3>Density plots of selected genes <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-density-plot" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="DENScell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="DENScell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.DENSgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="DENSgene"></p>
        <p>Split by:
        <select id="DENSsplit"></select></p>
        <p>Color by:
        <select id="DENSgrp"></select></p>
        <p>Bandwidth: <input id="DENSprec" type="number" step="0.1" value="0.5" min="0.1"/> controls how tightly the curve fits to the data, like the bin size in a histogram.</p>
      </div>
      <button class="DENSbt" onclick="DENSplot()"><b>Plot</b></button>
      <button class="DENSbt" onclick="ZoomIn('DENSresize')">Zoom In</button>
      <button class="DENSbt" onclick="ZoomOut('DENSresize')">Zoom Out</button>
      <button class="DENSpngBT" onclick="saveImg('DENSimg')" disabled>Save</button>
      <button onclick="hideShow($('#DENSinput'),$(this));">Hide Input</button>
      <pre id="DENSerror" style="color:red;"></pre>
      <div id="DENSspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="DENSresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="DENSfig"></p></div>

    </div><!--DENS-->

    <div  id="DENS2D" class="tabcontent">
      <div id="DENS2Dinput">
        <h3>Density Scatter plot shows the expression relationship of two genes. The color of each hexagon denotes the number of cells in it. <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-2d-density-plot" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="DENS2Dcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="DENS2Dcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose two genes ('Add Genes' first):</p>
        <p id="DENS2Dgene"></p>
        <p>Expression cutoff: <input id="DENS2Dcutoff" type="number" step="0.1" value="1"/></p>
        <p>Number of hexagons: <input id="DENS2Dbandwidth" type="number" step="1" value="50"/> The higher the number, the less cells in a hexagon.</p>
      </div>
      <button class="DENS2Dbt"  onclick="DENS2Dplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('DENS2Dresize')">Zoom In</button>
      <button onclick="ZoomOut('DENS2Dresize')">Zoom Out</button>
      <button class="DENS2DpngBT" onclick="saveImg('DENS2Dimg')" disabled>Save</button>
      <button onclick="hideShow($('#DENS2Dinput'),$(this));">Hide Input</button>
      <pre id="DENS2Dnote" style="color:red;"></pre>
      <div id="DENS2Dspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="DENS2Dresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="DENS2Dfig"></p></div>

    </div><!--DENS2D-->

    <div  id="DUAL" class="tabcontent">
      <div id="DUALinput">
        <h3>Embedding plot of two genes from selected cells <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-dual-genes" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="DUALcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="DUALcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p id="DUALgene"></p>
        <p>Expression cutoff: <input id="DUALcutoff" type="number" step="0.1" value="1"/></p>
        <p>Select the embedding layout:
        <select id="DUALlayout">
        </select>
        </p>
      </div>
      <button class="DUALbt"  onclick="DUALplot()"><b>Plot</b></button>
      <button onclick="ZoomIn('DUALresize')">Zoom In</button>
      <button onclick="ZoomOut('DUALresize')">Zoom Out</button>
      <button class="DUALpngBT" onclick="saveImg('DUALimg')" disabled>Save</button>
      <button onclick="hideShow($('#DUALinput'),$(this));">Hide Input</button>
      <pre id="DUALnote" style="color:red;"></pre>
      <div id="DUALspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="DUALresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6"><p id="DUALfig"></p></div>

    </div><!--DUAL-->

    <div  id="SANK" class="tabcontent">
      <div id="SANKinput">
        <h3>Interactive sankey diagram of annotations and/or genes from selected cells <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-sankey-diagram" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="SANKcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="SANKcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.SANKgenes').prop('checked', this.checked); SANKorderUpdate()">Uncheck / Check All</p>
        <p id="SANKgene"></p>
        <!-- <p>Choose user-defined gene sets:</p><p id="SANKgeneGrp"></p> -->
        <p>Number of equally divided gene expression ranges: <input id="sankBin" type="number" step="1" value="5"/></p>
        <p>Choose annotation(s):</p><p id="SANKgrp"></p>
        <p>Change the order by dragging items up and down:
        <button id="SANKorderBTs" onclick="SANKorderShow()" style="padding:2px;border-radius:45%"><b>&#43;</b></button>
        <button id="SANKorderBTh" onclick="SANKorderHide()" style="padding:2px;border-radius:40%"><b>&#8722;</b></button>
        </p>
        <div id='SANKorderGrp'>
          <ul id="SANKdynList" class="sortable-list"></ul><!--   -->
        </div>
        <p>Please specify the diagram <u>width per category</u>: <input id="SANKw" type="number" step="10" value="250"/> and the <u>diagram height</u>: <input id="SANKh" type="number" step="10" value="700"/></p>
      </div>
      <button class="SANKbt"  onclick="SANKplot()"><b>Plot</b> </button>
      <button onclick="hideShow($('#SANKinput'),$(this));">Hide Input</button>
      <button class="SANKbt" style="float:right;" onclick="for (let item of document.getElementsByClassName('node-label-guide')) {item.setAttribute('style', 'display:none');}; var a = document.createElement('a'); a.href = 'data:application/octet-stream;base64,' + btoa(document.getElementsByClassName('main-svg')[0].outerHTML); a.download = 'Sankey.svg'; a.click();">Save as SVG</button>

      <pre id="SANKnote" style="color:red;"></pre>
      <div id="SANKspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="SANKresize" style="height: 'auto'; width: 'auto'; padding: 0.5em;"></div>

    </div><!--SANK-->

    <div id="DEG" class="tabcontent">
      <div class="topTab">
        <button id="DEGtopBT1" class="DEGtoplinks" onclick="selTop(event, 'DEGtop1','DEGtop')">DEG</button>
        <button id="DEGtopBT2" class="DEGtoplinks" onclick="selTop(event, 'DEGtop2','DEGtop')">GSEA</button>
      </div>

      <div id="DEGtop1" class="DEGtopcontent">
        <div id="DEGinput">
          <h3>Find differentially expressed genes <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-deg-differential-expressed-genes" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
          <p>Selection 1 (<span class="cellN1" style="color:red;">0</span>) cells and selection 2 (<span class="cellN2" style="color:red;">0</span>) cells</p>
          <p>Differentially expressed genes will be detected between the two cell gorups defined above, unless two categories of the 'Custom_combine' annotation are selected,
          which enables DE analysis between those two:</p>
          <p id="DEGgrp"></p>
          <p>Select the number of top DEG:
          <select id="DEGtop">
            <option value=200>200</option>
            <option value=500>500</option>
            <option value=1000>1000</option>
            <option value=5000>5000</option>
            <option value='All'>All</option>
          </select></p>
          <p>Select the method:
          <select id="DEGmethod">
            <option value='default'>Welch's t-test (cellxgene)</option>
            <option value='t-test'>Welchs t-test (diffxpy)</option>
            <option value='rank'>Wilcoxon rank sum (diffxpy)</option>
            <option value='wald'>Wald test (diffxpy)</option>
          </select></p>
          <p>Select genes to be marked in the volcano plot:
          <input type="checkbox" onclick="$('.DEGgenes').prop('checked', this.checked);">Uncheck / Check All</p>
          <p id="DEGgene"></p>
          <p>Define the significance in FDR &#62; <input id="DEGsigFDR" type="number" step="0.001" value="0.05" min="0" max="1"/>
            and absolute log2(FC) &#62; <input id="DEGsigLogfc" type="number" step="0.1" value="1" min="0"/> </p>
          <p>Remove genes in volcano plot with the absolute log2(FC) larger than: <input id="DEGlogfcCut" type="number" step="1" value="10" min="0"/></p>
          <p>Label Size: <input id="DEGlabelSize" type="number" step="0.5" value="3" min="1" max="24"/>
          Dot Size: <input id="DEGdotSize" type="number" step="0.1" value="0.5" min="0.1" max="2"/>
          <p>Y-axis Min: <input id="DEGymin" type="number" step="1" value="-2" min="-10" max="10"/>
          Y-axis Max: <input id="DEGymax" type="number" step="1" value="30" min="1" max="5000"/></p>
          <fieldset>
            <legend>Gene Set Enrichment Analysis:
              <label class="switch"><input type="checkbox" id="GSEAenable" onclick="if($(this).prop('checked')){$('#GSEAopt').show()}else{$('#GSEAopt').hide()}"/>
                <span class="slider"></span></label></legend>
            <div id="GSEAopt" style="display:none;">
              Please select the gene sets: <select id="GSEAgs" style="width:150px;"></select>
              <a href="http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp" target="_blank">Gene set definitions and more.</a>
              <br>The gene sets with minimal size: <input id="GSEAminSize" type="number" step="1" value="15" min="1" />
                and maximal size: <input id="GSEAmaxSize" type="number" step="1" value="500" min="1"/>
              <br>The significant padj cutoff in the plot: <input id="GSEApadjCutoff" type="number" step="0.001" value="0.05" min="0" max="1"/>
              <br>Plot the significant top (ranked by pval) up: <select id="GSEAupN">
                <option value='0'>0</option><option value='5'>5</option><option value='10' selected>10</option><option value='50'>50</option></select>
              and down: <select id="GSEAdnN">
                <option value='0'>0</option><option value='5'>5</option><option value='10' selected>10</option><option value='50'>50</option></select>
              <br>Collapse similar gene sets in the plot: <label class="switch"><input type="checkbox" id="GSEAcollapse" checked><span class="slider"></span></label>
            </div>
          </fieldset>

        </div>

        <button class="DEGbt" onclick="DEGfind()"><b>Find</b></button>
        <button onclick="hideShow($('#DEGinput'),$(this));">Hide Input</button>
        <button id="DEGinteractiveBtn" onclick="hideShow($('#DEGfinal'),$(this),'Interactive');">Hide Interactive</button>
        <pre id="DEGnote" style="color:red;">Note: Unscaled data is used. Scaling option in Global Setting does not apply to DEG.</pre>
        <label id='DEGinfo'></label>
        <div id="DEGspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>

        <div id="DEGfinal">
          <select id="DEGplotBy">
            <option value='color'>Color by</option>
            <option value='label'>Label by</option>
          </select>
          <input type="radio" name="DEGcolorBy" value="significance" onclick="$('#DEGplotBySig').show();$('#DEGplotByGinfo').hide();$('#DEGplotByGsel').hide();" checked>Significance
          <input type="radio" name="DEGcolorBy" value="geneInfo" onclick="$('#DEGplotBySig').hide();$('#DEGplotByGinfo').show();$('#DEGplotByGsel').hide();">Gene Information
          <input type="radio" name="DEGcolorBy" value="selGene" onclick="$('#DEGplotBySig').hide();$('#DEGplotByGinfo').hide();$('#DEGplotByGsel').show();">Gene Selection
          <div id="DEGplotBySig">
            FDR: <input id="DEGfdr" type="number" step="0.001" value="0.05" min="0" max="1"/>
            Asolute log2(FC): <input id="DEGlogfc" type="number" step="0.1" value="1" min="0"/>
          </div>
          <div id="DEGplotByGinfo">
            <select id="DEGplotByGinfoSel"></select>
            <input type="radio" name="DEGplotByGinfoRd" value=">">Larger than
            <input type="radio" name="DEGplotByGinfoRd" value="=" checked>Equal to
            <input type="radio" name="DEGplotByGinfoRd" value="<">smaller than
            <input id="DEGplotByGinfoVal" type="text"/>
          </div>
          <div id="DEGplotByGsel">
            <input type="checkbox" onclick="$('.DEGselgenes').prop('checked', this.checked);">Uncheck / Check All
            <p id="DEGselgene"></p>
          </div>
          <div>
            <button id="DEGfigReset">Reset Axes</button>
            <button onclick="DEGplot('DEG')">Replot</button>
            <p id="DEGinteractive" style="height: 500px; width:500px; padding: 0.5em; border: 1px solid #B6B6B6"></p>
          </div>
        </div>

        <table id="DEGtable" class="display"></table>
        <button onclick="ZoomIn('DEGresize')">Zoom In</button>
        <button onclick="ZoomOut('DEGresize')">Zoom Out</button>
        <button class="DEGpngBT" onclick="saveImg('DEGimg')" disabled>Save</button>
        <div id="DEGresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
          <p id="DEGfig"></p>
        </div>
      </div>
      <div id="DEGtop2" class="DEGtopcontent">
        <h3>Gene Set Enrichment Analysis <span style="font-size:10px">by <a href="https://bioconductor.org/packages/release/bioc/html/fgsea.html" target="_blank">fgsea</a></span></h3>
        <table id="GSEAtable" class="display"></table>
        <button onclick="ZoomIn('GSEAresize')">Zoom In</button>
        <button onclick="ZoomOut('GSEAresize')">Zoom Out</button>
        <div id="GSEAresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
          <p id="GSEAfig"></p>
        </div>
        <table id="GSEAtable" class="display"></table>
      </div>

    </div><!--DEG-->

    <div id="preDEGpanel" class="tabcontent">
      <div class="topTab">
        <button id="DEGlinksPlot" class="DEGlinks" onclick="selTop(event, 'preDEGvolcano','DEG')">Plot DEG</button>
        <button id="DEGlinksMulti" class="DEGlinks" onclick="selTop(event, 'preDEGmulti','DEG')">Multi DEG</button>
        <button class="DEGlinks" onclick="selTop(event, 'preDEGfgsea','DEG')">GSEA</button>
      </div>

      <div id="preDEGvolcano" class="DEGcontent">
        <h3>Plot volcano for the pre-computed DEGs</h3>
        <select id="preDEGvolcanoTab"></select>
        <p>Select genes to be marked in the volcano plot:
        <input type="checkbox" onclick="$('.preDEGvolcanogenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="preDEGvolcanogene"></p>
        <p>Select the number of top DEG:
          <select id="preDEGvolcanotop">
            <option value=200>200</option>
            <option value=500>500</option>
            <option value=1000>1000</option>
            <option value=5000>5000</option>
            <option value='All'>All</option>
          </select></p>
        <p>Define the significance in FDR &#60; <input id="preDEGvolcanoFDR" type="number" step="0.001" value="0.05" min="0" max="1"/>
            and absolute log2(FC) &#62; <input id="preDEGvolcanoLogfc" type="number" step="0.1" value="1" min="0"/> </p>
        <p>Label Size: <input id="preDEGvolcanolabelSize" type="number" step="0.5" value="3" min="1" max="24"/>
        Dot Size: <input id="preDEGvolcanodotSize" type="number" step="0.1" value="0.5" min="0.1" max="2"/>
        <p>Y-axis Min: <input id="preDEGvolcanoymin" type="number" step="1" value="-2" min="-10" max="10"/>
        Y-axis Max: <input id="preDEGvolcanoymax" type="number" step="1" value="30" min="1" max="5000"/></p>

        <p hidden>Remove genes in volcano plot with the absolute log2(FC) larger than: <input id="preDEGvolcanologfcCut" type="number" step="1" value="100" min="0"/></p>
        <button class="preDEGvolcanobt" onclick="preDEGVolcanofind('preDEGvolcano')"><b>Plot</b></button>
        <button id="preDEGvolcanointeractiveBtn" onclick="hideShow($('#preDEGvolcanofinal'),$(this),'Interactive');">Hide Interactive</button>
        <pre id="preDEGvolcanonote" style="color:red;"></pre>
        <label id='preDEGvolcanoinfo'></label>
        <div id="preDEGvolcanospin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>

        <div id="preDEGvolcanofinal">
          <select id="preDEGvolcanoplotBy">
            <option value='color'>Color by</option>
            <option value='label'>Label by</option>
          </select>
          <input type="radio" name="preDEGvolcanocolorBy" value="significance" onclick="$('#preDEGvolcanoplotBySig').show();$('#preDEGvolcanoplotByGinfo').hide();$('#preDEGvolcanoplotByGsel').hide();" checked>Significance
          <input type="radio" name="preDEGvolcanocolorBy" value="geneInfo" onclick="$('#preDEGvolcanoplotBySig').hide();$('#preDEGvolcanoplotByGinfo').show();$('#preDEGvolcanoplotByGsel').hide();">Gene Information
          <input type="radio" name="preDEGvolcanocolorBy" value="selGene" onclick="$('#preDEGvolcanoplotBySig').hide();$('#preDEGvolcanoplotByGinfo').hide();$('#preDEGvolcanoplotByGsel').show();">Gene Selection
          <div id="preDEGvolcanoplotBySig">
            FDR: <input id="preDEGvolcanofdr" type="number" step="0.001" value="0.05" min="0" max="1"/>
            Asolute log2(FC): <input id="preDEGvolcanologfc" type="number" step="0.1" value="1" min="0"/>
          </div>
          <div id="preDEGvolcanoplotByGinfo">
            <select id="preDEGvolcanoplotByGinfoSel"></select>
            <input type="radio" name="preDEGvolcanoplotByGinfoRd" value=">">Larger than
            <input type="radio" name="preDEGvolcanoplotByGinfoRd" value="=" checked>Equal to
            <input type="radio" name="preDEGvolcanoplotByGinfoRd" value="<">smaller than
            <input id="preDEGvolcanoplotByGinfoVal" type="text"/>
          </div>
          <div id="preDEGvolcanoplotByGsel">
            <input type="checkbox" onclick="$('.preDEGvolcanoselgenes').prop('checked', this.checked);">Uncheck / Check All
            <p id="preDEGvolcanoselgene"></p>
          </div>
          <div>
            <button id="preDEGvolcanofigReset">Reset Axes</button>
            <button onclick="DEGplot('preDEGvolcano')">Replot</button>
            <p id="preDEGvolcanointeractive" style="height: 500px; width:500px; padding: 0.5em; border: 1px solid #B6B6B6"></p>
          </div>
        </div>

        <table id="preDEGvolcanotable" class="display"></table>
        <button class="preDEGvolcanobt" onclick="ZoomIn('preDEGvolcanoresize')">Zoom In</button>
        <button class="preDEGvolcanobt" onclick="ZoomOut('preDEGvolcanoresize')">Zoom Out</button>
        <button class="preDEGvolcanobt" onclick="saveImg('preDEGvolcanoimg')" disabled>Save</button>
        <div id="preDEGvolcanoresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
          <p id="preDEGvolcanofig"></p>
        </div>

      </div><!--preDEGvolcano-->

      <div id="preDEGmulti" class="DEGcontent">
        <h3>Plot the bubble heatmap for pre-computed DEGs from multiple comparisons</h3>
        <button class="preDEGmultibt" onclick="hideShow($('#preDEGmultiinput'),$(this));">Hide Input</button>
        <div id="preDEGmultiinput">
          <p>Select genes to be included in the bubble heatmap (rows):
          <input type="checkbox" onclick="$('.preDEGmultigenes').prop('checked', this.checked);">Uncheck / Check All</p>
          <p id="preDEGmultigene"></p>
          <p>Select comparisons from the table (shift/Ctrl for multi-selections) to be included in the bubble heatmap (columns):
          <!--
          <input type="checkbox" onclick="$('.preDEGmultigrps').prop('checked', this.checked);">Uncheck / Check All</p>
          <p id="preDEGmulticomp"></p>
          -->

          <table id="preDEGmulticompTable" class="display" width="100%"></table>
          <p>Change the order by dragging items up and down:
          <b><button id="preDEGmultiorderBT" onclick="hideShow($('#preDEGmultiorderGrp'),$(this),'');" style="padding:2px;border-radius:50%">&#8722;</button></b></p>
          <div id='preDEGmultiorderGrp'>
            <ul id="preDEGmultidynList" class="sortable-list"></ul><!--   -->
          </div>
          <p>Bubble distance in plot: <input id="preDEGmultiFigureScale" type="number" step="0.1" value="0.7"/></p>
        </div>
        <button class="preDEGmultibt"  onclick="preDEGmultiplot('preDEGmulti')"><b>Plot</b> </button>
        <div id="preDEGmultispin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
        <pre id="preDEGmultinote" style="color:red;"></pre>
        <button onclick="ZoomIn('preDEGmultiresize')">Zoom In</button>
        <button onclick="ZoomOut('preDEGmultiresize')">Zoom Out</button>
        <button class="preDEGmultibt" onclick="saveImg('preDEGmultiimg')" disabled>Save</button>
        <div id="preDEGmultiresize" style="height: 'auto'; width: 'auto'; padding: 0.5em;">
          <p id="preDEGmultifig"></p>
        </div>
      </div><!--preDEGmulti-->
      
      <div id="preDEGfgsea" class="DEGcontent"> <!--preDEG fgsea-->
        <h3>Gene Set Enrichment Analysis <span style="font-size:10px">by <a href="https://bioconductor.org/packages/release/bioc/html/fgsea.html" target="_blank">fgsea</a></span></h3>
        <div id='preDEGfgseainput'>
          <select id="preDEGfgseaTab"></select>
          <br>Please select the gene sets: <select id="preDEGfgseaGS" style="width:150px;"></select>
          <a href="http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp" target="_blank">Gene set definitions and more.</a>
          <br>The gene sets with minimal size: <input id="preDEGfgseaMinSize" type="number" step="1" value="15" min="1" />
          and maximal size: <input id="preDEGfgseaMaxSize" type="number" step="1" value="500" min="1"/>
          <br>The significant padj cutoff in the plot: <input id="preDEGfgseaPadjCutoff" type="number" step="0.001" value="0.05" min="0" max="1"/>
          <br>Plot the significant top (ranked by pval) up: <select id="preDEGfgseaUpN">
          <option value='0'>0</option><option value='5'>5</option><option value='10' selected>10</option><option value='50'>50</option></select>
          and down: <select id="preDEGfgseaDnN">
          <option value='0'>0</option><option value='5'>5</option><option value='10' selected>10</option><option value='50'>50</option></select>
          <br>Collapse similar gene sets in the plot: <label class="switch"><input type="checkbox" id="preDEGfgseaCollapse" checked><span class="slider"></span></label>
        </div>
        <button class="preDEGfgseabt" onclick="preDEGfgseaFind()"><b>Find</b> </button>
        <button class="preDEGfgseabt" onclick="hideShow($('#preDEGfgseainput'),$(this));">Hide Input</button>
        <div id="preDEGfgseaspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
        <pre id="preDEGfgseanote" style="color:red;"></pre>
        <table id="preDEGfgseaTable" class="display"></table>
        <button onclick="ZoomIn('preDEGfgsearesize')">Zoom In</button>
        <button onclick="ZoomOut('preDEGfgseaAresize')">Zoom Out</button>
        <div id="preDEGfgsearesize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
          <p id="preDEGfgseafig"></p>
        </div>
      </div>

    </div><!--preDEGpanel -->

    <div id='STACBAR' class='tabcontent'>
      <div id="STACBARinput">
        <h3>Interactive stacked barplot of cell distribution of selected cells over two annotations and/or genes <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-stacked-barplot" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
        <p>Choose cell selection(s):
          <label><input class="STACBARcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="STACBARcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose two from genes and annotations ('Add Genes' first):</p>
        <p id="STACBARgene"></p>
        <div id='STACBARgeneSel'>
          <p>Number of equally divided gene expression ranges: <input id="STACBARbin" type="number" step="1" value="5"/></p>
        </div>
        <p>Annotations:</p><p id="STACBARgrp"></p>
        <p>Color by <select id="STACBARcol"></select></p>
		    <p>Figure width: <input id="STACBARwidth" type="number" step="50" value="800"/> , height: <input id="STACBARheight" type="number" step="40" value="480"/></p>
        <p>x-Axis font size: <input id="STACBARxfontsize" type="number" step="2" value="16"/> ,
        label rotate: <input id="STACBARxlabelrotate" type="number" step="5" value="-60"/> ,
		     label shift: <input id="STACBARxlabelshift" type="number" step="2" value="-24"/></p>
      </div>


      <button class="STACBARbt"  onclick="STACBARplot()"><b>Plot</b> </button>
      <button onclick="hideShow($('#STACBARinput'),$(this));">Hide Input</button>
      <button class="STACBARbt" style="float:right;" onclick="var a = document.createElement('a'); a.href = 'data:application/octet-stream;base64,' + btoa(document.getElementById('STACBARsvg').outerHTML); a.download = 'Stackedbar.svg'; a.click();">Save as SVG</button>

      <pre id="STACBARnote" style="color:red;"></pre>
      <div id="STACBARspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id='STACBARresize' style="height: 'auto'; width: 'auto'; padding: 0.5em;"></div>
    </div><!--STACBAR -->

    <div id='BROW' class='tabcontent'>
      <h3>Jointly visualizing gene expression and DNA accessibility</h3>
      <div id="BROWinput">        
        <fieldset>
          <button onclick="hideShow($('#BROWtrackView'),$(this),'');">Hide</button>
          <div id="BROWtrackView">
            <input id="BROWtrackAll" type="checkbox" checked onclick="BROWreset();">
            <strong>The following tracks are available:</strong>
            <table id="BROWtable" class="display"></table>
            <div id="BROWinfo"></div>
          </div>

        </fieldset>
        Selected browser tracks:
        <small>Change the track orders by dragging items up or down, can also be removed by clicking <i class='fa fa-trash-o'></i></small>
        <ul id="BROWdynList" class="sortable-list"></ul>

        Define track region by:
        <label><input type="radio" name="BROWinputRadio" value="coordinate" onclick="$('.BROWregion').hide();$('#BROWregionCoor').show();$('#BROWselGene').show();">Coordinate</label>
        <label><input type="radio" name="BROWinputRadio" value="gene" onclick="$('.BROWregion').hide();$('#BROWregionGene').show();$('#BROWselGene').show();">Gene</label>
        <div id="BROWregionCoor" class="BROWregion" style="display:none">
          Chromosome: <input id="BROWregionChr" type="text" size="5"/>
          Start: <input id="BROWregionStart" type="number" style="width: 10em"/>
          End: <input id="BROWregionEnd" type="number" style="width: 10em"/>
        </div>
        <div id="BROWregionGene" class="BROWregion" style="display:none">
          Gene: <input id="BROWregionGeneName" type="text" size="10"/> <!-- <p id="BROWgene"></p> -->
          Extension upstream (bp): <input id="BROWregionUpEx" type="number" value="1000" min="0"/>
          Extension downstream (bp): <input id="BROWregionDnEx" type="number" value="1000" min="0"/>
        </div>
        <div id="BROWselGeneShow" style="display:none"><div id="BROWselGene" style="display:none">
          <br>
          <p>Add following gene expression violins to tracks ('Add Genes' first): <input type="checkbox" checked onclick="$('.BROWgenes').prop('checked', this.checked);">Uncheck / Check All</p>
          <p id="BROWgene"></p>
          <p>Expression cutoff: <input id="BROWgeneCutoff" type="number" step="0.1" value="-0.1" min="-0.1"/></p>
        </div></div>
      </div>
      <br>
      <button class="BROWbt"  onclick="BROWplot()"><b>Plot</b> </button>
      <button onclick="ZoomIn('BROWresize')">Zoom In</button>
      <button onclick="ZoomOut('BROWresize')">Zoom Out</button>
      <button class="BROWpngBT" onclick="saveImg('BROWimg')" disabled>Save</button>
      <button onclick="hideShow($('#BROWinput'),$(this));">Hide Input</button>

      <pre id="BROWnote" style="color:red;"></pre>
      <div id="BROWspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="BROWresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
          <p id="BROWfig"></p>
      </div>

    </div> <!-- Multiome -->
    
    <div id='COSMX' class='tabcontent'>
      <h3>Jointly visualizing histology image, cell boundaries, transcript locations</h3>
      <div id='COSMXinput'>
          Choose cosMx captures to plot
          <p id="COSMXsample"></p>
          <p>Select coordinates: 
            <input type="radio" name="COSMXcood" value="local" onclick="IMGcellSel()" checked>Local
            <input type="radio" name="COSMXcood" value="global" onclick="IMGcellSel()">Global
          </p>
          <p><label><input id="COSMXhistology" type="checkbox" checked/>Draw histology image</label></p>
          <p><label><input id="COSMXcell" type="checkbox" checked/>Highlight cell boundaries with color <input class="colorInput" id="COXMXcellColor" type='color' value='#ffffff'></label></p>
          Choose transcripts to highlight (need to 'Add Genes' first)
          <p id="COSMXgene"></p>
          Choose color for selected transcripts
          <p id="COSMXgeneColor"></p>
          <p>Choose the size for transcript markers: <input type='number' id="COSMXgeneSize" step="1" value="3"></p>
          <button class="COSMXbt"  onclick="COSMXplot()"><b>Plot</b> </button>
      </div>
      <button onclick="hideShow($('#COSMXinput'),$(this));">Hide Input</button>
      <br>
      <pre id="COSMXnote" style="color:red;"></pre>
      <div id="COSMXspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="COSMXresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
        <div id="COSMXfigs"></div>
      </div>
    </div> <!-- COSMX -->
    
    <div id='VISIUM' class='tabcontent'>
      <h3>Visualizing 10X visium data along with histology images</h3>
      <div id='VISIUMinput'>
          Choose visium samples to be included: <input type="checkbox" checked onclick="$('.VISIUMsamples').prop('checked', this.checked);">Uncheck / Check All
          <p id="VISIUMsample"></p>
          <p>Choose categorical annotation(s):</p>
          <p id="VISIUMgrp"></p>
          <p>Choose continuous annotation(s):</p>
          <p id="VISIUMmeta"></p>
          <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.VISIUMgenes').prop('checked', this.checked);">Uncheck / Check All</p>
          <p id="VISIUMgene"></p>
          <p>Group the plot by slides: <input id="VISIUMsortID" type="checkbox" checked/></p>
          <p>The number of plots per row: <input id="VISIUMslot" type="number" value="4" min="2"/></p>
          <p>The transparent alpha: <input id="VISIUMalpha" type="number" value="0.7" min="0" max="1"/></p>
          <p>The dotsize of the spot: <input id="VISIUMdotsize" type="number" value="2" min="1" max="6"/></p>
          <p>The size of subplots: <input id="VISIUMsubsize" type="number" value="4" min="3" max="8"/></p>
      </div>
      <button class="VISIUMbt" onclick="VISIUMplot()"><b>Plot</b> </button>
      <button onclick="ZoomIn('VISIUMresize')">Zoom In</button>
      <button onclick="ZoomOut('VISIUMresize')">Zoom Out</button>
      <button class="VISIUMpngBT" onclick="saveImg('VISIUMimg')" disabled>Save</button>
      <button onclick="hideShow($('#VISIUMinput'),$(this));">Hide Input</button>
      <br>
      <pre id="VISIUMnote" style="color:red;"></pre>
      <div id="VISIUMspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="VISIUMresize" style="height: 'auto'; padding: 0.5em; border: 1px solid #B6B6B6">
        <div id="VISIUMfig"></div>
      </div>
    </div> <!-- VISIUM -->

    <div id="IMG" class="tabcontent">
      <h3>Global setting for VIP <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-figure-option" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
      <p>Select cells cumulatively:
        <input type="radio" name="IMGcumuRd" value="Yes" onclick="IMGcellSel()">Yes
        <input type="radio" name="IMGcumuRd" value="No" onclick="IMGcellSel()" checked>No
        &nbsp;&nbsp;&nbsp;&nbsp;<button onclick="IMGcellClear()">Clear Cell Selections</button>&nbsp;<button onclick="selectAllCells()">Select All Cells</button></p>
        <div id='IMGrandomCellDiv'>
        <Blockquote>
          <label>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>&nbsp;&nbsp;|&nbsp;&nbsp;
          <label>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </Blockquote>
        <!--
        From selected cell group, <input id="IMGcellRandomN" type="number" step="1" min="1" style="width: 50px;" /> cells will be randomly <button onclick="IMGcellRandomChoose()">Chosen</button>
        -->
        </div>
      <div>
        <p>Scale gene expression to unit variance and zero mean
        <input type="radio" name="IMGscale" value="Yes" onclick="IMGset()">Yes
        <input type="radio" name="IMGscale" value="No" onclick="IMGset()">No
        <div id="IMGscaleShow">
          <Blockquote>
            <input id="IMGscaleZero" type="checkbox" onclick="IMGset()"/> Genes with no variation are retained and set expression values to 0 during this operation<br>
            <input id="IMGclipValue" type="checkbox" onclick="IMGset()"/> Clip to a maximum value of <input id="IMGscaleMax" type="number" step="0.1" onchange="IMGset()"/>
          </Blockquote>
        </div>
      </div>
      <p>Width of the central canvas of main window in pixel: <input id=canvasWidth name=canvasWidth value=1024 size=4>&nbsp;&nbsp;<button onclick="$('#cellxgene_container div')[0].children[1].style.width=$('#canvasWidth').val()+'px';window.dispatchEvent(new Event('resize'));">Apply</button>
      <p>Figure DPI (Dots Per Inch, for PNG only):
      <select id="IMGdpi" onchange="IMGset()">
        <option value=150>150</option>
        <option value=300>300</option>
        <option value=600>600</option>
      </select></p>
      <p>Figure format:
      <select id="IMGimg" onchange="IMGset()">
        <option value='png'>PNG</option>
        <option value='svg'>SVG</option>
      </select></p>
      <p>Font size:
      <input id="IMGfontsize" type="number" step="1" value="11" min="5" onchange="IMGset()"/></p>
      <p>Vector friendly:
        <input type="radio" name="IMGvector" value="Yes" onclick="IMGset()" checked>Yes, plot mass components using png
        <input type="radio" name="IMGvector" value="No" onclick="IMGset()">No, all are vectors.
      </p>
      <p>Transparent figure:
        <input type="radio" name="IMGtransparent" value="Yes" onclick="IMGset()">Yes
        <input type="radio" name="IMGtransparent" value="No" onclick="IMGset()" checked>No
      </p>
      <input type="radio" name="IMGscanpybranch" value="master" style="visibility: hidden;" checked>
      <!--
      <p>Scanpy plot style:
        <input type="radio" name="IMGscanpybranch" value="master" onclick="IMGset()">master
        <input type="radio" name="IMGscanpybranch" value="split_show" onclick="IMGset()" checked>split_show
      </p>
      -->
      <p>Colormap of Embedding Plot (all available), Density Scatter (limited to top 4 colormaps):
      <select id="IMGcolor" onchange="IMGset()">
        <option value='viridis'>viridis</option>
        <option value='plasma'>plasma</option>
        <option value='inferno'>inferno</option>
        <option value='magma'>magma</option>
        <option value='spring'>spring</option>
        <option value='summer'>summer</option>
        <option value='autumn'>Fall</option>
        <option value='winter'>winter</option>
      </select>
      <input id="IMGcolor_r" type="checkbox" onchange="IMGset()">Reverse
      </p>
      <table border=0>
        <tr><td>viridis</td><td rowspan=8><img src="https://interactivereport.github.io/cellxgene_VIP/static/color_map.png" height="180px"></tr>
        <tr><td>plasma</td></tr>
        <tr><td>inferno</td></tr>
        <tr><td>magma</td></tr>
        <tr><td>spring</td></tr>
        <tr><td>summer</td></tr>
        <tr><td>Fall</td></tr>
        <tr><td>winter</td></tr>
      </table>
    </div><!--IMG-->

    <div id="SPATIAL" class="tabcontent">
      <div class="topTab">
        <button id="SPATIALimage" class="SPATIALlinks" onclick="selTop(event, 'SPATIALimageV','SPATIAL')">Image</button>
        <button id="SPATIALlayout" class="SPATIALlinks" onclick="selTop(event, 'SPATIALlayoutV','SPATIAL'); SPATIALlist()">Layout</button>
      </div>
      <div id="SPATIALspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="SPATIALimageV" class="SPATIALcontent">
        <!--
        <p>Choose cell selection(s):
          <label><input class="SPATIALcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>
          <label><input class="SPATIALcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        -->
        <button class="SPATIALbt" onclick="SPATIAL()"><b>Get Spatial Image</b></button>
        <button onclick="SPATIALoverlay()"><b>Overlay Image</b></button>
        <button onclick="document.getElementById('spatial-layer').remove();"><b>Remove Image</b></button>
        <pre id="SPATIALnote" style="color:red;"></pre>
        <p>Image resolution:
          <input type="radio" name="IMGresolution" value="hires" checked>High
          <input type="radio" name="IMGresolution" value="lowres">Low </p>
        <p>Move embedding and image together:
          <input type="radio" name="IMGtogether" value="Yes" checked>Yes
          <input type="radio" name="IMGtogether" value="No">No </p>
        <p>Image on the front:
          <input type="radio" name="IMGfront" value="Yes" onclick="SpatialIMGfront();" checked>Yes
          <input type="radio" name="IMGfront" value="No" onclick="SpatialIMGfront();">No </p>

        <p><input type="range" min="0" max="1" value="0.75" step="0.05" id="image_opacity" oninput="SpatialIMGopacity(this.value);"> Image Opacity: <span id="image_opacity_val">0.75</span></p>
        <p><input type="range" min="0" max="1" value="1" step="0.05" id="embedding_opacity" oninput="SpatialEMBopacity(this.value);"> Embedding Opacity: <span id="embedding_opacity_val">1</span></p>
      </div>
      <div id="SPATIALlayoutV" class="SPATIALcontent">
        <button class="SPATIALbt" onclick="SPATIALarrange()"><b>Arrange Slides</b></button>
        <button onclick="saveLayout()"><b>Save Layout</b></button>
        <button onclick="hideShow($('#SLIDESinput'),$(this),'Slides');">Hide Slides</button>
        <div id="SLIDESinput">
          <p>Choose slides to arrange: <input type="checkbox" checked onclick="$('.SPATIALnames').prop('checked', this.checked);">Uncheck / Check All</p>
          <p id="SPATIALname"></p>
          <hr>
        </div>
        <p>Click on an image to operate: <span id="selected_image"></span></p>
        <p><input type="checkbox" id="flipx" onclick="SpatialIMGman();"> Flip X
           <input type="checkbox" id="flipy" onclick="SpatialIMGman();"> Flip Y
           <input type="range" style="width:50;" min="0" max="270" value="0" step="90" id="image_rotate" oninput="SpatialIMGman();"> Rotate: <span id="image_rotate_val">0</span> degree</p>
        <div class="grid-stack"></div>
      </div>
    </div><!--SPATIAL-->

    <div id="BRUSH" class="tabcontent">
      <h3>Brushed ranges <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-other-functions" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
      <div id="BRUSHset"></div>
    </div><!--BRUSH-->

    <div id="SYNC" class="tabcontent">
      <pre id="SYNCnote"></pre>
    </div><!--SYNC-->

    <div id="CLI" class="tabcontent">
      <h3>CLI to operate on data directly <a href="https://interactivereport.github.io/cellxgene_VIP/tutorial/docs/how-to-use-cellxgene-vip.html#vip-command-line-interface" target="_blank"><i class="fa fa-question-circle" style="font-size:20px"></i></a></h3>
      <div id="CLIinput">
        <p>Choose cell selection(s):
          <label><input class="CLIcell" type="checkbox" value="celllist1" checked/>Selection 1 (<span class="cellN1" style="color:red">0</span>)</label>


          <label><input class="CLIcell" type="checkbox" value="celllist2" checked/>Selection 2 (<span class="cellN2" style="color:red">0</span>)</label>
        </p>
        <p>Choose genes ('Add Genes' first): <input type="checkbox" checked onclick="$('.CLIgenes').prop('checked', this.checked);">Uncheck / Check All</p>
        <p id="CLIgene"></p>
        <p>Choose annotation(s):</p><p id="CLIgrp"></p>
        <p>Choose layout:</p><p id="CLIlayout"></p>
        <div id="CLIscript" style="height:380px;" ></div>
      </div>
      <div id="CLIhelp" style="display:none;">
      <ul>
        <li>adata: a python AnnData object created for you to use in CLI based on your selected cells and metadata.</li>
        <li>strPath: the full file name without suffix. strPath+'.h5ad' is the data file of adata.</li>
        <li>Python mini-Jupyter notebook (Vignette 1 & 2)
          <ul>
          <li>Use lines with leading '#' to seperate code for readability as empty lines will create new Jupyter notebook cells.</li>
          <li>R code cells should start with "%%R" as shown in Vignette 2. Add "%Rdevice svg" if SVG format is desired and numbers of elements in SVG plots are managable.</li>
          </ul>
        </li>
        <li>R Markdown (Vignette 3)
          <ul>
          <li>All objects created within Python chunks are available to R using the py object exported by the reticulate package.</li>
          <li>You can analagously access R objects within Python chunks via the r object. </li>
          </ul>
        </li>
      </ul>
      </div>
      <button class="CLIbt"  onclick="CLIplot()"><b>Run</b></button>
      <button class="CLIbt"  onclick="editor.getSession().setValue('')">Clear</button>
      <button class="CLIbt"  onclick="editor.getSession().setValue($('#CLIVignetteCode1').text().trim());selectAllCells();">Vignette 1</button>
      <button class="CLIbt"  onclick="editor.getSession().setValue($('#CLIVignetteCode2').text().trim());selectByCelltype('T cells');">Vignette 2</button>
      <button class="CLIbt"  onclick="editor.getSession().setValue($('#CLIVignetteCode3').text().trim());selectByCelltype('T cells');">Vignette 3</button>
      <button onclick="hideShow($('#CLIhelp'),$(this),'Help');">Show Help</button>
      <button onclick="hideShow($('#CLIinput'),$(this));">Hide Input</button>
      <button id="printNotebook" onclick="printNotebook();">Print Notebook</button>
      <button id="hideCode" onclick="hideShow($('.input'),$(this),'Code');">Hide Code</button>
      <button id="hideOutput" onclick="hideShow($('.output_text'),$(this),'Output');">Hide Output</button>
      <div id="CLIsvg"></div>
      <pre id="CLInote" style="color:red;"></pre>
      <div id="CLIspin" class="loading style-2" style="visibility: hidden;"><div class="loading-wheel"></div></div>
      <div id="CLIresize" style="height: 'auto'; width: 'auto'; padding: 0.5em;"></div>
    </div><!--CLI-->

    <div id="tVIP" class="tabcontent">
      <h3>Test VIP functionalities</h3>
      <p><button id="tVIPtest" class="tVIPbt" onclick="testVIP('tVIPnote')"><b>Test by the latest case</b></button></p>
      <fieldset>
        <legend>Create a test case</legend>
        <p>by: <select id="tVIPgrp"></select></p>
        <p><button class="tVIPbt tVIPbtA" onclick="tVIPcreate('random','tVIPnote')">Create</button><br>
          The process will<br>
          1. Split cells with selected annotation into two groups randomly;<br>
          2. Select 500 cells from each group;<br>
          3. Perform DEG (Differentially Expressed Genes) method to those two groups of cells;<br>
          4. Select top DEGs randomly into groups;<br>
          5. Choose plotting options in VIP randomly;<br>
          6. Perform plotting functions one by one;<br>
          7. Save the test case on the server for future use (overwrite previous).
        </p>
        <p><button class="tVIPbt tVIPbtA" onclick="tVIPcreate('cells','tVIPnote')">Create on selected cells</button><br>
          The test generation will ignore the first 2 steps in the above process and run the rest.
        </p>
        <p><button class="tVIPbt tVIPbtA" onclick="tVIPcreate('selections','tVIPnote')">Create on selections</button><br>
          The user has selected all options on VIP including the genes and gene sets (or loaded previous session).
          This process will process all VIP plots one by one and save it as a test case on the server (overwrite previous).
        </p>
      </fieldset>
      <div id="tVIPnote"></div>
    </div>

    <div id="LOAD" class="tabcontent">
      <h3>Load saved session</h3>
      <p>Load from a url?
        <input type="radio" name="LOADurlRd" value="Yes" onclick="$('#LOADurlDIV').show()">Yes
        <input type="radio" name="LOADurlRd" value="No" onclick="$('#LOADurlDIV').hide()" checked>No
      </p>
      <div id='LOADurlDIV' style="display:none;">
        <input type="url" id="LOADurl" size=80/>
        <input id="loadfile" type="file" onchange="load()" style="display:none"/>
      </div>
      <p>Run plotting after loading?
        <input type="radio" name="LOADplotRd" value="Yes">Yes
        <input type="radio" name="LOADplotRd" value="No" checked>No
      </p>
      <br><button onclick="loadClick()">load</button>
      <div id="LOADnote"></div>
    </div>
  </div>
</div>

<script type="text/javascript">
var heatCell = 10000;
var editor;
var VIPurl = '';
var VIPCellSelectionMsg = "Please use selection tools to put cells into selection bins";
var VIPGeneSelectionMsg = "Please choose a gene to plot";
var uniqID = 'VIPtesting';
var loadingCycle = 0;

$(document).ready(function(){
    let gSetFile=document.getElementById('Geneset_file');
    gSetFile.onchange 
})

function printNotebook(){
    var html = $('#CLIresize').html();
    var mywin=window.open('','_blank');
    mywin.document.write('<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.1.10/require.min.js"></script\>\n');
    mywin.document.write('<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/2.0.3/jquery.min.js"></script\>\n');
    mywin.document.write('<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script\>\n');
    mywin.document.write('<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.css" integrity="sha512-pli9aKq758PMdsqjNA+Au4CJ7ZatLCCXinnlSfv023z4xmzl8s+Jbj2qNR7RI8DsxFp5e8OvbYGDACzKntZE9w==" crossorigin="anonymous" />\n');
    mywin.document.write(html);
    mywin.document.close();
}

function checkLoading(){
    if(!window.store || !window.store.getState().categoricalSelection || !window.store.getState().annoMatrix._cache.obs.__columns){
      return false;
    }
    if (loadingCycle > 300) return true;
    //console.log("Loaded Annotation:"+store.getState().annoMatrix._cache.obs.__columns.length);
    var annoNum = store.getState().annoMatrix.schema.annotations.obs.columns.length;
    //var columns = store.getState().annoMatrix.schema.annotations.obs.columns;
    //for (var i=0;i<columns.length;i++) {
    //  if (columns[i].type === "categorical" && columns[i].categories.length<2) continue;
    //  annoNum ++;
    //}

    if (Object.keys(store.getState().annoMatrix.schema.annotations.obsByName).includes(store.getState().annoMatrix.schema.annotations.obs.index)) annoNum--;
    if (Object.keys(store.getState().annoMatrix.schema.annotations.obsByName).includes(">Description")) annoNum--;
	if (store.getState().annoMatrix._cache.obs.__columns.length < annoNum) {
      return false;
    }
    return true;
}
function enableButton(){
    if(!checkLoading()){
      for(const one of document.getElementsByClassName('tablinks')){
        one.disabled=true;
      }
      //$("#preDEGpanelBT").hide();
      //$("#metaCellBT").hide();
      //$("#VIPtestBT").hide();
      setTimeout(enableButton, 100);
      loadingCycle ++;
    }else{
      for(const one of document.getElementsByClassName('tablinks')){
        one.disabled=false;
      }
      //if (window.store.getState().annoMatrix.schema.annotations.obsByName[">Description"] !== undefined) {
      //  if(window.store.getState().annoMatrix.schema.annotations.obsByName[">Description"].categories[0].split("|").length>1){
      //    var ginit = window.store.getState().annoMatrix.schema.annotations.obsByName[">Description"].categories[0].split("|")[1].split(";");
      //    window.store.dispatch({type: "set layout choice", layoutChoice: ginit[0].split("by")[1].trim()});
      //    window.store.dispatch({type: "color by categorical metadata", colorAccessor: ginit[1].split("by")[1].trim()});
      //    window.store.dispatch({type: "show centroid labels for category", showLabels: true});
      //  }
      //}
      selectAllCells();
      init();
    }
    //var d = new Date();
    //console.log(d.getTime());
}

enableButton();

function dragElement( element, direction){
    var   md; // remember mouse down info
    const pLeft  = document.getElementById("panel-left");
    const pRight = document.getElementById("panel-right");

    element.onmousedown = onMouseDown;

    function onMouseDown( e )
    {
    	//console.log("mouse down: " + e.clientX);
    	md = {e,
    	      panelW:pLeft.parentElement.offsetWidth,
    	      lper:parseInt(pLeft.style.width)
    	};
      document.onmousemove = onMouseMove;
      document.onmouseup = () => {
          document.onmousemove = document.onmouseup = null;
      }
      console.log(md);
    }

    function onMouseMove( e )
    {
      var moveR = Math.ceil(100*(e.clientX - md.e.x)/md.panelW);
    	if (direction === "H" ) // Horizontal
    	{
    	    // prevent negative-sized elements
    	    moveR = (moveR < -md.lper) ? 0 : (moveR > (99-md.lper)) ? 0 : moveR;

          element.style.left = md.lper+moveR+"%";
          pLeft.style.width = md.lper+moveR+"%";
          pRight.style.width = (100-(md.lper+moveR)) +"%";
    	}
    }
}
dragElement( document.getElementById("separator"), "H" );

function selFun(evt,sel){
  var i, tabC, tabL;
  tabC = document.getElementsByClassName("tabcontent");
  for (i = 0; i < tabC.length; i++) {
    tabC[i].style.display = "none";
  }
  tabL = document.getElementsByClassName("tablinks");
  for (i = 0; i < tabL.length; i++) {
    tabL[i].className = tabL[i].className.replace(" active", "");
  }
  document.getElementById(sel).style.display = "block";
  evt.currentTarget.className += " active";
  sync();
}
function defaultTopSel(grp,sel){
  var initF = false;
  $("."+grp+"links").each(function(){
    initF ||= $(this).prop("class").includes("active");
  });
  if(!initF){
    $("#"+sel)[0].click();
  }
}
function selTop(evt,sel,grp){
  var i, tabC, tabL;
  tabC = document.getElementsByClassName(grp+"content");
  for (i = 0; i < tabC.length; i++) {
    tabC[i].style.display = "none";
  }
  tabL = document.getElementsByClassName(grp+"links");
  for (i = 0; i < tabL.length; i++) {
    tabL[i].className = tabL[i].className.replace(" active", "");
  }
  //console.log(sel);
  document.getElementById(sel).style.display = "block";
  evt.currentTarget.className += " active";
  sync();
}
function ZoomIn(strDiv){
  var r = $("#"+strDiv).width() / $("#"+strDiv).parent().width() * 100;
  r += 5;
  $("#"+strDiv).width(r+'%');//Math.min(95,r)
}
function ZoomOut(strDiv){
  var r = $("#"+strDiv).width() / $("#"+strDiv).parent().width() * 100;
  r -= 5;
  $("#"+strDiv).width(Math.max(10,r)+'%');
}
function showImg(eID,data,extraIMG=""){
  if(data.length<5){
    data = "iVBORw0KGgoAAAANSUhEUgAAAAUAAAAFCAYAAACNbyblAAAAEklEQVQYGWP4jwUwYBH7T6EgAG8TY53uGzeKAAAAAElFTkSuQmCC";
  }
  if (['SGVfig','HEATfig'].includes(eID)) {
    document.getElementById(eID).innerHTML = '<img id="'+eID.replace('fig','img')+'" src="data:image/'+window.figOpt['img'].replace('svg','svg+xml')+';base64,'+data+'" width="100%" height="100%"/>';
  } else {
    document.getElementById(eID).innerHTML = '<img id="'+eID.replace('fig','img')+'" src="data:image/'+window.figOpt['img'].replace('svg','svg+xml')+';base64,'+data+'" width="100%" height="auto" '+extraIMG+'/>';
  }
}
function saveImg(pngID){
  var gh = document.getElementById(pngID).src;
  var a = document.createElement('a');
  a.href = gh;
  a.download = pngID+"."+window.figOpt['img'];
  a.target="_blank";
  a.click();
}
function cellNupdate(){
  var n={'celllist1':0,'celllist2':0};
  var curCells = {};
  for(const one of Object.keys(n)){
    var curCells = window.store.getState().differential[one] || [];
    if(window.accumulateCell.cellaccumu){
      window.accumulateCell[one] = Int32Array.from([...new Set([...window.accumulateCell[one],...curCells])]);
      n[one] = window.accumulateCell[one].length;
    }else{
      n[one] = curCells.length;
    }
  }

  $('.cellN1').each(function(){
    $(this).text(n['celllist1']);
  });
  $('.cellN2').each(function(){
    $(this).text(n['celllist2']);
  });
}
function getCells(eID,sep=false){
  var cells={'celllist1':[],'celllist2':[]}, cellKey = Object.keys(cells);
  for(const one of cellKey){
    if(eID==null || $('.'+eID+'[value="'+one+'"]').prop('checked')){
      if(window.accumulateCell.cellaccumu){
        cells[one]=window.accumulateCell[one];
      }else{
        cells[one]=window.store.getState().differential[one] || [];
      }
    }
  }

  if(!sep){
    return Int32Array.from([...new Set([...cells[cellKey[0]],...cells[cellKey[1]]])]);
  }
  for(const one of cellKey){
    if(cells[one].length==0){
      delete cells[one];
    }
  }
  return cells;
}

//Page Initial functions
function createOpt(val){
  var opt = document.createElement("option");
  opt.text = val;
  opt.value = val;
  return(opt)
}
function addDescription(){
  $.ajax({
        type: "POST",
        url: VIPurl,
        data:JSON.stringify({'method':"Description",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
        contentType: 'application/json;charset=UTF-8',
        success: function(res){
          if(res.length==0){
            console.log('No description available!');
            return;
          }
          var appendDIV = $("[data-testclass='category']").parent();
          appendDIV.append("<div id='dataDescription' style='overflow-wrap: break-word;word-wrap: break-word;'>"+res+"</div>");
          for(const one of res.split("<br>")){
            if(one.startsWith("Scaling Default")){
              if(one.includes('disable')){
                $("input[name='IMGscale'][value='No']").trigger('click');
              }
            }
            if(one.startsWith("Initial Embedding")){
              window.store.dispatch({type: "set layout choice", layoutChoice: one.split(":")[1].trim()});
            }
            if(one.startsWith("Initial Coloring")){
              window.store.dispatch({type: "color by categorical metadata", colorAccessor: one.split(":")[1].trim()});
            }
            if(one.startsWith("Initial Session")){
              $("#LOADurl").val(one.split(/:(.+)/)[1].trim());
            }
            if(one.startsWith("Disable Panel")){
                var rmPanels = one.split(":")[1].split(",").map(function(item){return item.trim();});
                var a = $('.tablinks').each(function(i, obj) {
                    if(rmPanels.includes($(obj).text())){
                        $(obj).hide();
                    }
                });
            }
          }
          window.store.dispatch({type: "show centroid labels for category", showLabels: true});

        },
        error: function(request,status,error){
          console.log(request.status+': cannot get description ...');
        }
      });
}
function getGSEA(){
  $.ajax({
        type: "POST",
        url: VIPurl,
        data:JSON.stringify({'method':"GSEAgs",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
        contentType: 'application/json;charset=UTF-8',
        success: function(res){
          if(res.length==0) return;
          resD = JSON.parse(res);
          for(const one of resD){
            $('#GSEAgs').append(createOpt(one));
            $('#preDEGfgseaGS').append(createOpt(one));
          }
        },
        error: function(request,status,error){
          console.log(request.status+': cannot get gsea information ...');
        }
      });
}
function enablePreDEG(){
  $.ajax({
        type: "POST",
        url: VIPurl,
        data:JSON.stringify({'method':"preDEGname",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
        contentType: 'application/json;charset=UTF-8',
        success: function(res){
          if(res.length==0) return;
          resD = JSON.parse(res);
          $("#preDEGpanelBT").show();

          var htmlCheck="", tb=[];
          for(const one of resD){
            $('#preDEGvolcanoTab').append(createOpt(one));
            $('#preDEGfgseaTab').append(createOpt(one));
            //htmlCheck += '<label id="preDEGmulticompCK'+one+'"><input class="preDEGmultigrps" type="checkbox" value="'+one+'" onchange="preDEGmultiorderUpdate()"/> '+one+' </label>';//
          }//preDEGmulticomp
          $.each(resD,function(i,v){
            var one = v.split(/;(.+)/);
            if(one.length>1){
              tb.push(one[0].split("::").concat([one[1]]));
            }else{
              tb.push(one[0].split("::").concat([""]));
            }
          });
          var table = $('#preDEGmulticompTable').DataTable({
            destroy: true,
            select: true,
            lengthMenu:[[10,50,-1],[10,50,'all']],
            data: tb,
            columns:[{title:'Comparison'},{title:"Cell Type"},{title:"Tags"}]
          });
          table.on('select',function(e, dt, type, indexes){
            /*
            var selRow=[];
            $.each(table.rows(indexes).data().toArray(),function(i,v){
              var one = v[0]+"::"+v[1];
              if(v[2].length>0){
                one += ";"+v[2];
              }
              selRow.push(one);
            });*/
            preDEGmultiorderUpdate(table);
          }).on('deselect',function(e, dt, type, indexes){
            preDEGmultiorderUpdate(table);
          });

          //$('#preDEGmulticomp').html(htmlCheck);
        },
        error: function(request,status,error){
          console.log(request.status+': cannot get preDEG information ...');
        }
      });
}
function enableMetaCell(){
  $.ajax({
        type: "POST",
        url: VIPurl,
        data:JSON.stringify({'method':"isMeta",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
        contentType: 'application/json;charset=UTF-8',
        success: function(res){
          if(res.length==0) return;
          if(res == "TRUE"){
            $("#metaCellBT").show();
          }
        },
        error: function(request,status,error){
          console.log(request.status+': cannot get metaCell information ...');
        }
      });
}
function enableVIPtest(){
  $.ajax({
        type: "POST",
        url: VIPurl,
        data:JSON.stringify({'method':"testVIPready",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
        contentType: 'application/json;charset=UTF-8',
        success: function(res){
          $("#tVIPtest").prop('disabled',true);
          if(res == "TRUE"){
            $("#VIPtestBT").show();
          }
          if(res == "SHOW"){
            $("#tVIPtest").prop('disabled',false);
            $("#VIPtestBT").show();
          }
        },
        error: function(request,status,error){
          console.log(request.status+': cannot get VIP test information ...');
        }
      });

}
function enableBrowser(){
  $.ajax({
    type: "POST",
    url: VIPurl,
    data:JSON.stringify({'method':"getBWinfo",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      var browserinfo= JSON.parse(res);
      if(browserinfo['BWfile'].length > 0 && browserinfo['BWcluster'].length>0){
        //available track table
        var BWcluster=browserinfo['BWcluster'];
        var tableD = d3.csvParse(BWcluster,function(data){
          return Object.values(data);
        });
        var tableTitle = [];
        tableD.columns.forEach(function(d){tableTitle = tableTitle.concat({title:d})});
        var table = $('#BROWtable').DataTable({
          destroy: true,
          lengthMenu:[[10,50,-1],[10,50,"All"]],
          data:tableD,
          columns:tableTitle
        });
        $('#BROWtable tbody').on('click',"tr",function(){
          var data = table.row(this).data();
          BROWupdateTrack(data[0],null);
        });
        BROWreset();
        var html="";
        if(browserinfo['BWpeak'].length>4) html += "<br><br><strong>All peaks are defined</strong>";
        if(browserinfo['BWlink'].length>4) html += "<br><strong>Links between peaks and genes are avaiable</strong>";
        $("#BROWinfo").html(html.replaceAll('.bw',''));
        $("input[name='BROWinputRadio'][value='gene']").prop('disabled', true);
        if(browserinfo['BWannotation'].length>4) $("input[name='BROWinputRadio'][value='gene']").prop('disabled', false);
        if(browserinfo['BWcluster'].length>4) $("#BROWselGeneShow").show();
        $("#browserBT").show();
      }
    },
    error: function(request,status,error){
      console.log(request.status+': cannot get bigwig information ...');
    }
  });
}
function enableCOXMX(){
    $.ajax({
    type: "POST",
    url: VIPurl,
    data:JSON.stringify({'method':"checkCosMx",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      cosMxIDs= JSON.parse(res);
      if(cosMxIDs == null || cosMxIDs.length == 0){
          return;
      }
      //available coxMx samples
      var htmlCheck = "";
      for(const s of cosMxIDs){
          htmlCheck += '<label id="COSMXsampleCK_'+s+'"><input class="COSMXsamples" type="checkbox" value="'+s+'"/>'+s+'</label>';
      }
      $('#COSMXsample').append(htmlCheck);
      $("#cosMxBT").show();
    },
    error: function(request,status,error){
      console.log(request.status+': cannot get cosMx information ...');
    }
  });
}
function enableVISIUM(){
    $.ajax({
    type: "POST",
    url: VIPurl,
    data:JSON.stringify({'method':"checkVisium",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      VisiumIDs= JSON.parse(res);
      if(VisiumIDs == null || VisiumIDs.length == 0){
          return;
      }
      //available coxMx samples
      var htmlCheck = "";
      for(const s of VisiumIDs){
          htmlCheck += '<label id="VISIUMsampleCK_'+s+'"><input class="VISIUMsamples" type="checkbox" value="'+s+'"/>'+s+'</label>';
      }
      $('#VISIUMsample').append(htmlCheck);
      $("#visiumBT").show();
    },
    error: function(request,status,error){
      console.log(request.status+': cannot get visium information ...');
    }
  });
}
function init(){
  if(!('bioInit' in window)){
    //----------------
    $("#panel-left").prop('style')['width'] = "20%";
    $("#separator").prop('style')['left'] = "20%";
    $("#panel-right").prop('style')['width'] = "80%";
    //--------------------------------------------------------------------------
    //ABBRinit
    var grps = Object.keys(window.store.getState().categoricalSelection).sort();
    var grpL = window.store.getState().annoMatrix.schema.annotations.obsByName;
    var htmlCheck= "";
    for(i=0;i<grps.length;i++){
      var one = grps[i];
      if (one.startsWith(">")) continue;
      htmlCheck += '<label><input class="ABBRgrps" type="checkbox" value="'+grps[i]+'" onclick="ABBRlist(\'ABBR\')"/> '+grps[i]+' ('+grpL[grps[i]].categories.length+')</label>';
    }
    $('#ABBRgrp').html(htmlCheck);
    $('#ABBRcomBgrp').html(htmlCheck.replaceAll('ABBR','ABBRcomB'));

    var bioAbbr = {}
    for(i=0;i<grps.length;i++){
      var one = {}
      for(var j=0;j<grpL[grps[i]].categories.length;j++){
        if (grpL[grps[i]].categories[j] != null) {
          one[grpL[grps[i]].categories[j]] = grpL[grps[i]].categories[j].toString();
		} else {
          one[grpL[grps[i]].categories[j]] = "NA";
        }
      }
      bioAbbr[grps[i]] = one;
    }
    window.bioAbbr=bioAbbr;
    window.bioCombine={};
    window.bioCombineCell={};
    window.biogen=[];
    window.bioGeneGrp={};
    window.accumulateCell={celllist1:[],celllist2:[],cellaccumu:false};
    window.figOpt={scale:'No',scaleZero:'No',clipValue:'Yes',scaleMax:20,dpi:150,img:'png',fontsize:11,vectorFriendly:'Yes',transparent:'No',scanpybranch:'split_show',colorMap:'viridis_r'};
    //if (grpL[">Description"] !== undefined && grpL[">Description"].categories[0].split(";").length>1 && !grpL[">Description"].categories[0].split(";")[1].includes("scale")) {
    //  window.figOpt['scale'] = "Yes";
    //}

    //--------------------------------------------------------------------------
    //add dropdown annotation categories
    var sgv = document.getElementById("SGVgrp");
    var sgvGrp = document.getElementById("SGVcomparegrp");
    var sgvOrder = document.getElementById("SGVcompareorder");
    var pgv = document.getElementById("PGVgrp");
    var pgvRow = document.getElementById("PGVcompareRow");
    var pgvGrp = document.getElementById("PGVcompareGrp");
    var heat = document.getElementById("HEATorder");
    var dot = document.getElementById("DOTgrp");
    var trak = document.getElementById("TRAKgrp");
    var mark = document.getElementById("MARKgrp");
    var densS = document.getElementById("DENSsplit");
    var densG = document.getElementById("DENSgrp");
    var embed = document.getElementById("EMBEDsplit");
    var tVIP = document.getElementById("tVIPgrp");
    var GSPgrp = document.getElementById("GSPgrp");

    for(const one of grps){
      if (one.startsWith(">")) continue;
      sgv.add(createOpt(one));
      sgvGrp.add(createOpt(one));
      sgvOrder.add(createOpt(one));
      pgv.add(createOpt(one));
      pgvRow.add(createOpt(one));
      pgvGrp.add(createOpt(one));
      heat.add(createOpt(one));
      dot.add(createOpt(one));
      trak.add(createOpt(one));
      mark.add(createOpt(one));
      densS.add(createOpt(one));
      densG.add(createOpt(one));
      embed.add(createOpt(one));
      tVIP.add(createOpt(one));
      GSPgrp.add(createOpt(one));
    }
    densS.add(createOpt('None'));
    densG.add(createOpt('None'));
    //--------------------------------------------------------------------------
    //add layout to EMBED, DUAL, and CLI checkbox
    var embedLay = document.getElementById("EMBEDlayout");
    var dualLay = document.getElementById("DUALlayout");
    htmlCheck="";
    for(const ix of Object.keys(window.store.getState().annoMatrix.schema.layout.obs)){
      var one = window.store.getState().annoMatrix.schema.layout.obs[ix].name
      embedLay.add(createOpt(one));
      dualLay.add(createOpt(one));
      htmlCheck += '<label><input class="eIDlayouts" type="checkbox" value="'+one+'"/> '+one+' </label>';
    }
    $('#CLIlayout').html(htmlCheck.replace(/eID/g,'CLI'));
    //--------------------------------------------------------------------------
    htmlCheck= "";
    var iChar = 0, charCut=50;
    for(const one of grps){
      if (one.startsWith(">")) continue;
      if(iChar!=-1){
        iChar = iChar+one.length;
      }
      if(iChar>charCut){
        htmlCheck += '<span id="eIDgrpsHideShow" style="display: none;">';
        iChar=-1;
      }
      htmlCheck += '<label id="eIDgrpsCK'+one+'"><input class="eIDgrps" type="checkbox" value="'+one+'"/> '+one+' ('+grpL[one].categories.length+')</label>';
    }
    if(iChar==-1){
      htmlCheck += '</span>&nbsp;&nbsp;<a href="#" onclick="hideShow($(\'#eIDgrpsHideShow\'),$(this));return false;"><font size="+2">+</font></a>'
    }
    //HEAT, EMBED, SANK, STACBAR, CLI, VISIUM
    $('#HEATgrp').html(htmlCheck.replace(/eID/g,'HEAT').replace(/value/g,'onchange="HEATorder()" value'));
    $('#CLIgrp').html(htmlCheck.replace(/eID/g,'CLI'));
    $('#EMBEDgrp').html(htmlCheck.replace(/eID/g,'EMBED'));
    $('#VISIUMgrp').html(htmlCheck.replace(/eID/g,'VISIUM'));
    $('#SANKgrp').html(htmlCheck.replace(/eID/g,'SANK').replace(/value/g,'onchange="SANKorderUpdate()" value'));
    $('#STACBARgrp').html(htmlCheck.replace(/eID/g,'STACBAR').replace(/value/g,'onchange="STACBARref()" value'));

    //Continuous metadata
    htmlCheck= "";
    var names = window.store.getState().annoMatrix.schema.annotations.obsByName;
    iChar = 0;
    for(const one of Object.keys(names).filter(function(e) { return names[e].type != "categorical" && names[e].type != "string" ;})) {
      if(iChar!=-1){
        iChar = iChar+one.length;
      }
      if(iChar>charCut){
        htmlCheck += '<span id="eIDmetaHideShow" style="display: none;">';
        iChar=-1;
      }
      htmlCheck += '<label id="eIDmetasCK'+one+'"><input class="eIDmetas" type="checkbox" value="'+one+'"/> '+one+'</label>';
    }
    if(iChar==-1){
      htmlCheck += '</span>&nbsp;&nbsp;<a href="#" onclick="hideShow($(\'#eIDmetaHideShow\'),$(this));return false;"><font size="+2">+</font></a>'
    }
    $('#EMBEDmeta').html(htmlCheck.replace(/eID/g,'EMBED'));
    $('#VISIUMmeta').html(htmlCheck.replace(/eID/g,'VISIUM'));
    $('#HEATmeta').html(htmlCheck.replace(/eID/g,'HEAT'));

    //SANK
    $("#SANKorderGrp").hide();
    document.getElementById("SANKorderBTh").style.display="none";

    $('.sortable-list').sortable({
        connectWith: '.sortable-list',
        placeholder: 'sortable-list-dropholder'
        //console.log("init SANK")
    });
    $('#SGVcutoff').prop('min',-5);
    $('#PGVcutoff').prop('min',-5);
    $('#GDcutoff').prop('min',-5);
    $('#DUALcutoff').prop('min',-5);
    $('#DOTcutoff').prop('min',-5);

    //add text on select brush on histogram
    var html = "<table style='border: 1px solid gray; padding: 5px;'><tr><th style='border: 1px solid gray; padding: 5px;'>Name</th><th style='border: 1px solid gray; padding: 5px;'>Low end value</th><th style='border: 1px solid gray; padding: 5px;'>High end value</th><tr>";
    $("div[id^=histogram]").each(function(){
      var aID = $(this).prop('id').replace("histogram_","");
      html += "<tr><td style='border: 1px solid gray; padding: 5px;'>" +aID+"</td>";
      html += "<td style='border: 1px solid gray; padding: 5px;'><input type='number' id='input_"+aID+"_left' step=0.01 style='width: 12ch;border: 0' readonly></td>";
      html += "<td style='border: 1px solid gray; padding: 5px;'><input type='number' id='input_"+aID+"_right' step=0.01 style='width: 12ch;border: 0' readonly></td></tr>";
    });

    $("#BRUSHset").html(html);


    $("div[id^=histogram]").each(function(){initBrush($(this).prop('id'));});
    const unsubscribe = window.store.subscribe(() => {
      if (window.store.getState()["@@undoable/filterState"].prevAction) {
        var action = window.store.getState()["@@undoable/filterState"].prevAction;
        if (action.type.includes("color by continuous metadata")) {
          initBrush('histogram_'+action.colorAccessor);
        }
        if (action.type.includes("user defined gene success") || action.type.includes("store current cell selection as differential set")) {
          sync();
        }

        if (action.type.includes("continuous metadata histogram")) {
    //          $('#SYNCnote').text(JSON.stringify(window.store.getState().continuousSelection));
          $("div[id^=histogram]").each(function(){showBrush($(this).prop('id'));});
        }
        if ( (action.type.includes("resize") || action.type.includes("zoom"))) {
		}
        if (action.type.includes("set layout choice")) {
          if (document.getElementById('spatial-layer') != null) document.getElementById('spatial-layer').remove();
		}
      }
    });

    editor = ace.edit("CLIscript");
    editor.setTheme("ace/theme/iplastic");
    editor.getSession().setMode("ace/mode/python");
    editor.getSession().setUseWrapMode(true);
    editor.setFontSize("14px");
    editor.setShowPrintMargin(false);
    editor.getSession().setValue($("#CLIVignetteCode1").html().trim());



    var observer = new ResizeObserver(function(entries) {
        editor.resize();
    });
    observer.observe(document.getElementById('CLIscript'));

	document.getElementsByClassName('tablinks')[0].click();

    // sync image with embedding when zoom in/out
    var targetNode = document.getElementById("canvas-transformation-group-x").parentNode;
    var observer = new MutationObserver(function(entries) {
      if (document.getElementById("spatial-transform-layer") && $("input[name='IMGtogether']:checked").val() == "Yes") {
		SPATIALoverlay();
      }
    });
	observer.observe(targetNode, {
      attributes: true,
      subtree: true,
      attributeOldValue: true
    });
    // Click on "cellxgene" at the top-left corner to issue 'resize' event, so out-of-window VIP window will be back
    $(".container")[0].id = "cellxgene_container";
    $("#cellxgene_container span")[0].onclick = function () {window.dispatchEvent(new Event('resize'));}

    console.log("VIP (Visualization in Plugin) is ready on client side.");

    var tryurls = [window.location.href.replace(/\/+$/,'')+"/VIP",window.location.origin+"/VIP","/VIP","VIP"];
    for (var i=0;i<tryurls.length; i++) {
      $.ajax({
        type: "POST",
        url: tryurls[i],
        data:JSON.stringify({'method':"HELLO",'dataset': window.store.getState().config.displayNames.dataset+'.h5ad'}),
        contentType: 'application/json;charset=UTF-8',
        success: function(res, status, req){
          console.log(res + " url: "+this.url);
          if (!('bioInit' in window)) {
            window.bioInit=true;
            VIPurl = this.url;
            console.log("VIP ("+VIPurl+") is ready on server side.");
            //Add Description
            console.log("Checking description ...");
            addDescription();

            //get GSEA Information
            console.log("Loading GSEA ...");
            getGSEA();

            //for pre-computed DEG
            console.log("Checking pre-computed DEG ...");
            enablePreDEG();

            //for metaCell
            console.log("Checking Meta Cell ...");
            enableMetaCell();

            //for VIP test
            console.log("Checking VIP test ...");
            enableVIPtest();

            // for browser track plot
            console.log("Checking browser track ...");
            enableBrowser();
            
            // for cosMx availability
            console.log("Checking cosMx ...");
            enableCOXMX();
            
            // for cosMx availability
            console.log("Checking visium ...");
            enableVISIUM();

            console.log("Initial server communication is completed!");
          }
        },
        error: function(request,status,error){
          //console.log(request.status+': handshake failed, keep trying ...');
        }
      });
    }
  }
}
function initBrush(aID){
  for(const i of ["left","right"]){
    var one = d3.select("[id='"+aID+"'] g.brush").append("text").attr("id",aID+"_"+i).attr("y",$('[id="'+aID+'"] rect.selection').prop("y").baseVal.value).attr("fill","#000");
    if(i=="left") one.attr("text-anchor", "end");
  }
}
function showBrush(aID){
  var val = window.store.getState().continuousSelection[aID.replace('histogram','obsAnno')];
  if(val===undefined){
    val = ["",""];
  }else{
    var valMax = Math.max(...val);
    var decimals = (valMax<10) ? 2 : (valMax<100) ? 1:0;
    val = [val[0].toFixed(decimals),val[1].toFixed(decimals)];
  }
  var pos = ['left','right']
  for(i in [0,1]){
    var one = d3.select("[id='"+aID+"_"+pos[i]+"']").attr("x",$('[id="'+aID+'"] rect.selection').prop("x").baseVal.value).text(val[i]);
    if(i==1) one.attr("x",$('[id="'+aID+'"] rect.selection').prop("x").baseVal.value+$('[id="'+aID+'"] rect.selection').prop("width").baseVal.value);
    $('[id="'+aID.replace('histogram','input')+'_'+pos[i]+'"]').val(val[i]);
  }
}
//Page sync functions
function sync(){
  cellNupdate();
  ABBRref();
  ADDGref();
  GDref();
  SGVref();
  SGVcompareref();
  PGVref();
  PGVcomparref();
  HEATref();
  DOTref();
  TRAKref();
  DUALref();
  DEGref();
  EMBEDref();
  IMGref();
  DENSref();
  DENS2Dref();
  SANKref();
  STACBARref();
  BROWref();
  CLIref();
  preDEGVolcanoref("preDEGvolcano");
  preDEGmultiref();
  GSPref();
  COSMXref();
  VISIUMref();
  if(window.store.getState().colors.colorAccessor != null){
    $('#tVIPgrp').val(window.store.getState().colors.colorAccessor);
  }
}
function ABBRref(){
}
function ADDGref(){
  var html =  "";
  for(const one of window.biogen){
    html += '<button class="link" style="cursor: crosshair;" id="ADDG_'+one+'" onclick=\'ADDGdel("'+one+'")\'>'+one+'</button>';
  }
  $("#ADDGsel").html(html);
  //$("#ADDGsel").text(window.biogen.join(", "));

  var html= "";
  var grpKeys = Object.keys(window.bioGeneGrp);
  for(const one of grpKeys){
    html += "<p><button onclick=\"ADDGsetDel('"+one+"')\"><b>Delete</b></button> <b>"+one+"</b>: " + (""+window.bioGeneGrp[one]).replaceAll(",",", ")+"</p>";
  }
  $("#ADDGsets").html(html);
}
function GDref(){
}

function selectAllCells() {
  $('[data-testid="subset-button"]').get(0).click();
  $('[data-testid="reset-subset-button"]').get(0).click();
  window.store.dispatch({type: "store current cell selection as differential set 1", data:Int32Array.from({length:window.store.getState().annoMatrix.nObs},(v,k)=>k)});
}

function selectByCelltype(celltype) {
  var selectionMap = new Map();
  var categories = window.store.getState().annoMatrix.schema.annotations.obsByName['cell_type'].categories;
  for (var j=0;j<categories.length; j++) {
    if (categories[j] == celltype) {
      selectionMap.set(categories[j], true);
    } else {
      selectionMap.set(categories[j], false);
    }
  }
  window.store.dispatch({type: "categorical metadata filter all of these", obsCrossfilter: window.store.getState().obsCrossfilter, metadataField: 'cell_type', labelSelectionState: selectionMap});
  /* The selection is a promise, an asynchronous operation */
  window.store.getState().obsCrossfilter.select("obs","cell_type",{mode:"none"}).then(function(response) {
    console.log(response.obsCrossfilter);
    window.store.getState().obsCrossfilter.obsCrossfilter = response.obsCrossfilter.select("obs/cell_type",{mode:"exact", values:celltype});
    window.store.dispatch({type: "store current cell selection as differential set 1", data:window.store.getState().obsCrossfilter.allSelectedLabels()});
    window.store.dispatch({type: "store current cell selection as differential set 2", data:null});
  });
}
function delay(time) {
  return new Promise(resolve => setTimeout(resolve, time));
}
function addCustomCombine(element) {
  var x = document.getElementById(element);
  //var grps = Object.keys(window.store.getState().categoricalSelection).filter(function(e) { return !(e.startsWith('>') || e.startsWith('|')) });
  $("#"+element+" option[value='Custom_combine']").remove();
  var grpKeys = Object.keys(window.bioCombine);
  if(grpKeys.length>0){
    x.add(createOpt("Custom_combine"));
  }
  //if(grpKeys.length>0 && grps.length==x.length){
  //    x.add(createOpt("Custom_combine"));
  //  }else if(grpKeys.length==0){// && grps.length<x.length
  //    $("#"+element+" option[value='Custom_combine']").remove();
    //x.remove(grps.length);
  //  }
}
function addCustomCheck(aID){
  $('#'+aID+'grpsCKcustom').remove();
  var grpKeys = Object.keys(window.bioCombine);
  if(grpKeys.length>0){
    var htmlCheck = $('#'+aID+'grp').html();
    htmlCheck += '<label id="'+aID+'grpsCKcustom"><input class="'+aID+'grps" type="checkbox" value="Custom_combine"/> Custom_combine ('+Object.keys(window.bioCombineCell).length+')</label>';
  }
  $('#'+aID+'grp').html(htmlCheck);
}
function MARKref(){
  addCustomCombine("MARKgrp");
}
function dropGene(aID){
  var genes = window.store.getState().controls.userDefinedGenes;
  genes = genes.concat(window.biogen);
  genes = [...new Set(genes)];
  optG = [];
  var i,x = document.getElementById(aID+"gene");
  //remove options which are not in the gene list
  $("#"+aID+"gene option").each(function(){
    optG.push($(this).val());
    if(!genes.includes($(this).val())){
      x.remove($(this).index())
    }
  });
  //add options which are missing from gene list
  if(genes.length>0){
    for(const g of genes){
      if(!optG.includes(g)){
        x.add(createOpt(g));
      }
    }
  }
}
function SGVref(){
  dropGene("SGV");
  addCustomCombine("SGVgrp");
  //if($('#SGVgene').val()) $('#SGVnote').text('');
  if($('#SGVgene').val() == null){
    $('#SGVnote').text('Please add genes either from the VIP menu or the main window');
  }else if($('#SGVnote').text().includes("VIP menu")){
    $('#SGVnote').text('');
  }
}
function SGVcompareref(){
  dropGene("SGVcompare");
  //addCustomCombine("SGVcomparegrp");
  //addCustomCombine("SGVcompareorder");

  if($('#SGVcomparegene').val() == null){
    $('#SGVcomparenote').text('Please add genes either from the VIP menu or the main window');
  }else if ($('#SGVcomparenote').text().includes("VIP menu")){
    $('#SGVcomparenote').text("");
  }
}
function showGeneSet(aID){
  var gSets = Object.keys(window.bioGeneGrp);
  var optG = [], NcheckG=[];
  //remove checkbox which are not in the gene list
  $("."+aID+"sets").each(function(){
    optG.push($(this).val());
    if(!gSets.includes($(this).val())){
      $('#'+aID+'ck'+$(this).val()).remove();
    }else if($(this).prop("checked") == false){
      NcheckG.push($(this).val());
    }
  })
  //add options which are missing from gene list
  if(gSets.length>0){
    var htmlCheck = $('#'+aID+'geneGrp').html();
    for(const g of gSets){
      if(!optG.includes(g)){
        htmlCheck += '<label id="'+aID+'ck'+g+'"><input class="'+aID+'sets" type="checkbox" value="'+g+'" checked/>'+g+'</label>';// '+g+'
      }
    }
    $('#'+aID+'geneGrp').html(htmlCheck);
  }
  $("."+aID+"sets").each(function(){
    if(NcheckG.includes($(this).val())){
      $(this).prop("checked",false);
    }
  })

}
function showGene(aID, addInfo='', ckDefault='checked'){
  //console.log(aID+":"+ckDefault+":"+addInfo);
  var genes = window.store.getState().controls.userDefinedGenes;
  genes = genes.concat(window.biogen);
  genes = [...new Set(genes)];
  var optG = [], NcheckG=[],checkG=[];
  //remove checkbox which are not in the gene list
  $("."+aID+"genes").each(function(){
    optG.push($(this).val());
    if(!genes.includes($(this).val())){
      $('#'+aID+'ck'+$(this).val()).remove();
    }else if($(this).prop("checked") === false){// == false
      NcheckG.push($(this).val());
    }else if($(this).prop("checked")){
      checkG.push($(this).val());
    }
  });
  //add options which are missing from gene list
  var missG = genes.filter(i=>!optG.includes(i));
  if(missG.length>0){
    var htmlCheck = "";//$('#'+aID+'gene').html();
    for(const g of missG){
      htmlCheck += '<label id="'+aID+'ck'+g+'"><input class="'+aID+'genes" type="checkbox" value="'+g+'" '+ckDefault+' '+addInfo+'/>'+g+'</label>';// '+g+'
    }
    $('#'+aID+'gene').append(htmlCheck);//html(htmlCheck);
  }
  //$("."+aID+"genes").each(function(){
  //  if(NcheckG.includes($(this).val())){
  //    $(this).prop("checked",false);
  //  }else if(checkG.includes($(this).val())){
  //    $(this).prop("checked",true);
  //  }
  //});
  if(ckDefault==='checked') $.merge(checkG,missG);
  return checkG;
}
function showGeneGrpCollSet(aID){
  if(Object.keys(window.bioGeneGrp).length>0){
    $('#'+aID+'geneGrpCollapsing').show();
  }else{
    $('#'+aID+'geneGrpCollapsing').hide();
  }
  if($("input[name='"+aID+"geneGrpColl']:checked").val() == 'Yes'){
    $('#'+aID+'geneGrpCollShow').show();
  }else{
    $('#'+aID+'geneGrpCollShow').hide();
  }
}
function PGVref(){
  showGene('PGV');
  showGeneSet('PGV');
  addCustomCombine("PGVgrp");
  radioCheckTriger("PGVgeneGrpColl");
}
function PGVcomparref(){
  showGene('PGVcompare');
}
function GSPref(){
  showGene("GSP");
}
function HEATref(){
  showGene('HEAT');

  var nCells = 0;
  var x = document.getElementsByClassName('HEATcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked && !!window.store.getState().differential[x[i].value]){
      nCells += window.store.getState().differential[x[i].value].length;
    }
  }
  var gL = Object.keys(window.store.getState().categoricalSelection).length-1; //not considering the last one ">Description"
  addCustomCombine("HEATorder");

  var HEATorder=$.map($("#HEATorder option"),function(a) {
    return a.value;
  });
  var HEATcomplexSortby=$.map($("#HEATcomplexSortby option"),function(a) {
    return a.value;
  });
  if(nCells<heatCell){
    if(!HEATorder.includes("Expression")){
      x = document.getElementById("HEATorder");
      x.add(createOpt("Expression"));
    }
    if(!HEATcomplexSortby.includes("Expression")){
      x = document.getElementById("HEATcomplexSortby");
      x.add(createOpt("Expression"));
    }
  }else{
    if(HEATorder.includes("Expression")){
      $("#HEATorder option[value='Expression']").remove();
    }
    if(HEATcomplexSortby.includes("Expression")){
      $("#HEATcomplexSortby option[value='Expression']").remove();
      $("#HEATcomplexSortby").change();
    }
  }

  addCustomCheck("HEAT");
}
function HEATrm(eID){
  console.log(eID);
  var selElem = $("#HEATdynList").sortable('toArray');
  selElem.splice(selElem.indexOf(eID),1);
  var htmlCheck = "";
  for(const one of selElem){
    htmlCheck += "<li id='"+one+"'><label>"+one.replace("li","")+"</label> <a onClick='HEATrm(\'"+one+"\')' style='cursor: pointer; cursor: hand;'><i class='fa fa-trash-o'></i></a></li>";
  }
  $("#HEATdynList").html(htmlCheck);
}
function HEATorder(){
  var eID="HEAT";
  selElem=[];
  $("."+eID+"grps").each(function(){
    if($(this).prop("checked")){
      selElem.push($(this).val());
  }})
  var htmlCheck = "";
  for(const one of selElem){
    htmlCheck += "<li id='"+one+"'><label>"+one.replace(eID+"li","")+"</label> <a onClick=\"HEATrm('"+one+"')\" style='cursor: pointer; cursor: hand;'><i class='fa fa-trash-o'></i></a></li>";
  }
  $("#"+eID+"dynList").html(htmlCheck);
}
function DOTref(){
  showGene('DOT');
  showGeneSet('DOT');
  addCustomCombine("DOTgrp");
  showGeneGrpCollSet("DOT");
}
function TRAKref(){
  showGene('TRAK');
  showGeneSet('TRAK');
  addCustomCombine("TRAKgrp");
  showGeneGrpCollSet("TRAK");
}
function DENSref(){
  var genes = window.store.getState().controls.userDefinedGenes;
  genes = genes.concat(window.biogen);
  genes = [...new Set(genes)];
  var optG = [], NcheckG=[];
  //remove checkbox which are not in the gene list
  $(".DENSgenes").each(function(){
    optG.push($(this).val());
    if(!genes.includes($(this).val())){
      $('#DENSck'+$(this).val()).remove();
    }else if($(this).prop("checked") == false){
      NcheckG.push($(this).val());
    }
  })
  //add options which are missing from gene list
  if(genes.length>0){
    var htmlCheck = $('#DENSgene').html();
    for(const g of genes){
      if(!optG.includes(g)){
        htmlCheck += '<label id="DENSck'+g+'"><input class="DENSgenes" type="checkbox" value="'+g+'" checked/>'+g+'</label>';//
      }
    }
    $('#DENSgene').html(htmlCheck);
  }
  $(".DENSgenes").each(function(){
    if(NcheckG.includes($(this).val())){
      $(this).prop("checked",false);
    }
  })
  addCustomCombine("DENSgrp");
  addCustomCombine("DENSsplit");
}
function DENS2Dref(){
  if(false){
    var genes = window.store.getState().controls.userDefinedGenes;
    genes = genes.concat(window.biogen);
    genes = [...new Set(genes)];
    var optG = [], checkG=[];
    //remove checkbox which are not in the gene list
    $(".DENS2Dgenes").each(function(){
      optG.push($(this).val());
      if(!genes.includes($(this).val())){
        $('#DENS2Dck'+$(this).val()).remove();
      }else if($(this).prop("checked")){
        checkG.push($(this).val());
      }
    })
    //add options which are missing from gene list
    var missG = genes.filter(i=>!optG.includes(i));
    if(missG.length>0){
      var htmlCheck = "";//$('#DENS2Dgene').html();
      for(const g of missG){
        htmlCheck += '<label id="DENS2Dck'+g+'"><input class="DENS2Dgenes" type="checkbox" value="'+g+'" onclick="DENS2Dref()"/>'+g+'</label>';// '+g+'
      }
      $('#DENS2Dgene').append(htmlCheck);
    }
  }
  var iEnabled=showGene('DENS2D','onclick="DENS2Dref()"','').length;
  //$(".DENS2Dgenes").each(function(){
  //  if(checkG.includes($(this).val())){
  //    $(this).prop("checked",true);
  //    iEnabled++;
  //  }
  //})
  if(iEnabled>1){
    $(".DENS2Dgenes:checkbox:not(:checked)").prop("disabled",true);
  }else{
    $('.DENS2Dgenes').prop("disabled",false);
  }

}
function EMBEDref(){
  showGene('EMBED');
}
function DUALref(){
  if(false){
    var genes = window.store.getState().controls.userDefinedGenes;
    genes = genes.concat(window.biogen);
    genes = [...new Set(genes)];
    var optG = [], checkG=[];
    //remove checkbox which are not in the gene list
    $(".DUALgenes").each(function(){
      optG.push($(this).val());
      if(!genes.includes($(this).val())){
        $('#DUALck'+$(this).val()).remove();
      }else if($(this).prop("checked")){
        checkG.push($(this).val());
      }
    })
    //add options which are missing from gene list
    if(genes.length>0){
      var htmlCheck = $('#DUALgene').html();
      for(const g of genes){
        if(!optG.includes(g)){
          htmlCheck += '<label id="DUALck'+g+'"><input class="DUALgenes" type="checkbox" value="'+g+'" onclick="DUALref()"/>'+g+'</label>';// '+g+'
        }
      }
      $('#DUALgene').html(htmlCheck);
    }
  }
  var iEnabled = showGene('DUAL','onclick="DUALref()"','').length;
  if(iEnabled>1){
    $(".DUALgenes:checkbox:not(:checked)").prop("disabled",true);
  }else{
    $('.DUALgenes').prop("disabled",false);
  }

}
function orderUpdate(eID,choseGene,selElem=[]){
  //var selElem = [];
  $("."+eID+"grps").each(function(){
    if($(this).prop("checked")){
      selElem.push($(this).val());
  }})
  if(choseGene) selElem = selElem.concat(getGenes(eID));
  selElem=selElem.map(function(one){return eID+'li'+one;})

  for(const one of $("#"+eID+"dynList").sortable('toArray')){
    if(!selElem.includes(one)){
      $(document.getElementById(one)).remove();
    }
  }
  selElem = [...new Set($("#"+eID+"dynList").sortable('toArray').concat(selElem))];

  var htmlCheck = "";
  for(const one of selElem){
    htmlCheck += "<li id='"+one+"'><label>"+one.replace(eID+"li","")+"</label></li>";
  }
  $("#"+eID+"dynList").html(htmlCheck);
}
function SANKref(){
  showGene('SANK','onchange="SANKorderUpdate()"');
  //showGeneSet('SANK');
  SANKorderUpdate();
}
function SANKorderUpdate(){
  orderUpdate("SANK",true);
  return;
  var selSank = [];
  $(".SANKgrps").each(function(){
    if($(this).prop("checked")){
      selSank.push($(this).val());
  }})
  selSank = selSank.concat(getGenes('SANK'));
  selSank=selSank.map(function(one){return 'SANKli'+one;})

  for(const one of $("#SANKdynList").sortable('toArray')){
    if(!selSank.includes(one)){
      $(document.getElementById(one)).remove();
    }
  }
  selSank = [...new Set($("#SANKdynList").sortable('toArray').concat(selSank))];

  var htmlCheck = "";
  for(const one of selSank){
    htmlCheck += "<li id='"+one+"'><label>"+one.replace("SANKli","")+"</label></li>";
  }
  $("#SANKdynList").html(htmlCheck);
}
function SANKorderShow(){
  $("#SANKorderGrp").show();
	document.getElementById("SANKorderBTs").style.display="none";
  document.getElementById("SANKorderBTh").style.display="inline";
}
function SANKorderHide(){
  $("#SANKorderGrp").hide();
	document.getElementById("SANKorderBTh").style.display="none";
  document.getElementById("SANKorderBTs").style.display="inline";
}
function STACBARref(){
  var selGene=showGene('STACBAR','onchange="STACBARref()"','');
  $("#STACBARcol").empty();
  var colSel = document.getElementById("STACBARcol");
  for(const one of selGene){
    colSel.add(createOpt(one));
  }
  var iEnabled=selGene.length;
  //$(".STACBARgenes").each(function(){
  //  if($(this).prop("checked")){
  //    colSel.add(createOpt($(this).val()));
  //  }
  //})
  $("#STACBARgeneSel").hide();
  if(iEnabled>0){
    $("#STACBARgeneSel").show();
  }
  $(".STACBARgrps").each(function(){
    if($(this).prop("checked")){
      iEnabled++;
      colSel.add(createOpt($(this).val()));
    }
  })
  if(iEnabled>1){
    $(".STACBARgenes:checkbox:not(:checked)").prop("disabled",true);
    $(".STACBARgrps:checkbox:not(:checked)").prop("disabled",true);
  }else{
    $(".STACBARgenes").prop("disabled",false);
    $(".STACBARgrps").prop("disabled",false);
  }

}
function BROWref(){
  //var selGene=showGene('BROW','onchange="BROWref()"','');
  showGene('BROW','','');
}
function radioCheckTriger(eName){
  $("input[name='"+eName+"']:checked").trigger("click");
}
function DEGref(){
  showGene('DEG',"","");
  showGene('DEGselgene',"","");
  var cName=[],cVal=Object.keys(window.bioCombineCell);
  for(const one of Object.values(window.bioCombineCell)){
    cName.push(one.cName);
  }

  var optG = [], checkG=[];
  //remove checkbox which are not in the annotation list
  $(".DEGgrps").each(function(){
    optG.push($(this).val());
    if(!cVal.includes($(this).val())){
      $('#DEGck'+$(this).val()).remove();
    }else if($(this).prop("checked")){
      checkG.push($(this).val());
    }
  })
  //add options which are missing from annotation list
  if(cVal.length>0){
    var htmlCheck = $('#DEGgrp').html();
    for(var i=0;i<cVal.length;i++){
      if(!optG.includes(cVal[i])){
        htmlCheck += '<label id="DEGck'+cVal[i]+'"><input class="DEGgrps" type="checkbox" value="'+cVal[i]+'" onclick="DEGref()"/>'+cName[i]+'</label>';
      }
    }
    $('#DEGgrp').html(htmlCheck);
  }

  //check the originally checked
  var iEnabled=0;
  $(".DEGgrps").each(function(){
    if(checkG.includes($(this).val())){
      $(this).prop("checked",true);
      iEnabled++;
    }
  })
  //disable when 2 are checked or enable when less than 2 is checked
  if(iEnabled>1){
    $(".DEGgrps").each(function(){
      if($(this).prop("checked")==false){
        $(this).prop("disabled",true);
      }
    })
  }else{
    $('.DEGgrps').prop("disabled",false);
  }
  radioCheckTriger("DEGcolorBy");
  if($("#DEGfig").html().length==0 && !$("#DEGinteractiveBtn").prop('disabled')){
    $("#DEGfinal").hide();
    $("#DEGinteractiveBtn").prop('disabled', true);
  }
}
function preDEGVolcanoref(eID){
  showGene(eID,"","");
  showGene(eID+'selgene',"","");
  //console.log(eID);

  radioCheckTriger(eID+"colorBy");
  if($("#"+eID+"fig").html().length==0 && !$("#"+eID+"interactiveBtn").prop('disabled')){
    $("#"+eID+"final").hide();
    $("#"+eID+"interactiveBtn").prop('disabled', true);
  }
}
function preDEGmultiref(){
  showGene('preDEGmulti',"","");
}
function preDEGmultiorderUpdate(tb){
  //orderUpdate("preDEGmulti",false);
  var selRow=[];
  $.each(tb.rows({selected:true}).data().toArray(),function(i,v){
    var one = v[0]+"::"+v[1];
    if(v[2].length>0){
      one += ";"+v[2];
    }
    selRow.push(one);
  });
  orderUpdate("preDEGmulti",false,selRow);
}
function IMGref(){
  $('#IMGscaleZero').prop('checked',(window.figOpt['scaleZero']=="Yes"));
  $('#IMGclipValue').prop('checked',(window.figOpt['clipValue']=="Yes"));
  $('#IMGscaleMax').val(window.figOpt['scaleMax']);
  if(window.figOpt['scale'] == "Yes"){
    $('#IMGscaleShow').show();
    $("input[name='IMGscale'][value='Yes']").prop('checked', true);
  }else{
    $('#IMGscaleShow').hide();
    $("input[name='IMGscale'][value='No']").prop('checked', true);
  }

  $('#IMGdpi').val(window.figOpt['dpi']);
  $('#IMGimg').val(window.figOpt['img']);
  $('#IMGfontsize').val(window.figOpt['fontsize']);
  $('#IMGvector').val(window.figOpt['vectorFriendly']);
  $('#IMGtransparent').val(window.figOpt['transparent']);
  $('#IMGscanpybranch').val(window.figOpt['scanpybranch']);
  $('#IMGcolor').val(window.figOpt['colorMap'].replace('_r',''));
  $('#IMGcolor_r').prop('checked',window.figOpt['colorMap'].includes('_r'));
}
function IMGcellSel(){
  //IMGcumuRd
  window.accumulateCell.cellaccumu=$("input[name='IMGcumuRd']:checked").val()=='Yes';
  /*
    $("#IMGrandomCellDiv").hide();
    if(window.accumulateCell.cellaccumu){
      $("#IMGrandomCellDiv").show();
    }
  */
  cellNupdate();
}
function IMGcellClear(){
  // cellNupdate();
  window.store.dispatch({type: "store current cell selection as differential set 1",data:null});
  window.store.dispatch({type: "store current cell selection as differential set 2",data:null});
  $('.cellN1').each(function(){
    $(this).text(0);
  });
  $('.cellN2').each(function(){
    $(this).text(0);
  });
  window.accumulateCell.celllist1=[];
  window.accumulateCell.celllist2=[];
}
function IMGcellRandomChoose(){
  var cGrp = $("input[name='IMGcell']:checked").val();
  var cNum = $('#IMGcellRandomN').val();
  if(window.accumulateCell[cGrp].length<cNum || cNum==0){
    alert("Cannot randomly select "+cNum+" cells from "+window.accumulateCell[cGrp].length+" cells");
    return;
  }
  var sel=[... new Set(Array.from({length: cNum*2}, () => Math.floor(Math.random() * window.accumulateCell[cGrp].length)))];
  sel = sel.slice(0,Math.min(cNum,sel.length));
  window.accumulateCell[cGrp] = Int32Array.from(sel.map(i=>window.accumulateCell[cGrp][i]));
  window.store.dispatch({type: "store current cell selection as differential set "+cGrp.slice(-1),data:window.accumulateCell[cGrp]});
  cellNupdate();
}
function SpatialIMGfront() {
  if ($("input[name='IMGfront']:checked").val() == "Yes") {
    document.getElementById('spatial-layer').style.zIndex='0';

    document.getElementsByClassName('graph-canvas')[0].style.opacity='1';
    document.getElementById('embedding_opacity').value=1;
    $('#embedding_opacity_val').text('1');

    $('#spatial-layer')[0].style.opacity='0.75';
    document.getElementById('image_opacity').value=0.75;
    $('#image_opacity_val').text('0.75');
  } else {
    document.getElementById('spatial-layer').style.zIndex='-1';

    document.getElementsByClassName('graph-canvas')[0].style.opacity='0.35';
    document.getElementById('embedding_opacity').value=0.35;
    $('#embedding_opacity_val').text('0.35');

    $('#spatial-layer')[0].style.opacity='1';
    document.getElementById('image_opacity').value=1;
    $('#image_opacity_val').text('1');
  }
}
function SpatialIMGopacity(val) {
  $('#spatial-layer')[0].style.opacity=val;
  $('#image_opacity_val').text(val);
}
function SpatialIMGman() {
  $('#image_rotate_val').text($('#image_rotate').val());
  var trans ="";
  if ($('#flipx').is(':checked')) trans += 'scaleX(-1)';
  if ($('#flipy').is(':checked')) trans += ' scaleY(-1)';
  if ($('#image_rotate').val() !=0 ) trans += ' rotate(' + $('#image_rotate').val() +'deg)';
  $('#'+$('#selected_image').text()+'_img').css({transform: trans});
}
function SpatialEMBopacity(val) {
  document.getElementsByClassName('graph-canvas')[0].style.opacity=val;
  $('#embedding_opacity_val').text(val);
}
function COSMXref(){
    showGene('COSMX','onchange="COSMXsetGeneColor()"','');
    COSMXsetGeneColor();
}
function COSMXsetGeneColor(){
    selG = getGenes("COSMX");
    if(selG.length>5){
        $(".COSMXgenes:checkbox:not(:checked)").prop("disabled",true);
    }else{
        $(".COSMXgenes").prop("disabled",false);
    }
    html = "";
    for(const one of selG){
        html += "<label>"+one+"<input class='COSMXgeneColors colorInput' id='COSMXgeneColor_"+one+"'";
        if($("#COSMXgeneColor_"+one).val()!==undefined){
            html += " value='"+$("#COSMXgeneColor_"+one).val()+"'";
        }
        html += " type='color'></label>";
    }
    $("#COSMXgeneColor").html(html);
}
function VISIUMref(){
    showGene('VISIUM');
}
function IMGset(){
  window.figOpt['scale'] = $("input[name='IMGscale']:checked").val();
  if(window.figOpt['scale'] == "Yes"){
    $('#IMGscaleShow').show();
    window.figOpt['scaleZero'] = $('#IMGscaleZero').prop('checked')?"Yes":"No";
    window.figOpt['clipValue'] = $('#IMGclipValue').prop('checked')?"Yes":"No";
    window.figOpt['scaleMax'] = $('#IMGscaleMax').val();
  }else{
    $('#IMGscaleShow').hide();
  }

  window.figOpt['dpi'] = $('#IMGdpi').val();
  window.figOpt['img'] = $('#IMGimg').val();
  window.figOpt['fontsize'] = Math.max($('#IMGfontsize').val(),$('#IMGfontsize').prop('min'));
  window.figOpt['vectorFriendly'] = $("input[name='IMGvector']:checked").val();
  window.figOpt['transparent'] = $("input[name='IMGtransparent']:checked").val();
  window.figOpt['scanpybranch'] = 'split_show'; //$("input[name='IMGscanpybranch']:checked").val();
  window.figOpt['colorMap'] = $('#IMGcolor').val() + ($('#IMGcolor_r').prop('checked')?'_r':'');
console.log(window.figOpt['colorMap']);
  //IMGref();
}
function CLIref(){
  showGene('CLI');
  $(".CLIgrps:checkbox[value='cell_type']").prop("checked","true");
  $(".CLIlayouts:checkbox[value='umap_harmony']").prop("checked","true");
}
// process annotation
function ABBRlist(eID){
  var grps = {},nrows=0,i;
  $.each($('.'+eID+'grps:checkbox:checked'),function(index, one){
    grps[one.value]=window.store.getState().annoMatrix.schema.annotations.obsByName[one.value].categories.map(String);
    nrows = Math.max(nrows,grps[one.value].length);
  });

  var grpKey = Object.keys(grps);
  var htmlTable = "<table style='width:100%;border-collapse: collapse;'><tr>";
  for(i=0;i<grpKey.length;i++){
    htmlTable += "<th colspan='2' style='background-color: #4CAF50;color: white;'><label>"
    if(eID=="ABBRcomB"){
      htmlTable += '<input type="checkbox" onclick="$(\'.ABBRcomBgrp_' + $.escapeSelector(grpKey[i]) + '\').prop(\'checked\',$(this).prop(\'checked\'))" checked/>';
    }
    htmlTable += grpKey[i]+"</label></th>"
  }
  htmlTable += "</tr><tr>";
  for(i=0;i<nrows;i++){
    for(var j=0;j<grpKey.length;j++){
      if(j>0){
        htmlTable += "<td align='right' style='border-left:2px solid #888'>"
      }else{
        htmlTable += "<td align='right'>"
      }
      if(i<grps[grpKey[j]].length){//grpKey[j]+"','"+grps[grpKey[j]][i]
        if(eID=="ABBRcomB"){
          htmlTable += '<input class="ABBRcomBgrp_'+grpKey[j]+'" type="checkbox" value="'+grps[grpKey[j]][i]+'" checked/> </td><td>' + grps[grpKey[j]][i] + '</td>';
        }else{
          htmlTable += grps[grpKey[j]][i]+'</td><td><input type="text" id="'+grpKey[j]+"|"+grps[grpKey[j]][i]+'" value="' +window.bioAbbr[grpKey[j]][grps[grpKey[j]][i]]+'" onchange=ABBRupdate(["'+grpKey[j].replace(/ /g,";;")+'","'+grps[grpKey[j]][i].replace(/ /g,";;")+'"])></td>';
        }
      }else{
        htmlTable += "</td><td></td>";
      }
    }
    htmlTable += "</tr><tr>"
  }
  htmlTable += "</tr></table>"
  $('#'+eID+'table').html(htmlTable);
}
function ABBRcomBlist(eID){
  ABBRlist(eID)
}
function ABBRupdate(eID){
  for(var i=0;i<eID.length;i++){
    eID[i] = (eID[i].replace(/;;/g," ")).replace(/\./g,"\\.");
  }
  window.bioAbbr[eID[0]][eID[1]] = $('#'+eID.join("\\|").replace(/ /g,"\\ ")).val();
}
function ABBRclear(){
  var grpKeys = Object.keys(window.store.getState().categoricalSelection);
  for(var i=0;i<grpKeys.length;i++) {
    var selectionMap = new Map();
    var categories = window.store.getState().annoMatrix.schema.annotations.obsByName[grpKeys[i]].categories;
    for (var j=0;j<categories.length; j++) {
      selectionMap.set(categories[j], false);
    }
    window.store.dispatch({type: "categorical metadata filter none of these", metadataField: grpKeys[i], labelSelectionState: selectionMap});
  }
  return;
}
function ABBRall(){
  selectAllCells();
  return;
  var grpKeys = Object.keys(window.store.getState().categoricalSelection);
  for(var i=0;i<grpKeys.length;i++) {
    var selectionMap = new Map();
    var categories = window.store.getState().annoMatrix.schema.annotations.obsByName[grpKeys[i]].categories;
    for (var j=0;j<categories.length; j++) {
      selectionMap.set(categories[j], true);
    }
    window.store.dispatch({type: "categorical metadata filter all of these", metadataField: grpKeys[i], labelSelectionState: selectionMap});
  }
  return;
}
//const indexOfAll = (arr, val) => arr.reduce((acc, el, i) => (el === val ? [...acc, i] : acc), []);
function ABBRcombineCell(){
  var annKeys = Object.keys(window.store.getState().categoricalSelection);
  var obsMap = window.store.getState().annoMatrix._cache.obs.colIndex.index;
  var grpKeys = Object.keys(window.bioCombine);
  var cName = [], cVal = [], combCell = {};//, selC=[]
  for(const key of grpKeys){
    var k = annKeys.indexOf(key);
    if(cName.length==0){
      for(const v of window.bioCombine[key]){
        cName.push(window.bioAbbr[key][v]);
        cVal.push(k+"_"+window.store.getState().annoMatrix.schema.annotations.obsByName[key].categories.indexOf(v));
        //selC.push(indexOfAll(window.store.getState().annoMatrix._cache.obs.__columns[obsMap.get(key)],v));
      }
    }else{
      var newGrp = [], newVal = [];//, newC = []
      for(var i=0;i<cName.length;i++){
        window.bioCombine[key].forEach(function(v){
          newGrp.push(cName[i]+":"+window.bioAbbr[key][v]);
          newVal.push(cVal[i]+"__"+k+"_"+window.store.getState().annoMatrix.schema.annotations.obsByName[key].categories.indexOf(v));
          //var oneSel = indexOfAll(window.store.getState().annoMatrix._cache.obs.__columns[obsMap.get(key)],v);
          //newC.push(selC[i].filter(v => oneSel.includes(v)));
        });
        if(newGrp.length>$("#ABBRcomBmax").val()){
          $("#ABBRcomBnote").html("<p style='color:red;'>Max annotation combination reached, either increase the number of reduce selected annotations.</p>");
          return combCell;
        }

      }
      cName = newGrp;
      cVal = newVal;
      //selC = newC;
    }
  }
  for(var i=0;i<cName.length;i++){
    //cName[i] += " ("+selC[i].length+")";
    //var cells = new Int32Array(selC[i].length);
    //cells.set(selC[i]);
    combCell[cVal[i]] = {cName:cName[i]};//,selC:cells
  }
  return combCell;
}
function ABBRcombineNew(){
  window.bioCombine={};
  var combGrp = {};
  $.each($('.ABBRcomBgrps:checkbox:checked'),function(index, anno){
    var oneG=[];
    $.each($('.ABBRcomBgrp_'+$.escapeSelector(anno.value)+':checkbox:checked'),function(ix,sel){
      oneG.push(sel.value);
    });
    if(oneG.length>0){
      combGrp[anno.value]=oneG;
    }
  });
  if(Object.keys(combGrp).length<2){
    $("#ABBRcomBnote").html("Please select annotations from at least two categories.");
    return;
  }
  window.bioCombine=combGrp;
  window.bioCombineCell = ABBRcombineCell();
  if(Object.keys(window.bioCombineCell).length<2){
    return;
  }
  var htmlCheck = "";
  for(const one of Object.values(window.bioCombineCell)){
    htmlCheck += "<li data-val='"+one.cName+"'><label>"+one.cName+"</label></li>";
  }
  $("#ABBRdynList").html(htmlCheck);
  $("#ABBRcombineList").show();

}
function ABBRcombine(){
  window.bioCombine={};
  var combGrp = {};
  var grpKeys = Object.keys(window.store.getState().categoricalSelection);
  for(var i=0;i<grpKeys.length;i++) {
    if(window.store.getState().categoricalSelection[grpKeys[i]].size == 0){
      combGrp[grpKeys[i]] = window.store.getState().annoMatrix.schema.annotations.obsByName[grpKeys[i]].categories;
    }else{
      var one = [];
      for (let [key, value] of window.store.getState().categoricalSelection[grpKeys[i]]) {
        if(value) {
          one.push(key);
        }
      }
      if(one.length>0){
        combGrp[grpKeys[i]] = one;
      }
    }
  }
  window.bioCombine=combGrp;
  window.bioCombineCell = ABBRcombineCell();
  var htmlCheck = "";
  for(const one of Object.values(window.bioCombineCell)){
    htmlCheck += "<li data-val='"+one.cName+"'><label>"+one.cName+"</label></li>";
  }
  $("#ABBRdynList").html(htmlCheck);
  $("#ABBRcombineList").show();
}
// process genes
function ADDGadd(){
  var addG = $("#ADDGinput").val().trim().split(/\s+|\s*,\s*|\s*;\s*/);
  var genes =  window.biogen;
  var notG = [];
  var totalG = window.store.getState().annoMatrix._cache.var.__columns[0];
  var totalGlower = totalG.map(function(s){return s.toLowerCase();});

  for(var i=0;i<addG.length;i++){
    var one = addG[i].trim();
    if (addG[i].trim().length<1)
      continue;
    var ix = totalGlower.indexOf(one.toLowerCase());
    if(ix !== -1){
      var oneG = totalG[ix];
      if(!genes.includes(oneG)){
        genes.push(oneG);
      }
    }else{
      notG.push(one);
    }
  }
  $("#ADDGdel").text(notG.join(", "));
  $("#ADDGinput").val("");
  ADDGref();
}
function ADDGdel(gName){
  const ix = window.biogen.indexOf(gName);
  window.biogen.splice(ix,1);
  $("#ADDG_"+gName).remove();
}
function ADDGsetAdd(idx){//Geneset_Name Geneset_Genes
  $("#ADDGsetError").text("");
  eMsg=AddGsetAddOne($("#Geneset_Name"+idx).val().trim(),$("#Geneset_Genes"+idx).val().trim().split(/\s+|\s*,\s*|\s*;\s*/))
  $("#ADDGsetError").text(eMsg);
  if(!eMsg.startsWith("Error")){
      $("#Geneset_Name"+idx).val("");
      $("#Geneset_Genes"+idx).val("");
  }
}
function AddGsetAddOne(grpName,gNames){
  var grpKeys = Object.keys(window.bioGeneGrp);
  if(grpName.length<1){
    return "Error: Please provide a gene set name with at least one character!";
  }
  if(grpKeys.includes(grpName)){
    return "Error: The gene set name exists!";
  }
  var definedG = [];
  for(var i=0;i<grpKeys.length;i++){
    definedG = definedG.concat(window.bioGeneGrp[grpKeys[i]]);
  }
  var notG = [],geneNames=[];
  var totalG = window.store.getState().annoMatrix._cache.var.__columns[0];
  var totalGlower = totalG.map(function(s){return s.toLowerCase();});
  for(var i=0;i<gNames.length;i++){
    var one = gNames[i].trim();
    var ix = totalGlower.indexOf(one.toLowerCase());
    if(ix !== -1){
      var oneG = totalG[ix];
      if (one.length<1) continue;
      if(!definedG.includes(oneG)){
        geneNames.push(oneG);
      }else{
        notG.push(one);
      }
    }else{
      notG.push(one);
    }
  }
  
  var eMsg= "";
  if(notG.length>0){
    eMsg ="The following genes cannot be found in the data or defined in previous gene sets: "+notG;
  }
  if(geneNames.length==0){
    eMsg = "Error: No valid gene names is provided! " + eMsg;
  }else{
      window.bioGeneGrp[grpName] = geneNames;
      geneNames = geneNames.concat(window.biogen);
      geneNames = [...new Set(geneNames)];
      window.biogen = geneNames;
      ADDGref();
  }
  return eMsg;
}

function ADDGgsetFile(ele){
    Global_F = ele.files;
    const reader = new FileReader();
    reader.onload = (e) => {
        eMsg=""
        e.target.result.split(/\r\n|\n/).forEach(function(line,ix){
            res = line.split(":");
            msg = AddGsetAddOne(res[0].trim(),res[1].trim().split(/\s+|\s*,\s*|\s*;\s*/));
            if(msg.length>2){
                eMsg += "<br/>"+res[0]+": "+msg;
            }
        })
        $("#ADDGsetError").html(eMsg);
    }
    for (let file of ele.files) {
        reader.readAsText(file)
    }
    //Global_reader = reader;
}
function ADDGsetDel(grpName){
  delete window.bioGeneGrp[grpName];
  ADDGref();
}
function hideShow(section, button, label) {
  section.toggle();
  if (label == null) label = 'Input';
  //button.text(button.text()=='Hide '+label?'Show '+label:'Hide '+label);
  if(button.text().includes('Hide')){
    button.text('Show '+label);
  }else if(button.text().includes('Show')){
    button.text('Hide '+label);
  }else if(button.text()==""){
    button.text("+");
  }else if(button.text()=="+"){
    button.text("");
  }
}
// spatial function
function SPATIAL(){
  var state = window.store.getState();
  if (!state.layoutChoice.current.toLowerCase().includes("spat")) {
    $('#SPATIALnote').text("The current layout is "+state.layoutChoice.current+". Please select a spatial embedding with 'spat' in the name.");
    return;
  }
  $('#SPATIALnote').text("");
  if (document.getElementById("spatial-layer")) document.getElementById("spatial-layer").remove();

  var D = {'method':'SPATIAL',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'spots':SPATIALspots(state.layoutChoice.current),
            'embedding':state.layoutChoice.current,
            'resolution':$("input[name='IMGresolution']:checked").val(),
            'figOpt':window.figOpt};
  $(".SPATIALbt").prop('disabled',true);
  document.getElementById("SPATIALspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      $(".SPATIALbt").prop("disabled",false);
      document.getElementById("SPATIALspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#SPATIALnote').text(res);
        return;
      }
      var resD = JSON.parse(res);
      var scale = resD[0][0].scale;

      var element = $('#canvas-transformation-group-x').parent()[0];
      var mtx = document.getElementById("camera-transformation-group").getCTM();
      var trans = 'translate(' +document.getElementById("graph-wrapper").offsetParent.offsetLeft+ ',0) matrix('+mtx.a+','+mtx.b+','+mtx.c+','+mtx.d+','+mtx.e+','+mtx.f+')';
      $("body").append('<svg id="spatial-layer" width="' +element.attributes.width.value+ '" height="' +element.attributes.height.value+ '" pointer-events="none" style="position:relative;opacity:0.75;z-index:0"><g id="spatial-transform-layer" transform="' +trans+ '"><g id="spatial-image-layer" transform="translate(' +resD[0][0].translatex+ ',' +resD[0][0].translatey+ ')"><image transform="matrix(' +scale+ ',0,0,-' +scale+ ',0,0)" href="data:image/png;base64,' +resD[1]+ '"></image></g></g></svg>');
      window.store.dispatch({type: 'change graph interaction mode', data: 'zoom'});
    },
    error: function(request,status,error){
      $('#SPATIALnote').text(request.status+request.responseText);
      $(".SPATIALbt").prop("disabled",false);
      document.getElementById("SPATIALspin").style.visibility="hidden";
    }
  });
}

function SPATIALlist() {
  if ($('#SPATIALname').text().length > 1) return;
  var D = {'method':'SPATIAL',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'embedding': 'get_spatial_list',
            'figOpt':window.figOpt};
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      if(res.startsWith('ERROR')){
        $('#SPATIALnote').text(res);
        return;
      }
      var spatialEmb = JSON.parse(res).list.filter(val => val !== "spatial_Merged");
      var htmlCheck = "";
      for(const g of spatialEmb){
        htmlCheck += '<label id="SPATIAL'+'ck'+g+'"><input class="SPATIALnames" type="checkbox" value="'+g+'" checked/>'+g.replace("spatial_","")+'</label>';
      }
      $('#SPATIALname').append(htmlCheck);
    },
    error: function(request,status,error){
      $('#SPATIALnote').text(request.status+request.responseText);
    }
  });
}

function SPATIALarrange() {
  let grid = GridStack.init({minRow: 1, cellHeight: 200,resizable: {autoHide: true, handles: 'none'},float: true});
  grid.column(6, 'moveScale');
  grid.removeAll();

  $(".SPATIALbt").prop('disabled',true);
  document.getElementById("SPATIALspin").style.visibility="visible";

  var spatialEmb = [];
  $(".SPATIALnames").each(function(){
    if($(this).prop("checked")){
      spatialEmb.push($(this).val());
    }
  });

  var col = Math.ceil(Math.sqrt(spatialEmb.length));
  var row = Math.ceil(spatialEmb.length/col);

  var count = 0;
  for (i=0;i<row;i++) {
    for (j=0;j<col;j++) {
      var D = {'method':'SPATIAL',
                'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
                'spots':SPATIALspots(spatialEmb[count]),
                'embedding':spatialEmb[count],
                'resolution':'lowres',
                'row':i,
                'col':j,
                'figOpt':window.figOpt};
      $.ajax({
        type:"POST",
        url: VIPurl,
        data:JSON.stringify(D),
        contentType: 'application/json;charset=UTF-8',
        success: function(res){
          $(".SPATIALbt").prop("disabled",false);
          document.getElementById("SPATIALspin").style.visibility="hidden";
          if(res.startsWith('ERROR')){
            $('#SPATIALnote').text(res);
            return;
          }
          var resD = JSON.parse(res);
          var input = JSON.parse(this.data);
          var n={};
          n.x=input.col;
          n.y=input.row;
          n.w=1;
          n.h=1;
          n.text=input.embedding.replace("spatial_","");
          n.content = '<button onClick="GridStack.init().removeWidget(this.parentNode.parentNode)">X</button><br><div id="' + n.text + '_content"><div style="word-break:break-all;">' + n.text + '</div><image id="' + n.text +'_img" width="100%" src="data:image/png;base64,' +resD[1]+ '"></image></div>';
          var el = grid.addWidget(n);
          $('#'+n.text + '_content').click (function(){
            var trans =  $('#'+ n.text + '_img')[0].style.transform;
            trans.match(/scaleX/) ? $("#flipx").prop( "checked", true ) : $("#flipx").prop( "checked", false ) ;
            trans.match(/scaleY/) ? $("#flipy").prop( "checked", true ) : $("#flipy").prop( "checked", false ) ;
            if (trans.match(/([0-9]+)deg/)) {
              var rot = RegExp.$1;
              $('#image_rotate_val').text(rot);
              $('#image_rotate').val(rot);
            } else {
              $('#image_rotate_val').text(0);
              $('#image_rotate').val(0);
            }
            $('#selected_image').text(n.text);
          });
        },
        error: function(request,status,error){
          $('#SPATIALnote').text(request.status+request.responseText);
          $(".SPATIALbt").prop("disabled",false);
          document.getElementById("SPATIALspin").style.visibility="hidden";
        }
      });
      count ++;
      if (count >= spatialEmb.length) break;
	}
  }
}

function SPATIALspots(embedding){
  var state = window.store.getState();

  var axis0 = state.annoMatrix.schema.layout.obsByName[embedding].dims[0];
  var axis1 = state.annoMatrix.schema.layout.obsByName[embedding].dims[1];
  var emb = state.annoMatrix._cache.emb;
  var spots = {};
  var x_index = emb.colIndex.index.get(axis0);
  var y_index = emb.colIndex.index.get(axis1);
  spots['spot0_x'] = emb.__columns[x_index][0];
  spots['spot0_y'] = emb.__columns[y_index][0];

  for (var i=1;i<emb.length;i++) {
    if (Math.abs(emb.__columns[y_index][0] - emb.__columns[y_index][i]) > 0.001) {
      spots['spoti_i'] = i;
      spots['spoti_x'] = emb.__columns[x_index][i];
      spots['spoti_y'] = emb.__columns[y_index][i];
      break;
    }
  }
  return spots;
}

function SPATIALoverlay() {
  var mtx = document.getElementById("camera-transformation-group").getCTM();
  var trans = 'translate(' +document.getElementById("graph-wrapper").offsetParent.offsetLeft+ ',0) matrix('+mtx.a+','+mtx.b+','+mtx.c+','+mtx.d+','+mtx.e+','+mtx.f+')';
  $('#spatial-transform-layer').attr('transform',trans);
}

// plot function
function GDplot(){
  var cells = getCells(null,true);
  if(Object.keys(cells).length==0){
    $('#GDnote').text(VIPCellSelectionMsg);
    return;
  }

  if(+$('#GDcutoff').prop('min') > +$("#GDcutoff").val()){
    $('#GDnote').text("The minimal expression level is: "+$('#GDcutoff').prop('min'));
    return;
  }

  var D = {'method':'GD',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'cutoff':$("#GDcutoff").val(),
            'figOpt':window.figOpt};
  $(".GDbt").prop('disabled',true);
  $(".GDpngBT").prop('disabled',true);
  document.getElementById("GDspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      $(".GDbt").prop("disabled",false);
      document.getElementById("GDspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#GDnote').text(res);
        document.getElementById("GDfig").innerHTML = '<img id="GDpng" src="data:image/png;base64," width=100% height="auto"/>';
        return;
      }
      showImg('GDfig',res);
      $(".GDpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#GDnote').text(request.status+request.responseText);
      document.getElementById("GDfig").innerHTML = '<img id="GDpng" src="data:image/png;base64," width=100% height="auto"/>';
      $(".GDbt").prop("disabled",false);
      document.getElementById("GDspin").style.visibility="hidden";
    }
  });

}
function MARKplot(){
  $('#MARKnote').text("");
  var cells=getCells('MARKcell');
  if(cells.length==0){
    $('#MARKnote').text(VIPCellSelectionMsg);
    return;
  }

  if($("#MARKn").val()<3){
    $('#MARKnote').text("Minimal number of markers per group is 3");
    return;
  }
  var grp = [$("#MARKgrp option:selected").val()];
  var combGrp = [],combGrpOrder=[];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
    combGrpOrder = $("#ABBRdynList").sortable('toArray',{ attribute: 'data-val' });
  }

  $(".MARKbt").prop("disabled",true);
  $(".MARKpngBT").prop('disabled',true);
  document.getElementById("MARKspin").style.visibility="visible";
  var D = {'method':'MARK',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':[],
            'grp':grp,
            'markMethod':$("#MARKmethod option:selected").val(),
            'geneN':$("#MARKn").val(),
            'abb':window.bioAbbr,
            'combine':combGrp,
            'combineOrder':combGrpOrder,
            'figOpt':window.figOpt};
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      document.getElementById("MARKspin").style.visibility="hidden";
      $(".MARKbt").prop("disabled",false);
      if(res.startsWith('ERROR')){
        $('#MARKnote').text(res);
        $('#MARKtable').empty();
        document.getElementById("MARKfig").innerHTML = '';
        return;
      }

      resD = JSON.parse(res);
      var tD = resD[0];
      var keyD = tD.shift();
      var titleD = [];
      for(const one of keyD){
        titleD.push({'title':one})
      }
      $('#MARKtable').empty();
      var table = $('#MARKtable').DataTable({
        destroy: true,
        dom:'Blfrtip',
        lengthMenu:[[10,50,100,200,-1],[10,50,100,200,"All"]],
        order:[[1,"dsc"]],
        buttons:['csv'],
        data:tD,
        columns:titleD
      });

      $('#MARKtable tbody').on('click',"tr",function(){
        var data = table.row(this).data();
        //alert('Your clicked on '+data[0]+" row");
        var genes =  window.biogen;
        if(genes.includes(data[0])){
          genes.splice(genes.indexOf(data[0]),1);
        }else{
          genes.push(data[0]);
        }
        $('#MARKnote').text("Selected genes: "+genes.join(", "));
      });
      showImg('MARKfig',resD[1]);
      $(".MARKpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#MARKnote').text(request.status+request.responseText);
      $('#MARKtable').empty();
      document.getElementById("MARKfig").innerHTML = '';
      $(".MARKbt").prop("disabled",false);
      document.getElementById("MARKspin").style.visibility="hidden";
    }
  })

}
function SGVplot(acc){
  $('#SGVnote').text("");
  if($("#SGVgene").val()==null){
    $('#SGVnote').text(VIPGeneSelectionMsg);
    return;
  }
  if($("#SGVgrp option:selected").val()=="NA"){
    $('#SGVnote').text("Please select a category to plot");
    return;
  }
  var cells=getCells('SGVcell');
  if(cells.length==0){
    $('#SGVnote').text(VIPCellSelectionMsg);
    return;
  }
  if(+$("#SGVcutoff").val() < +$("#SGVcutoff").prop('min')){
    $('#SGVnote').text("The minimal expression level is "+$("#SGVcutoff").prop('min'));
    return;
  }
  //$('#SGVnote').text("Plotting on "+cells.length+" selected cells");
  var grp = [$("#SGVgrp").val()];
  var combGrp = [],combGrpOrder=[];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
    combGrpOrder = $("#ABBRdynList").sortable('toArray',{ attribute: 'data-val' });
  }

  var D = {'method':'SGV',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':[$("#SGVgene").val()],
            'grp':grp,
            'abb':window.bioAbbr,
            'cutoff':$("#SGVcutoff").val(),
            'combine':combGrp,
            'combineOrder':combGrpOrder,
            'figOpt':window.figOpt};
  if(acc=='data') D['method']='VIOdata';
  $(".SGVbt").prop("disabled",true);
  $(".SGVsaveBT").prop('disabled',true);
  $(".SGVpngBT").prop('disabled',true);
  document.getElementById("SGVspin").style.visibility="visible";

  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".SGVbt").prop("disabled",false);
      $(".SGVsaveBT").prop('disabled',false);
      document.getElementById("SGVspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#SGVnote').text(res);
        document.getElementById("SGVfig").innerHTML = '';
        return;
      }
      if(acc=='data'){
        var hiddenE = document.createElement('a');
        hiddenE.href="data:attachment/text,"+encodeURIComponent(res);
        hiddenE.target='_blank';
        hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.violinData.csv';
        hiddenE.click();
      }else{
        showImg('SGVfig',res);
      }
      //document.getElementById("SGVfig").innerHTML = '<img id="SGVpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".SGVpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#SGVnote').text(request.status+request.responseText);
      document.getElementById("SGVfig").innerHTML = '<img id="SGVpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".SGVbt").prop("disabled",false);
      document.getElementById("SGVspin").style.visibility="hidden";
    }
  })
}
function SGVcompareplot(){
  $('#SGVcomparenote').text("");
  if($("#SGVcomparegene option:selected").val()=="NA"){
    $('#SGVcomparenote').text(VIPGeneSelectionMsg);
    return;
  }
  if($("#SGVcomparegrp option:selected").val()=="NA"){
    $('#SGVcomparenote').text("Please select a category to plot");
    return;
  }
  var cells=getCells('SGVcomparecell');
  if(cells.length==0){
    $('#SGVcomparenote').text(VIPCellSelectionMsg);
    return;
  }
  if(+$("#SGVcomparecutoff").val() < +$("#SGVcomparecutoff").prop('min')){
    $('#SGVcomparenote').text("The minimal expression level is "+$("#SGVcomparecutoff").prop('min'));
    return;
  }
  //$('#SGVnote').text("Plotting on "+cells.length+" selected cells");
  let grp = [...new Set([$("#SGVcompareorder").val(),$("#SGVcomparegrp").val()])];


  //var combGrp = [],combGrpOrder=[];
  //if(grp[0]=="Custom_combine"){
  //  grp = Object.keys(window.bioCombine);
  //  combGrp = window.bioCombine;
  //  combGrpOrder = $("#ABBRdynList").sortable('toArray',{ attribute: 'data-val' });
  //}

  var D = {'method':'SGVcompare',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':[$("#SGVcomparegene").val()],
            'grp':grp,
            'abb':window.bioAbbr,
            'cutoff':$("#SGVcomparecutoff").val(),
            'cellCutoff':$("#SGVcompareCellcutoff").val(),
            'dotsize':1,
            //'combine':combGrp,
            //'combineOrder':combGrpOrder,
            'figOpt':window.figOpt};
  //if(acc=='data') D['method']='VIOdata';
  $(".SGVcomparebt").prop("disabled",true);
  $(".SGVcomparepngBT").prop('disabled',true);
  document.getElementById("SGVcomparespin").style.visibility="visible";

  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".SGVcomparebt").prop("disabled",false);
      document.getElementById("SGVcomparespin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#SGVcomparenote').text(res);
        showImg('SGVcomparefig',"");
        return;
      }
      showImg('SGVcomparefig',res);
      $(".SGVcomparepngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#SGVcomparenote').text(request.status+request.responseText);
      showImg('SGVcomparefig',"");
      $(".SGVcomparebt").prop("disabled",false);
      document.getElementById("SGVcomparespin").style.visibility="hidden";
    }
  })
}
function getGeneSet(aID){
  var grpKeys=[];
  $("."+aID+"sets").each(function(){
    if($(this).prop("checked")){
      grpKeys.push($(this).val());
    }
  });
  var selSets={};
  for(const one of grpKeys){
    selSets[one]=window.bioGeneGrp[one];
  }
  return(selSets)
}
function getGenes(aID){
  var genes = [];
  $("."+aID+"genes").each(function(){
    if($(this).prop("checked")){
      genes.push($(this).val());
    }
  });
  return(genes);
}
function PGVplot(acc){
  $('#PGVnote').text("");
  var genes =getGenes('PGV');
  var geneSets = getGeneSet('PGV');

  if(genes.length==0&&Object.keys(geneSets).length==0){
    $('#PGVnote').text("Please select at least a gene or a gene set to plot");
    return;
  }

  if($("#PGVgrp option:selected").val()=="NA"){
    $('#PGVnote').text("Please select a category to plot");
    return;
  }
  var cells=getCells('PGVcell');
  if(cells.length==0){
    $('#PGVnote').text(VIPCellSelectionMsg);
    return;
  }
  if(+$("#PGVcutoff").val()< +$("#PGVcutoff").prop('min')){
    $('#PGVnote').text("The minimal expression level is "+$("#PGVcutoff").prop('min'));
    return;
  }

  var grp = [$("#PGVgrp").val()];
  var combGrp = [],combGrpOrder=[];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
    combGrpOrder = $("#ABBRdynList").sortable('toArray',{ attribute: 'data-val' });
  }
  var geneGrpColl = $("input[name='PGVgeneGrpColl']:checked").val()
  if( geneGrpColl== 'Yes'){
    geneGrpColl = $("input[name='PGVgeneGrpCollCal']:checked").val()
  }
  var color = $("#PGVcolor").val();
  if ($("#PGVcolor_r")[0].checked) color += '_r';
  var D = {'method':'PGV',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'geneGrp':geneSets,
            'geneGrpColl':geneGrpColl,
            'grp':grp,
            'color':color,
            'by':$("#PGVgrpXY").val(),
            'abb':window.bioAbbr,
            'cutoff':$("#PGVcutoff").val(),
            'combine':combGrp,
            'combineOrder':combGrpOrder,
            'figOpt':window.figOpt};
  if(acc=='data') D['method']='VIOdata';
  $(".PGVbt").prop('disabled',true);
  $(".PGVsaveBT").prop('disabled',true);
  $(".PGVpngBT").prop('disabled',true);
  document.getElementById("PGVspin").style.visibility="visible";

  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".PGVbt").prop("disabled",false);
      $(".PGVsaveBT").prop("disabled",false);
      document.getElementById("PGVspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#PGVnote').text(res);
        document.getElementById("PGVfig").innerHTML = '';
        return;
      }
      if(acc=='data'){
        var hiddenE = document.createElement('a');
        hiddenE.href="data:attachment/text,"+encodeURIComponent(res);
        hiddenE.target='_blank';
        hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.violinData.csv';
        hiddenE.click();
      }else{
        showImg('PGVfig',res);
      }
      //document.getElementById("PGVfig").innerHTML = '<img id="PGVpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".PGVpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#PGVnote').text(request.status+request.responseText);
      document.getElementById("PGVfig").innerHTML = '<img id="PGVpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".PGVbt").prop("disabled",false);
      document.getElementById("PGVspin").style.visibility="hidden";

    }
  })
}
function PGVcomparebt(){
  $('#PGVcomparenote').text("");
  var genes =getGenes('PGVcompare');
  if(genes.length<2){
    $('#PGVcomparenote').text("Please add/select at least two genes to plot");
    return;
  }
  var cells=getCells('PGVcomparecell');
  if(cells.length==0){
    $('#PGVcomparenote').text(VIPCellSelectionMsg);
    return;
  }
  var grp = [...new Set([$("#PGVcompareRow").val(),$("#PGVcompareGrp").val()])];
  if(grp.length<2){
    $('#PGVcomparenote').text("Please selected two different group annotations");
    return;
  }
  var D = {'method':'PGVcompare',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'cutoff':$("#PGVcomparecutoff").val(),
            'abb':window.bioAbbr,
            'width':$("#PGVcomparewidth").val(),
            'height':$("#PGVcompareheight").val(),
            'figOpt':window.figOpt};

  $(".PGVcomparebt").prop('disabled',true);
  $(".PGVcomparepngBT").prop('disabled',true);
  document.getElementById("PGVcomparespin").style.visibility="visible";

  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".PGVcomparebt").prop("disabled",false);
      document.getElementById("PGVcomparespin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#PGVcomparenote').text(res);
        document.getElementById("PGVcomparefig").innerHTML = '';
        return;
      }
      //$('#PGVcomparenote').text("Success!");
      showImg('PGVcomparefig',res);
      $(".PGVcomparepngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#PGVcomparenote').text(request.status+request.responseText);
      document.getElementById("PGVcomparefig").innerHTML = '<img id="PGVcomparepng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".PGVcomparebt").prop("disabled",false);
      document.getElementById("PGVcomparespin").style.visibility="hidden";
    }
  })

}
function violinData(){

}
function HEATgeneFileUpdate(){
  //get genes annotation from file
  const fileSelector = document.getElementById('HEATgeneFile');
  if(fileSelector.files.length>0){
    let fr = new FileReader();
    fr.readAsText(fileSelector.files[0]);
    fr.onload=function(){
      var geneAnno = d3.csvParse(fr.result);
      var totalG = window.store.getState().annoMatrix._cache.var.__columns[0];
      var totalGlower = totalG.map(function(s){return s.toLowerCase();});
      //$("#ADDGinput").val(geneAnno.map(function(d){return(d.gene)}).join(','));
      //ADDGadd();
      //sync();
      var iG = geneAnno.map(function(d){
        var ix = totalGlower.indexOf(d.gene.trim().toLowerCase());
        if(ix !== -1){
          d.gene = totalG[ix];
        }else{
          d=null;
        }
        return d;
      });
      window.HEATgene=iG.filter(n=>n);
    }
    fr.onerror=function(){
      alert("Error in read user CSV files!");
    }
    $("#HEATgeneAnnoCK").prop('checked', false);
    $('#HEATgeneAnno').hide();
    if($("#HEATgenesAll").is(':checked')){
      $('#HEATgenesAll').trigger('click');
    }
    $('#HEATgenesAll').prop('checked',false);
  }else{
    $("#HEATgeneAnnoCK").prop('checked', true);
    $('#HEATgeneAnno').show();
    if(!$("#HEATgenesAll").is(':checked')){
      $('#HEATgenesAll').trigger('click');
    }
  }
}
function HEATcheck(){
  $('#HEATnote').text("");
  var cells=getCells('HEATcell');
  if(cells.length==0){
    $('#HEATnote').text(VIPCellSelectionMsg);
    return;
  }
  //get genes annotation from file
  const fileSelector = document.getElementById('HEATgeneFile');
  var genes=[],gAnno=[],gAnnoDef=$("#HEATgeneAnnoCK").is(':checked');
  if(fileSelector.files.length>0){
    genes = window.HEATgene.map(function(d){return(d.gene)});
    gAnno = window.HEATgene;
    gAnnoDef = false;
    $('#HEATnote').text("Genes obtained from gene annotation file!");
  }else{
    genes = getGenes('HEAT');
  }
  if(genes.length<2){
    $('#HEATnote').text("It needs at least two genes");
    return;
  }

  var grps = $(".HEATgrps:checked").map(function(){return this.value;}).get();
  //x = document.getElementsByClassName('HEATgrps');
  //for(i=0;i<x.length;i++){
  //  if(x[i].checked){
  //    grps.push(x[i].value);
  //  }
  //}
  var numGrp = $(".HEATmetas:checked").map(function(){return this.value;}).get();

  var orderGrp=[];
  if($("#HEATplotMethod").val()=="cHeatmap"){
    if($("#HEATcomplexSortby").val()=="Annotation"){
      orderGrp = $("#HEATdynList").sortable('toArray');
    }else{
      orderGrp=["Expression"];
    }
  }else{
    orderGrp = $("#HEATorder").val();
    if(orderGrp=="Expression" && cells.length>heatCell){
      $('#HEATnote').text("Cluster by expression is not available for the number of selected cells larger than "+Math.round(heatCell/1000)+"k.");
      return;
    }
    if(orderGrp!="Expression" && !grps.includes(orderGrp)){
      grps.push(orderGrp);
    }
    orderGrp = [orderGrp];
  }

  var grp=[],combGrp = [],combGrpOrder=[];
  if(orderGrp.includes("Custom_combine") || grps.includes("Custom_combine")){
    grp = grps.filter(function(e) { return e !== 'Custom_combine' });
    grps = [...new Set(grp.concat(Object.keys(window.bioCombine)))];
    combGrp = window.bioCombine;
    combGrpOrder = $("#ABBRdynList").sortable('toArray',{ attribute: 'data-val' });
  }

  if(grps.length==0){
    $('#HEATnote').text("Please select annotations to plot");
    return;
  }
  document.getElementById("HEATspin").style.visibility="visible";
  $(".HEATbt").prop('disabled',true);
  $(".HEATpngBT").prop('disabled',true);
  $(".HEATsaveBT").prop('disabled',true);
  var color = $("#HEATcolor").val();
  if ($("#HEATcolor_r")[0].checked) color += '_r';
  if($("#HEATplotMethod").val()=="cHeatmap" && orderGrp=="Expression"){
    $('#HEATnote').text("The cells with sd=0 and value<1 across all selected genes will be removed");
  }
  var D = {'method':'HEAT',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'gAnno':gAnno,
            'gAnnoDef':gAnnoDef,
            'order':orderGrp,
            'norm':$("#HEATnorm").val(),
            'grp':grps,
            'grpNum':numGrp,
            'color':color,
            'plotMethod':$("#HEATplotMethod").val(),
            'width':$("#HEATwidth").val(),
            'height':$("#HEATheight").val(),
            //'columnFormat':$("#HEATcolumnFormat").val().replaceAll("'","\\'"),
            //'rowFormat':$("#HEATrowFormat").val().replaceAll("'","\\'"),
            //'annoFormat':$("#HEATannoFormat").val().replaceAll("'","\\'"),
            'swapAxes':$("input[name='HEATswapAxes']:checked").val(),
            'legendRow':$("#HEATcomplexSettingLegendRow").val(),
            'fontadj':$("#HEATcomplexSettingFontadj").val(),
            'abb':window.bioAbbr,
            'combine':combGrp,
            'combineOrder':combGrpOrder,
            'addGrp':grp,
            'figOpt':window.figOpt};
  return D;
}
function HEATplot(){
  D = HEATcheck();
  if(D==undefined) return;
  D['method'] = 'HEATplot';
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".HEATbt").prop("disabled",false);
      document.getElementById("HEATspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#HEATnote').text(res);
        document.getElementById("HEATfig").innerHTML = '';
        return;
      }

      showImg('HEATfig',res);//resD[0]
      //document.getElementById("HEATfig").innerHTML = '<img id="HEATpng" src="data:image/png;base64,'+resD[0]+'" width=100% height="auto"/>';// height="450" width="450"
      $(".HEATpngBT").prop('disabled',false);
      $(".HEATsaveBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#HEATnote').text(request.status+request.responseText);
      document.getElementById("HEATfig").innerHTML = '<img id="HEATpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".HEATbt").prop("disabled",false);
      document.getElementById("HEATspin").style.visibility="hidden";

    }
  });
}
function HEATdata(){
  D = HEATcheck();
  if(D==undefined) return;
  D['method'] = 'HEATdata';
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".HEATbt").prop("disabled",false);
      document.getElementById("HEATspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#HEATnote').text(res);
        return;
      }

      var hiddenE = document.createElement('a');
      hiddenE.href="data:attachment/text,"+encodeURIComponent(res);
      hiddenE.target='_blank';
      hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.heatmapData.csv';
      hiddenE.click();

      $(".HEATpngBT").prop('disabled',false);
      $(".HEATsaveBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#HEATnote').text(request.status+request.responseText);
      document.getElementById("HEATfig").innerHTML = '<img id="HEATpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".HEATbt").prop("disabled",false);
      document.getElementById("HEATspin").style.visibility="hidden";

    }
  });
}
function DEGselCells(){
  var selAnno = {};
  var selNames = [];
  //var selC = {};
  var grpKey = Object.keys(window.store.getState().categoricalSelection);
  $(".DEGgrps").each(function(){
    if($(this).prop("checked")){
      selNames.push(window.bioCombineCell[$(this).val()].cName);
      //selC[window.bioCombineCell[$(this).val()].cName.split(' ')[0]] = window.bioCombineCell[$(this).val()].selC;
      //window.store.getState().annoMatrix.schema.annotations.obsByName[key].categories.indexOf(v)
      $(this).val().split("__").forEach(function(one){
        [k,v]=one.split("_");
        if(selAnno[grpKey[k]]==undefined){selAnno[grpKey[k]]=[]}
        if(!selAnno[grpKey[k]].includes(window.store.getState().annoMatrix.schema.annotations.obsByName[grpKey[k]].categories[v])){
          selAnno[grpKey[k]].push(window.store.getState().annoMatrix.schema.annotations.obsByName[grpKey[k]].categories[v]);
        }
      });
    }
  })
  return [selAnno,selNames];
  //return selC;
}
function operator(a,b,op){
  switch(op){
    case ">": return a>b;
    case "=": return a==b;
    case "<": return a<b;
  }
}
function getSelOpt(eID){
  var selOpt=[];
  if($("input[name='"+eID+"colorBy']:checked").val()=='significance'){
    selOpt.push({column:'absLogfc',opr:'>',val:+$("#"+eID+"logfc").val()});
    selOpt.push({column:'qval',opr:'<',val:+$("#"+eID+"fdr").val()});
  }else if($("input[name='"+eID+"colorBy']:checked").val()=='geneInfo'){
    selOpt.push({column:$("#"+eID+"plotByGinfoSel").val(),
                  opr:$("input[name='"+eID+"plotByGinfoRd']:checked").val(),
                  val:$("#"+eID+"plotByGinfoVal").val()});
  }else if($("input[name='"+eID+"colorBy']:checked").val()=='selGene'){
    selOpt.push({column:'gene',opr:'include',val:getGenes('DEGsel')});
  }else{
    $('#'+eID+'note').text("Unknown selection!");
  }
  return selOpt;
}
function updateDEGdata(rawD,selOpts){
  return d3.csvParse(rawD,function(d){
    for (var key in d) {
     d[key] = isNaN(+d[key]) ? d[key]:(+d[key]);
     d['absLogfc'] = Math.abs(d.log2fc)
     //for color-----------
     selCol = selOpts['color'];
     d['color'] = "unselected";
     if(selCol[0].opr=="include"){
       if(selCol[0].val.includes(d[selCol[0].column])) d['color'] = "selected";
     }else if(isNaN(+selCol[0].val)){
       d['color'] = d[selCol[0].column]
     }else{
       var flag=true;
       selCol.forEach(function(ref){
         flag = flag && operator(d[ref.column],ref.val,ref.opr);
       })
       if(flag) d['color'] = "selected";
     }
     //for label -----------
    selLab = selOpts['label'];
     d['label'] = true;
     if(selLab[0].opr=="include"){
       if(!selLab[0].val.includes(d[selLab[0].column])) d['label'] = false;
     }else if(isNaN(+selLab[0].val)){
       if(d['label'] != d[selLab[0].column]) d['label'] = false;
     }else{
       selLab.forEach(function(ref){
         d['label'] = d['label'] && operator(d[ref.column],ref.val,ref.opr);
       })
     }
    }
    var selD = {};
    ['gene','color','log2fc','qval','label'].forEach(function(i){
      selD[i] = d[i];
    })
    return selD;
  });
}
function DEGplot(eID){//window.DEGraw,window.DEGsel 'DEG'
  window[eID+'sel'][$("#"+eID+"plotBy").val()] = getSelOpt(eID);
  var DEGdata=updateDEGdata(window[eID+'raw'],window[eID+'sel']).slice(0,5000);//maximun 5000 dots for interactive volcano plot
  //console.log("interactive colcano dots number:" + DEGdata.length);
  $("#"+eID+"interactive").html("");
  var vol = volcanoPlot()
      .width($("#"+eID+"interactive").width())
      .height($("#"+eID+"interactive").height())
      .xAxisLabel('log<tspan baseline-shift="sub">2</tspan>Fold-change')
      .yAxisLabel('-log<tspan baseline-shift="sub">10</tspan>False Discovery Rate')
      .vLines({a:-$("#"+eID+"logfc").val(),b:+$("#"+eID+"logfc").val()})
      .hLines({['FDR='+$("#"+eID+"fdr").val()]:+$("#"+eID+"fdr").val()})
      .sampleID("gene")
      .xColumn("log2fc")
      .yColumn("qval")
      .colorColumn('color')
      .labelColumn('label')
      .maxY(300)
      .tooltips(['gene','log2fc','qval'])
      .resetBtnID(eID+"figReset");

  d3.select('#'+eID+'interactive')
    .data([DEGdata.filter(function(d){return Math.abs(d.log2fc)< +$("#"+eID+"logfcCut").val();})])
    .call(vol);

}
function DEGfind(){
  $('#DEGnote').text("");
  var genes = getGenes('DEG');
  var D = {'method':'DEG',
           'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
           'topN':$("#DEGtop").val(),
           'DEmethod':$("#DEGmethod").val(),
           'genes':genes,
           'logFC':$("#DEGlogfcCut").val(),
           'sigFDR':$("#DEGsigFDR").val(),
           'sigFC':$("#DEGsigLogfc").val(),
           'labelSize':$("#DEGlabelSize").val(),
           'dotSize':$("#DEGdotSize").val(),
           'ymin':$("#DEGymin").val(),
           'ymax':$("#DEGymax").val(),
           'gsea':{'enable':$('#GSEAenable').prop('checked'),
                   'gs':$('#GSEAgs').val(),
                   'gsMin':$('#GSEAminSize').val(),
                   'gsMax':$('#GSEAmaxSize').val(),
                   'padj':$('#GSEApadjCutoff').val(),
                   'up':$('#GSEAupN').val(),
                   'dn':$('#GSEAdnN').val(),
                   'collapse':$('#GSEAcollapse').prop('checked')},
           'figOpt':window.figOpt}
  const [combGrp,combName] = DEGselCells();
  var cells = {};
  if(combName.length!=2){
    var selCells = getCells(null,true);
    if(Object.keys(selCells).length!=2){
      $('#DEGnote').text("Please select cells into both selection 1 and selection 2");
      return;
    }else{
      cells['group1'] = selCells.celllist1;
      cells['group2'] = selCells.celllist2;
    }
    D['cells'] = cells;
    D['cellN'] = window.store.getState().annoMatrix.nObs;
    D['comGrp'] = Object.keys(cells);
  }else{
    D['combine']=combGrp;
    D['abb']=window.bioAbbr;
    D['grp'] = Object.keys(combGrp);
    D['cells'] = new Int32Array(Array(window.store.getState().annoMatrix.nObs).keys());
    D['comGrp'] = combName.sort();
  }
  $('#DEGinfo').text("DEG: "+D['comGrp'].join(" vs "));

  document.getElementById("DEGspin").style.visibility="visible";
  $(".DEGbt").prop('disabled',true);
  $(".DEGpngBT").prop('disabled',true);
  $('#DEGtable').empty();
  $('#GSEAtable').empty();
  //$('#GSPtable').empty();
  showImg('DEGfig',"");
  showImg('GSEAfig',"");
  $("#DEGinteractive").html("");
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      document.getElementById("DEGspin").style.visibility="hidden";
      $(".DEGbt").prop("disabled",false);
      $(".DEGpngBT").prop('disabled',false);

      if(res.startsWith('ERROR')){
        $('#DEGnote').text(res);
        return;
      }
      var resD = JSON.parse(res);
      //showImg('DEGfig',resD[1]);
      // the default color and label
      var selOpt=[];
      selOpt.push({column:'absLogfc',opr:'>',val:+$("#DEGlogfc").val()});
      selOpt.push({column:'qval',opr:'<',val:+$("#DEGfdr").val()});
      window.DEGsel = {};
      $("#DEGplotBy option").each(function(){
        window.DEGsel[$(this).val()]=selOpt;
      });
      // gene information
      var gInfo = resD[0].split("\n")[0].split(',').slice(4);
      gInfo.forEach(function(d){
        $('#DEGplotByGinfoSel').append('<option value="'+d+'">'+d+'</option>')
      });
      // raw DEG table
      window.DEGraw = resD[0]; //needs for the following plot!
      DEGplot('DEG');
      $("#DEGfinal").show();
      $("#DEGinteractiveBtn").prop('disabled', false);

      // process DEG table
      var tableD = d3.csvParse(resD[0],function(data){
        return [].concat([data.gene,
                  (+data.log2fc).toFixed(2),
                  (+data.pval).toExponential(2),
                  (+data.qval).toExponential(2)],
                  Object.values(data).slice(4));
        });
      var tableTitle = [];
      [].concat(['Gene',"Log Fold Change","p-value","padj"],gInfo).forEach(function(d){tableTitle = tableTitle.concat({title:d})});
      var table = $('#DEGtable').DataTable({
        destroy: true,
        dom:'Blfrtip',
        lengthMenu:[[10,50,100,200],[10,50,100,200]],
        order:[3,"asc"],
        buttons:['csv'],
        data:tableD,
        columns:tableTitle
      });
      $('#DEGtable tbody').on('click',"tr",function(){
        var data = table.row(this).data();
        //alert('Your clicked on '+data[0]+" row");
        var genes =  window.biogen;
        if(genes.includes(data[0])){
          genes.splice(genes.indexOf(data[0]),1);
        }else{
          genes.push(data[0]);
        }
        $('#DEGnote').text("Selected genes: "+genes.join(", "));
      });
      showImg('DEGfig',resD[1]);

      // process GSEA
      if(resD[2].length>10){
        showImg('GSEAfig',resD[3]);
        var gseaTable = d3.csvParse(resD[2],function(data){
          return [].concat([data.pathway,
                    data.pval.length===0? 1:(+data.pval).toExponential(2),
                    data.padj.length===0? 1:(+data.padj).toExponential(2),
                    data.ES.length===0? 'NaN':(+data.ES).toFixed(2),
                    data.NES.length===0? 'NaN':(+data.NES).toFixed(2)],
                    Object.values(data).slice(5));
        });
        var gseaTitle = [];
        [].concat(resD[2].substring(0,500).
          split("\n")[0].split(',').
          forEach(function(d){gseaTitle=gseaTitle.concat({title:d.replace('pval','p value').replace('padj','adjusted p value')})}));
        var gTable = $('#GSEAtable').DataTable({
          destroy: true,
          dom:'Blfrtip',
          lengthMenu:[[10,50,100,200],[10,50,100,200]],
          order:[[1,"asc"]],
          buttons:['csv'],
          data:gseaTable,
          columns:gseaTitle
        });
      }
    },
    error: function(request,status,error){
      $('#DEGnote').text(request.status+request.responseText);
      document.getElementById("DEGspin").style.visibility="hidden";
      $(".DEGbt").prop("disabled",false);
    }
  })
}
function preDEGVolcanofind(eID){
  //var genes = getGenes(eID);
  var D = {'method':eID,
           'topN':$("#"+eID+"top").val(),
           'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
           'compSel':$("#"+eID+"Tab").val(),
           'genes':getGenes(eID),
           'logFC':10,
           'sigFC':$("#"+eID+"Logfc").val(),
           'sigFDR':$("#"+eID+"FDR").val(),
           'labelSize':$("#"+eID+"labelSize").val(),
           'dotSize':$("#"+eID+"dotSize").val(),
           'ymin':$("#"+eID+"ymin").val(),
           'ymax':$("#"+eID+"ymax").val(),
           'figOpt':window.figOpt};
  document.getElementById(eID+"spin").style.visibility="visible";
  $("."+eID+"bt").prop('disabled',true);
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      document.getElementById(eID+"spin").style.visibility="hidden";
      $("."+eID+"bt").prop("disabled",false);
      $('#'+eID+'table').empty();

      if(res.startsWith('ERROR')){
        $('#'+eID+'note').text(res);
        return;
      }
      resD = JSON.parse(res);
      var selOpt=[];
      selOpt.push({column:'absLogfc',opr:'>',val:+$("#"+eID+"logfc").val()});
      selOpt.push({column:'qval',opr:'<',val:+$("#"+eID+"fdr").val()});
      window[eID+"sel"] = {};
      $("#"+eID+"plotBy option").each(function(){
        window[eID+"sel"][$(this).val()]=selOpt;
      });
      // gene information
      var gInfo = resD[0].split("\n")[0].split(',').slice(5);
      gInfo.forEach(function(d){
        $('#'+eID+'plotByGinfoSel').append('<option value="'+d+'">'+d+'</option>')
      });
      // raw DEG table
      window[eID+"raw"] = resD[0];
      DEGplot(eID);
      $("#"+eID+"final").show();
      $("#"+eID+"interactiveBtn").prop('disabled', false);

      // process DEG table
      var tableD = d3.csvParse(resD[0],function(data){
        return [].concat([data.gene,
                          (+data.log2fc).toFixed(2),
                          (+data.pval).toExponential(2),
                          (+data.qval).toExponential(2)],
                         Object.values(data).slice(5));
      });
      var tableTitle = [];
      [].concat(['Gene',"Log Fold Change","p-value","padj"],gInfo).forEach(function(d){tableTitle = tableTitle.concat({title:d})});
      var table = $('#'+eID+'table').DataTable({
        destroy: true,
        dom:'Blfrtip',
        lengthMenu:[[10,50,100,200,-1],[10,50,100,200,"All"]],
        order:[[3,"asc"]],
        buttons:['csv'],
        data:tableD,
        columns:tableTitle
      });
      showImg(eID+'fig',resD[1]);

      $('#'+eID+'table tbody').on('click',"tr",function(){
        var data = table.row(this).data();
        //alert('Your clicked on '+data[0]+" row");
        var genes =  window.biogen;
        if(genes.includes(data[0])){
          genes.splice(genes.indexOf(data[0]),1);
        }else{
          genes.push(data[0]);
        }
        $('#'+eID+'note').text("Selected genes: "+genes.join(", "));
      });
    },
    error: function(request,status,error){
      $('#'+eID+'note').text(request.status+request.responseText);
      $('#'+eID+'table').empty();
      document.getElementById(eID+"spin").style.visibility="hidden";
      $("."+eID+"bt").prop("disabled",false);
    }
  });
}
function preDEGmultiplot(eID){
  $('#'+eID+'note').text("");
  var genes = getGenes(eID);
  if(genes.length<1){
    $('#'+eID+'note').text('Please select genes to be ploted!');
    return;
  }
  var selComp = $("#"+eID+"dynList").sortable('toArray');
  selComp=selComp.map(function(one){return one.replace(eID+"li","");});
  if(selComp.length<1){
    $('#'+eID+'note').text('Please select the comparisons to be ploted!');
    return;
  }
  var D = {'method':eID,
           'genes':genes,
           'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
           'compSel':selComp,
           'scale':$("#"+eID+"FigureScale").val(),
           'figOpt':window.figOpt}
  document.getElementById(eID+"spin").style.visibility="visible";
  $("."+eID+"bt").prop('disabled',true);
  showImg(eID+'fig',"iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mNkYAAAAAYAAjCB0C8AAAAASUVORK5CYII=");
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      //window.testD = res;
      document.getElementById(eID+"spin").style.visibility="hidden";
      $("."+eID+"bt").prop("disabled",false);
      $('#'+eID+'table').empty();

      if(res.startsWith('ERROR')){
        $('#'+eID+'note').text(res);
        return;
      }

      showImg(eID+'fig',res);
    },
    error: function(request,status,error){
      $('#'+eID+'note').text(request.status+request.responseText);
      document.getElementById(eID+"spin").style.visibility="hidden";
      $("."+eID+"bt").prop("disabled",false);
    }
  });

}
function preDEGfgseaFind(){
  var D = {'method':'preDEGfgsea',
           'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
           'compSel':$("#preDEGfgseaTab").val(),
           'gs':$('#preDEGfgseaGS').val(),
           'gsMin':$('#preDEGfgseaMinSize').val(),
           'gsMax':$('#preDEGfgseaMaxSize').val(),
           'padj':$('#preDEGfgseaPadjCutoff').val(),
           'up':$('#preDEGfgseaUpN').val(),
           'dn':$('#preDEGfgseaDnN').val(),
           'collapse':$('#preDEGfgseaCollapse').prop('checked'),
            'figOpt':window.figOpt
  };
  document.getElementById("preDEGfgseaspin").style.visibility="visible";
  $(".preDEGfgseabt").prop('disabled',true);
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      document.getElementById("preDEGfgseaspin").style.visibility="hidden";
      $(".preDEGfgseabt").prop("disabled",false);
      $('#preDEGfgseaTable').empty();

      if(res.startsWith('ERROR')){
        $('#preDEGfgseanote').text(res);
        return;
      }
      resD = JSON.parse(res);
      // process GSEA
      showImg('preDEGfgseafig',resD[1]);
      var gseaTable = d3.csvParse(resD[0],function(data){
        return [].concat([data.pathway,
                  data.pval.length===0? 1:(+data.pval).toExponential(2),
                  data.padj.length===0? 1:(+data.padj).toExponential(2),
                  data.ES.length===0? 'NaN':(+data.ES).toFixed(2),
                  data.NES.length===0? 'NaN':(+data.NES).toFixed(2)],
                  Object.values(data).slice(5));
      });
      var gseaTitle = [];
      [].concat(resD[0].substring(0,500).
        split("\n")[0].split(',').
        forEach(function(d){gseaTitle=gseaTitle.concat({title:d.replace('pval','p value').replace('padj','adjusted p value')})}));
      var gTable = $('#preDEGfgseaTable').DataTable({
        destroy: true,
        dom:'Blfrtip',
        lengthMenu:[[10,50,100,200],[10,50,100,200]],
        order:[[1,"asc"]],
        buttons:['csv'],
        data:gseaTable,
        columns:gseaTitle
      });

    },
    error: function(request,status,error){
      $('#preDEGfgseanote').text(request.status+request.responseText);
      document.getElementById("preDEGfgseaspin").style.visibility="hidden";
      $(".preDEGfgseabt").prop("disabled",false);
    }
  })
}
function GSPplot(){
  $('#GSPnote').text("");
  var genes =getGenes('GSP');
  if(genes.length==0){
    $('#GSPnote').text("All genes will be used, please be patient!");
  }
  var cells=getCells('GSPcell');
  if(cells.length==0){
    $('#GSPnote').text(VIPCellSelectionMsg);
    return;
  }
  var grp = [$("#GSPgrp").val()];
  var D={'method':'GSP',
          'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
          'cells':cells,
          'genes':genes,
          'grp':grp,
          'abb':window.bioAbbr,
          'figOpt':window.figOpt};
  $(".GSPbt").prop('disabled',true);
  $("#GSPspin").show();
  $("#GSPtableDIV").empty();
  $("#GSPtableDIV").append('<table id="GSPtable" class="display" width="100%"></table>');
  $("#GSPfig").empty();

  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".GSPbt").prop("disabled",false);
      $("#GSPspin").hide();
      if(res.startsWith('ERROR')){
        $('#GSPnote').text(res);
        document.getElementById("GSPfig").innerHTML = '';
        return;
      }
      var resD = JSON.parse(res);
      showImg('GSPfig',resD[1])
      // process table
      var tableD = d3.csvParse(resD[0],function(data){
        return Object.values(data);
        });
      var tableTitle = [];
      resD[0].split("\n")[0].split(',').forEach(function(d){tableTitle = tableTitle.concat({title:d})});
      //create download file name
      var dt= new Date($.now());
      const offset = dt.getTimezoneOffset();
      var dt1 = new Date(dt.getTime() - (offset*60*1000));
      var filename="VIP_geneSpecificity_"+dt1.toISOString().replaceAll("T","_").replaceAll(":","");

      var table = $('#GSPtable').DataTable({
        destroy: true,
        dom:'Blfrtip',
        lengthMenu:[[10,50,100,200],[10,50,100,200]],
        //order:[3,"asc"],
        buttons:[{
          extend: 'csv',
          filename:filename,
        }],
        data:tableD,
        columns:tableTitle
      });
      $(".GSPpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#GSPnote').text(request.status+request.responseText);
      document.getElementById("GSPfig").innerHTML = '<img id="GSPpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".GSPbt").prop("disabled",false);
      $("#GSPspin").hide();
      //document.getElementById("GSPspin").style.visibility="hidden";
    }
  })
}

function DOTplot(){
  $('#DOTnote').text("");
  var genes =getGenes('DOT');
  var geneSets = getGeneSet('DOT');
  if(genes.length==0&&Object.keys(geneSets).length==0){
    $('#DOTnote').text("Please select at least a gene or a gene set to plot");
    return;
  }

  var cells=getCells('DOTcell');
  if(cells.length==0){
    $('#DOTnote').text(VIPCellSelectionMsg);
    return;
  }

  if(+$('#DOTcutoff').prop('min') > +$("#DOTcutoff").val()){
    $('#DOTnote').text("The minimal expression level is: "+$('#DOTcutoff').prop('min'));
    return;
  }

  var grp = [$("#DOTgrp").val()];
  var combGrp = [],combGrpOrder=[];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
    combGrpOrder = $("#ABBRdynList").sortable('toArray',{ attribute: 'data-val' });
  }
  var geneGrpColl = $("input[name='DOTgeneGrpColl']:checked").val()
  if( geneGrpColl== 'Yes'){
    geneGrpColl = $("input[name='DOTgeneGrpCollCal']:checked").val()
  }

  var color = $("#DOTcolor").val();
  if ($("#DOTcolor_r")[0].checked) color += '_r';

  var D = {'method':'DOT',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'abb':window.bioAbbr,
            'combine':combGrp,
            'combineOrder':combGrpOrder,
            'geneGrp':geneSets,
            'geneGrpColl':geneGrpColl,
            'cutoff':$("#DOTcutoff").val(),
            'mean_only_expressed':$("input[name='mean_only_expressed']:checked").val(),
            'legendW':$('#DOTlegendW').val(),
            'color':color,
            'figOpt':window.figOpt};
  $(".DOTbt").prop('disabled',true);
  document.getElementById("DOTspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".DOTbt").prop("disabled",false);
      document.getElementById("DOTspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#DOTnote').text(res);
        document.getElementById("DOTfig").innerHTML = '';
        return;
      }

      showImg('DOTfig',res)
      //document.getElementById("DOTfig").innerHTML = '<img id="DOTpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".DOTpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#DOTnote').text(request.status+request.responseText);
      document.getElementById("DOTfig").innerHTML = '<img id="DOTpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".DOTbt").prop("disabled",false);
      document.getElementById("DOTspin").style.visibility="hidden";
    }
  })
}
function TRAKplot(){
  $('#TRAKnote').text("");
  var genes =getGenes('TRAK');
  var geneSets = getGeneSet('TRAK');
  if(genes.length==0&&Object.keys(geneSets).length==0){
    $('#TRAKnote').text("Please select at least a gene or a gene set to plot");
    return;
  }

  var cells=getCells('TRAKcell');
  if(cells.length==0){
    $('#TRAKnote').text(VIPCellSelectionMsg);
    return;
  }

  var grp = [$("#TRAKgrp").val()];
  var combGrp = [],combGrpOrder=[];
  if(grp[0]=="Custom_combine"){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
    combGrpOrder = $("#ABBRdynList").sortable('toArray',{ attribute: 'data-val' });
  }
  var geneGrpColl = $("input[name='TRAKgeneGrpColl']:checked").val()
  if( geneGrpColl== 'Yes'){
    geneGrpColl = $("input[name='TRAKgeneGrpCollCal']:checked").val()
  }

  var D = {'method':'TRAK',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'abb':window.bioAbbr,
            'combine':combGrp,
            'combineOrder':combGrpOrder,
            'geneGrp':geneSets,
            'geneGrpColl':geneGrpColl,
            'figOpt':window.figOpt};
  $(".TRAKbt").prop('disabled',true);
  document.getElementById("TRAKspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".TRAKbt").prop("disabled",false);
      document.getElementById("TRAKspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#TRAKnote').text(res);
        document.getElementById("TRAKfig").innerHTML = '';
        return;
      }

      showImg('TRAKfig',res);
      //document.getElementById("TRAKfig").innerHTML = '<img id="TRAKpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".TRAKpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#TRAKnote').text(request.status+request.responseText);
      document.getElementById("TRAKfig").innerHTML = '<img id="TRAKpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".TRAKbt").prop("disabled",false);
      document.getElementById("TRAKspin").style.visibility="hidden";
    }
  })

}
function DENSplot(){
  $('#DENSerror').text("");
  var cells=getCells('DENScell');
  if(cells.length==0){
    $('#DENSerror').text(VIPCellSelectionMsg);
    return;
  }

  var genes = getGenes('DENS');
  if(genes.length<1){
    $('#DENSerror').text("It needs at least a gene");
    return;
  }
  if(+$("#DENSprec").val() < +$("#DENSprec").prop('min')){
    $('#DENSerror').text("The minimal bandwidth is "+$("#DENSprec").prop('min'));
    return;
  }

  if($("#DENSsplit").val()==$("#DENSgrp").val()){
    $('#DENSerror').text("Please choose different annotations for Split and Color!");
    return;
  }
  var category = [$("#DENSsplit").val(),$("#DENSgrp").val()];
  var combGrp = [];
  var grp= [],combGrpOrder=[];
  if(category.includes("Custom_combine")){
    grp = Object.keys(window.bioCombine);
    combGrp = window.bioCombine;
    combGrpOrder = $("#ABBRdynList").sortable('toArray',{ attribute: 'data-val' });
  }
  if($("#DENSsplit").val()!="Custom_combine" && $("#DENSsplit").val()!='None'){
    grp.push($("#DENSsplit").val())
  }
  if($("#DENSgrp").val()!="Custom_combine" && $("#DENSgrp").val()!='None'){
    grp.push($("#DENSgrp").val())
  }

  var bw = $("#DENSprec").val();
  var D = {'method':'DENS',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'bw':bw,
            'category':category,
            'abb':window.bioAbbr,
            'combine':combGrp,
            'combineOrder':combGrpOrder,
            'figOpt':window.figOpt};
  $(".DENSbt").prop('disabled',true);
  document.getElementById("DENSspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".DENSbt").prop("disabled",false);
      document.getElementById("DENSspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#DENSerror').text(res);
        document.getElementById("DENSfig").innerHTML = '';
        return;
      }

      showImg('DENSfig',res);
      //document.getElementById("TRAKfig").innerHTML = '<img id="TRAKpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".DENSpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#DENSerror').text(request.status+request.responseText);
      document.getElementById("DENSfig").innerHTML = '<img id="DENSpng" src="data:image/png;base64," width=100% height="auto"/>';// height="450" width="450"
      $(".DENSbt").prop("disabled",false);
      document.getElementById("DENSspin").style.visibility="hidden";
    }
  })

}
function DENS2Dplot(){
  $('#DENS2Dnote').text("");
  var cells=getCells('DENS2Dcell');
  if(cells.length==0){
    $('#DENS2Dnote').text(VIPCellSelectionMsg);
    return;
  }

  var genes = getGenes('DENS2D');
  //$(".DENS2Dgenes").each(function(){
  //  if($(this).prop("checked")){
  //    genes.push($(this).val());
  //  }
  //})
  if(genes.length!=2){
    $('#DENS2Dnote').text("Please select two genes to plot");
    return;
  }

  document.getElementById("DENS2Dspin").style.visibility="visible";
  $(".DENS2Dbt").prop('disabled',true);
  $(".DENS2DpngBT").prop('disabled',true);

  var D = {'method':"DENS2D",
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':[],
            //'layout':$("#DUALlayout").val(),
            'cutoff':$("#DENS2Dcutoff").val(),
            'bandwidth':$("#DENS2Dbandwidth").val(),
            'figOpt':window.figOpt};
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".DENS2Dbt").prop("disabled",false);
      document.getElementById("DENS2Dspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#DENS2Dnote').text(res);
        document.getElementById("DENS2Dfig").innerHTML = '';
        return;
      }

      showImg('DENS2Dfig',res);
      $(".DENS2DpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#DENS2Dnote').text(request.status+request.responseText);
      document.getElementById("DENS2Dfig").innerHTML = '';// height="450" width="450"
      $(".DENS2Dbt").prop("disabled",false);
      document.getElementById("DENS2Dspin").style.visibility="hidden";
    }
  })

}
function EMBEDplot(){
  $('#EMBEDnote').text("");
  var cells=getCells('EMBEDcell');
  if(cells.length==0){
    $('#EMBEDnote').text(VIPCellSelectionMsg);
    return;
  }
  var genes = getGenes('EMBED');
  var grps = $(".EMBEDgrps:checked").map(function(){return this.value;}).get();
  var grpMeta = $(".EMBEDmetas:checked").map(function(){return this.value;}).get();

  if(genes.length==0) {
    if ((grps.length+grpMeta.length)==0){
      $('#EMBEDnote').text("Select an annotation and/or a gene");
      return;
    }
    if ($("#EMBEDsplit").val()!="NONE") {
      $('#EMBEDnote').text("Splitting only apply to genes");
      $("#EMBEDsplit").val("NONE");
    }
  }

  if($('#EMBEDslot').val()<2){
    $('#EMBEDnote').text("The minimal number of plots per row is 2");
    return;
  }


  document.getElementById("EMBEDspin").style.visibility="visible";
  $(".EMBEDbt").prop('disabled',true);
  $(".EMBEDpngBT").prop('disabled',true);

  var D = {'method':"EMBED",
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grps,
            'grpNum':grpMeta,
            'ncol':$('#EMBEDslot').val(),
            'layout':$("#EMBEDlayout").val(),
            //'abb':window.bioAbbr,
            'figOpt':window.figOpt};
  if($("#EMBEDsplit").val()!="NONE" && genes.length>0){
    D['splitGrp'] = $("#EMBEDsplit").val();
    grps.push($("#EMBEDsplit").val());
    D['grp'] = [...new Set(grps)];
  }
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".EMBEDbt").prop("disabled",false);
      document.getElementById("EMBEDspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#EMBEDnote').text(res);
        document.getElementById("EMBEDfig").innerHTML = '';
        return;
      }

      showImg('EMBEDfig',res);
      //document.getElementById("EMBEDfig").innerHTML = '<img id="EMBEDpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".EMBEDpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#EMBEDnote').text(request.status+request.responseText);
      document.getElementById("EMBEDfig").innerHTML = '';// height="450" width="450"
      $(".EMBEDbt").prop("disabled",false);
      document.getElementById("EMBEDspin").style.visibility="hidden";
    }
  })

}
function DUALplot(){
  $('#DUALnote').text("");
  var cells=getCells('DUALcell');
  if(cells.length==0){
    $('#DUALnote').text(VIPCellSelectionMsg);
    return;
  }

  var genes = [];
  $(".DUALgenes").each(function(){
    if($(this).prop("checked")){
      genes.push($(this).val());
    }
  })
  if(genes.length!=2){
    $('#DUALnote').text("Please select two genes to plot");
    return;
  }
  if(+$('#DUALcutoff').prop('min') > +$("#DUALcutoff").val()){
    $('#DUALnote').text("The minimal expression level is: "+$('#DUALcutoff').prop('min'));
    return;
  }

  document.getElementById("DUALspin").style.visibility="visible";
  $(".DUALbt").prop('disabled',true);
  $(".DUALpngBT").prop('disabled',true);

  var D = {'method':"DUAL",
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':[],
            'layout':$("#DUALlayout").val(),
            'cutoff':$("#DUALcutoff").val(),
            'figOpt':window.figOpt};
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".DUALbt").prop("disabled",false);
      document.getElementById("DUALspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#DUALnote').text(res);
        document.getElementById("DUALfig").innerHTML = '';
        return;
      }

      showImg('DUALfig',res);
      //document.getElementById("DUALfig").innerHTML = '<img id="DUALpng" src="data:image/png;base64,'+res+'" width=100% height="auto"/>';// height="450" width="450"
      $(".DUALpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#DUALnote').text(request.status+request.responseText);
      document.getElementById("DUALfig").innerHTML = '';// height="450" width="450"
      $(".DUALbt").prop("disabled",false);
      document.getElementById("DUALspin").style.visibility="hidden";
    }
  })
}
function SANKcheck(){
  $('#SANKnote').text("");
  var cells=getCells('SANKcell');
  if(cells.length==0){
    $('#SANKnote').text(VIPCellSelectionMsg);
    return;
  }
  var genes =getGenes('SANK');
  //var geneSets = getGeneSet('SANK');
  var grp = [];
  $(".SANKgrps").each(function(){
    if($(this).prop("checked")){
      grp.push($(this).val());
  }})

  var selSank = $("#SANKdynList").sortable('toArray');
  selSank=selSank.map(function(one){return one.replace("SANKli","");});
  if (selSank.length < 2) {
    $('#SANKnote').text("Please select at least two features (gene or annotation) to plot");
    return;
  }
  var D = {'method':'SANK',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            //'geneGrp':geneSets,
            'grp':grp,
            'sankOrder':selSank,
            'sankBin':$("#sankBin").val(),
            'imgW':$("#SANKw").val(),
            'imgH':$("#SANKh").val(),
            'abb':window.bioAbbr,
            'figOpt':window.figOpt};
  return(D);
}
function SANKplot(){
  D = SANKcheck();
  if (!D) return;
  $(".SANKbt").prop('disabled',true);
  document.getElementById("SANKspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".SANKbt").prop("disabled",false);
      document.getElementById("SANKspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#SANKnote').text(res);
        $('#SANKresize').text("");
        return;
      }
      $('#SANKresize').html(res);

    },
    error: function(request,status,error){
      $('#SANKnote').text(request.status+request.responseText);
      $(".SANKbt").prop("disabled",false);
      document.getElementById("SANKspin").style.visibility="hidden";

    }
  })
}
function SANKimgSave(){
  var D = SANKcheck();
  if (!D) return;
  D['imgSave'] = 'svg';
  $(".SANKbt").prop('disabled',true);
  document.getElementById("SANKspin").style.visibility="visible";
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".SANKbt").prop("disabled",false);
      document.getElementById("SANKspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#SANKnote').text(res);
        $('#SANKresize').text("");
        return;
      }
      var a = document.createElement('a');
      a.href = "data:image/svg+xml;base64,"+res;//
      a.download = "Sankey.svg";
      a.target="_blank";
      a.click();
    },
    error: function(request,status,error){
      $('#SANKnote').text(request.status+request.responseText);
      $(".SANKbt").prop("disabled",false);
      document.getElementById("SANKspin").style.visibility="hidden";

    }
  })

}
function STACBARplot(){
  $('#STACBARnote').text("");
  var cells=getCells('STACBARcell');
  if(cells.length==0){
    $('#STACBARnote').text(VIPCellSelectionMsg);
    return;
  }
  var genes =getGenes('STACBAR');
  //var geneSets = getGeneSet('SANK');
  var grp = [];
  $(".STACBARgrps:checkbox:checked").each(function(){
    grp.push($(this).val());
  })
  if((genes.length+grp.length)<2){
    $('#STACBARnote').text("Please select at least two features (genes and/or annotations) to plot");
    return;
  }

  var D = {'method':'STACBAR',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grp,
            'colorBy':$("#STACBARcol").val(),
            'Nbin':$("#STACBARbin").val(),
            'abb':window.bioAbbr,
            'figOpt':window.figOpt};

  $(".STACBARbt").prop('disabled',true);
  document.getElementById("STACBARspin").style.visibility="visible";
  $("#STACBARresize").html("");
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".STACBARbt").prop("disabled",false);
      document.getElementById("STACBARspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#STACBARnote').text(res);
        $('#STACBARresize').text("");
        return;
      }
      res = JSON.parse(res);
      //console.log(res);
      stackedBar('STACBARresize',res);
    },
    error: function(request,status,error){
      $('#STACBARnote').text(request.status+request.responseText);
      $(".STACBARbt").prop("disabled",false);
      document.getElementById("STACBARspin").style.visibility="hidden";

    }
  });

}
function BROWupdateTrack(addID,rmID){
  var selElem = $("#BROWdynList").sortable('toArray');
  if(addID){
    if(!selElem.includes(addID)){
      selElem=selElem.concat(addID);
    }
  }
  if(rmID){
    if(selElem.includes(rmID)){
      selElem.splice(selElem.indexOf(rmID),1);
    }
  }
  var htmlCheck = "";
  for(const one of selElem){
    htmlCheck += "<li id='"+one+"'><label>"+one.replace("li","")+"</label> <a onClick='BROWupdateTrack(null,\""+one+"\")' style='cursor: pointer; cursor: hand;'><i class='fa fa-trash-o'></i></a></li>";
  }
  $("#BROWdynList").html(htmlCheck);
}
function BROWreset(){
  if($("#BROWtrackAll").is(':checked')){
    var table = $('#BROWtable').DataTable();
    table.rows().every(function ( rowIdx, tableLoop, rowLoop ) {
      var data = this.data();
      BROWupdateTrack(data[0],null);
    });
  }else{
    console.log("UNchecked!");
    $("#BROWdynList").html("");
  }
  

}
function BROWplot(){
  $('#BROWnote').text("");
  var region = "", exUP=0, exDN=0,genes=[];
  if($("input[name='BROWinputRadio'][value='coordinate']").prop('checked')){
    region = $("#BROWregionChr").val()+"-"+ $("#BROWregionStart").val()+"-"+$("#BROWregionEnd").val();
  }else if($("input[name='BROWinputRadio'][value='gene']").prop('checked')){
    region = $("#BROWregionGeneName").val().trim();
    exUP = $("#BROWregionUpEx").val();
    exDN = $("#BROWregionDnEx").val();
  }else{
    $('#BROWnote').text("region is required");
    return;
  }
  var selBW = $("#BROWdynList").sortable('toArray');
  if(selBW.length<1){
    $('#BROWnote').text("Please select at least one track!");
    return;
  }
  var D = {'method':'plotBW',
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'bw':selBW,
            'cells':Int32Array.from([...Array(window.store.getState().annoMatrix.nObs).keys()]),
            'genes':getGenes('BROW'),
            'cutoff':$('#BROWgeneCutoff').val(),
            'region':region,
            'exUP':exUP,
            'exDN':exDN,
            'figOpt':window.figOpt};

  $(".BROWbt").prop('disabled',true);
  $(".BROWpngBT").prop('disabled',true);
  document.getElementById("BROWspin").style.visibility="visible";
  $("#BROWfig").html("");
  $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      BWimg=res;
      $(".BROWbt").prop("disabled",false);
      document.getElementById("BROWspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#BROWnote').text(res);
        $('#BROWfig').html("");
        return;
      }
      showImg('BROWfig',res);
      $(".BROWpngBT").prop('disabled',false);
    },
    error: function(request,status,error){
      $('#BROWnote').text(request.status+request.responseText);
      $(".BROWbt").prop("disabled",false);
      document.getElementById("BROWspin").style.visibility="hidden";
    }
  });

}
function COSMXplot(){
    $('#COSMXnote').text("");
    var sIDs =[];
    $(".COSMXsamples").each(function(){
        if($(this).prop("checked")){
            sIDs.push($(this).val());
        }
    });
    if(sIDs.length==0){
        $('#COSMXnote').text("Please select at least one capture sample to plot");
        return;
    }
    var genes={};
    $(".COSMXgeneColors").each(function(){
        genes[this.id.replace('COSMXgeneColor_','')] = this.value;
    });
    if(Object.keys(genes).length==0 && !$("#COSMXhistology").prop('checked') && !$("#COSMXcell").prop('checked')){
        $('#COSMXnote').text("Please choose at least one to plot: histology image, cell boundaries, transcripts");
        return;
    }
    if($("#COSMXgeneSize")<1 || $("#COSMXgeneSize")>10){
        $('#COSMXnote').text("Please choose transcript marker size between 1 and 10");
        return;
    }
    if(sIDs.length==1 && $("input[name='COSMXcood']:checked").val()=='global'){
        $('#COSMXnote').text("'Local' is set for plotting one capture");
        $("input[name='COSMXcood'][value='local']").prop('checked',true);
    }
    var D= {'method':'plotCOSMX',
        'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
        'cood': $("input[name='COSMXcood']:checked").val(),
        'sIDs':sIDs,
        'genes':genes,
        'geneSize':$("#COSMXgeneSize").val(),
        'histology':$("#COSMXhistology").prop('checked'),
        'cell': $("#COSMXcell").prop('checked')? $("#COXMXcellColor").val() : false,
        'figOpt':window.figOpt
    };
    $(".COSMXbt").prop('disabled',true);
    //$(".BROWpngBT").prop('disabled',true);
    document.getElementById("COSMXspin").style.visibility="visible";
    $("#COSMXfigs").html("");
    $.ajax({
        type:"POST",
        url:VIPurl,
        data:JSON.stringify(D),
        contentType: 'application/json;charset=UTF-8',//
        success:function(res){
            if(res.startsWith('ERROR')){
                $('#COSMXnote').text(res);
                $(".COSMXbt").prop("disabled",false);
                document.getElementById("COSMXspin").style.visibility="hidden";
                return;
            }
            var imgs = JSON.parse(res);
            var html = "";
            for(const one of Object.keys(imgs)){
                html += "<div style='text-align: center; overflow: scroll;'>"+one+"<figure class='zoom' onmousemove='localZoom(event)' id='COSMXfig_"+one+"'></figure></div>";
            }
            $("#COSMXfigs").html(html);
            for(const one of Object.keys(imgs)){
                showImg('COSMXfig_'+one,imgs[one]);//,'onmouseover="localZoom(this)" onmousemove="localZoom(this)"';
                $("#COSMXfig_"+one).css("background-image","url("+$("#COSMXimg_"+one).prop('src')+")");
            }
            $(".COSMXbt").prop("disabled",false);
            document.getElementById("COSMXspin").style.visibility="hidden";
        },
        error: function(request,status,error){
            $('#COSMXnote').text(request.status+request.responseText);
            $(".COSMXbt").prop("disabled",false);
            document.getElementById("COSMXspin").style.visibility="hidden";
        }
    });
}
function VISIUMplot(){
    $('#VISIUMnote').text("");
    var sIDs =[];
    $(".VISIUMsamples").each(function(){
        if($(this).prop("checked")){
            sIDs.push($(this).val());
        }
    });
    if(sIDs.length==0){
        $('#VISIUMnote').text("Please select at least one visium slide to plot");
        return;
    }
    var genes = getGenes('VISIUM');
    var grps = $(".VISIUMgrps:checked").map(function(){return this.value;}).get();
    var grpMeta = $(".VISIUMmetas:checked").map(function(){return this.value;}).get();
    var D= {'method':'plotVisium',
        'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
        'sIDs':sIDs,
        'genes':genes,
        'grp':grps,
        'grpNum':grpMeta,
        'sortID':$("#VISIUMsortID").prop('checked'),
        'ncol':$("#VISIUMslot").val(),
        'alpha':$("#VISIUMalpha").val(),
        'dotsize':$("#VISIUMdotsize").val(),
        'subsize':$("#VISIUMsubsize").val(),
        'figOpt':window.figOpt
    };
    $(".VISIUMbt").prop('disabled',true);
    $(".VISIUMpngBT").prop('disabled',true);
    document.getElementById("VISIUMspin").style.visibility="visible";
    $("#VISIUMfig").html("");
    $.ajax({
        type:"POST",
        url:VIPurl,
        data:JSON.stringify(D),
        contentType: 'application/json;charset=UTF-8',
        success:function(res){
            $(".VISIUMpngBT").prop('disabled',false);
            $(".VISIUMbt").prop("disabled",false);
            document.getElementById("VISIUMspin").style.visibility="hidden";
            if(res.startsWith('ERROR')){
                $('#VISIUMnote').text(res);
                return;
            }
            //var imgs = JSON.parse(res);
            visiumData = res;
            showImg('VISIUMfig',res);
        },
        error: function(request,status,error){
            $('#VISIUMnote').text(request.status+request.responseText);
            $(".VISIUMbt").prop("disabled",false);
            document.getElementById("VISIUMspin").style.visibility="hidden";
        }
    });
}
function localZoom(e){
    //console.log(event.offsetX," ",event.offsetX);
    var zoomer = e.currentTarget;
    e.offsetX ? offsetX = e.offsetX : offsetX = e.touches[0].pageX;
    e.offsetY ? offsetY = e.offsetY : offsetX = e.touches[0].pageY;
    x = offsetX/zoomer.offsetWidth*100;
    y = offsetY/zoomer.offsetHeight*100;
    zoomer.style.backgroundPosition = x + '% ' + y + '%';
}
function CLIplot(){
  $('#CLInote').text("");
  $('#CLIsvg').html("");
  var cells=getCells('CLIcell');
  if(cells.length==0){
    $('#CLInote').text(VIPCellSelectionMsg);
    return;
  }

  var genes = getGenes('CLI');
  if(genes.length==0){
    if(!confirm("No genes is sepecified. Are you sure you want to continue with all genes (might be very slow or out of memory)?")){
      return;
    }
  }

  $('#printNotebook').show();
  $('#hideCode').show();
  $('#hideOutput').show();
  var mywin;
  if (editor.getSession().getValue().includes("```")) {
    mywin=window.open('','_blank');
    mywin.document.write('Please wait for the run to finish ... ...! Check the VIP window for status (spinning wheel).');
    $('#printNotebook').hide();
    $('#hideCode').hide();
    $('#hideOutput').hide();
  }

  var grps = [];
  $(".CLIgrps:checkbox:checked").each(function(){
    grps.push($(this).val());
  })
  if(grps.length==0){
    $('#CLInote').text("Please select at least one annotation");
    return;
  }

  var layouts = [];
  $(".CLIlayouts:checkbox:checked").each(function(){
    layouts.push($(this).val());
  })
  if(layouts.length==0){
    $('#CLInote').text("Please select at least one layout");
    return;
  }

  document.getElementById("CLIspin").style.visibility="visible";
  $(".CLIbt").prop('disabled',true);

  var D = {'method':"CLI",
            'dataset': window.store.getState().config.displayNames.dataset+'.h5ad',
            'cells':cells,
            'genes':genes,
            'grp':grps,
            'layout':layouts,
            'script':editor.getSession().getValue()};
   $.ajax({
    type:"POST",
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',//
    success: function(res){
      $(".CLIbt").prop("disabled",false);
      document.getElementById("CLIspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        $('#CLInote').text(res);
        $("#CLIresize").html("");
        return;
      }
      window.testD=res;
      if (editor.getSession().getValue().includes("```")) {
        mywin.document.body.innerText='';
        mywin.document.write(res);
        mywin.document.close();
        $('#CLInote').text("Success! Please go to tab named as "+mywin.document.title);
      }
      var filteredRes = '<style type="text/css"> pre { display: block; font-size: 12px; line-height: 1.42857143; word-break: break-all; word-wrap: break-word; color: #333333; background-color: #f5f5f5; border: 1px solid #ccc; border-radius: 2px; }';
      var readIt = false;
      res.split(/(\r\n|\n|\r)/g).forEach(function(line) {
        if (line.includes('div.inner_cell') || line.includes('<body>')) readIt = true;
        if (readIt) {
  //          line = line.replace('output_subarea','');
          filteredRes += line;
        }
        if (line.includes('<\/body>')) readIt = false;
      });
      // console.log(filteredRes);
      $("#CLIresize").html(res);

      var num = 0;
      for (let item of document.getElementsByClassName('output_svg')) {
        if (item.getElementsByTagName('svg')[0] != null) {
          makeIdsUnique(item.getElementsByTagName('svg')[0], false);
          var onclickCode = "var link = document.createElement('a'); link.href = 'data:application/octet-stream;base64,' + btoa(unescape(encodeURIComponent(document.getElementsByClassName('output_svg')["+num+"].getElementsByTagName('svg')[0].outerHTML))); link.download = 'CLI-"+num+".svg'; link.click();"
        } else {
          makeIdsUnique(item.getElementsByTagName('img')[0], false);
          var onclickCode = "var link = document.createElement('a'); link.href = document.getElementsByClassName('output_svg')["+num+"].getElementsByTagName('img')[0].src; link.download = 'CLI-"+num+".svg'; link.click();"
        }
        item.innerHTML += '<p><button class="CLIbt" onclick="' + onclickCode + '">Save as SVG</button></p>';
        num ++;
      }
    },
    error: function(request,status,error){
      $('#CLInote').text(request.status+request.responseText);
      $("#CLIresize").html("");
      $(".CLIbt").prop("disabled",false);
      document.getElementById("CLIspin").style.visibility="hidden";
    }
  });
}
function metaSel(){
  document.getElementById("metaCellBTspin").style.visibility="visible";
  var cells=[], x = document.getElementsByClassName('CLIcell');
  for(i=0;i<x.length;i++){
    if(x[i].checked){
      if(!!window.store.getState().differential[x[i].value]){
        if(cells.length==0){
          cells = window.store.getState().differential[x[i].value];
        }else{
          var tmp = new Int32Array(cells.length+window.store.getState().differential[x[i].value].length);
          tmp.set(cells);
          tmp.set(window.store.getState().differential[x[i].value],cells.length);
          cells = tmp;
        }
      }
    }
  }
  if(cells.length==0){
    $('#CLInote').text("Please use selection tools to put meta cells into groups");
    return;
  }
  console.log(cells.length)
  var D = {'method':'mergeMeta',
            'cells':cells,
            'metaPostfix': uniqID};
  $.ajax({
    type:'POST',
    url: VIPurl,
    data:JSON.stringify(D),
    contentType: 'application/json;charset=UTF-8',
    success: function(res){
      document.getElementById("metaCellBTspin").style.visibility="hidden";
      if(res.startsWith('ERROR')){
        alert(res);
        return;
      }
      window.open(res, uniqID);
    },
    error: function(request,status,error){
      document.getElementById("metaCellBTspin").style.visibility="hidden";
      alert(request.status+request.responseText);
    }
  });
}
//save and load
function saveContent(){
  sync();

  var content = {'saveName':'cellxgene.'+window.store.getState().config.displayNames.dataset};
  content['selCells'] = cellSave();
  content['selGenes'] = geneSave();
  content['dropdown'] = selectSave();
  content['checkbox'] = checkSave();
  content['radio'] = radioSave();
  content['number'] = numberSave();
  content['sortable'] = sortableSave();
  content['eID'] = eIDsave(['GSEAenable','GSEAcollapse']);//
  return content;
}
function saveLocal(){
  var content = saveContent();
  var hiddenE = document.createElement('a');
  hiddenE.href="data:attachment/text,"+encodeURIComponent(JSON.stringify(content));
  hiddenE.target='_blank';
  hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.info.txt';
  hiddenE.click();
}
function saveLayout(){
  var nodes = GridStack.init().engine.nodes;

  var content = '{"total":' + nodes.length + ',"layout":[';
  for (var i=0; i<nodes.length; i++) {
    content += '{"name":"' + nodes[i].text +'","x":'+nodes[i].x+',"y":'+nodes[i].y+',';
    var trans =  $('#'+ nodes[i].text + '_img')[0].style.transform;
    trans.match(/scaleX/) ? content += '"flipx":1,' : content += '"flipx":0,' ;
    trans.match(/scaleY/) ? content += '"flipy":1,' : content += '"flipy":0,' ;
    var rot=0;
    if (trans.match(/([0-9]+)deg/)) var rot = RegExp.$1;
    content += '"rotate":' +rot+ '},';
  }
  content = content.slice(0,-1) + ']}';
  console.log(JSON.parse(content));

  var hiddenE = document.createElement('a');
  hiddenE.href="data:attachment/text,"+encodeURIComponent(content);
  hiddenE.target='_blank';
  hiddenE.download='cellxgene.'+window.store.getState().config.displayNames.dataset+'.json.txt';
  hiddenE.click();
}
function loadClick(){
  if($("input[name='LOADurlRd']:checked").val()=='Yes'){
    testVIPgetData($("#LOADurl").val(),urlload);
  }else{
    $("#loadfile").trigger('click');
  }
}
function urlload(){
  var content = JSON.parse(decodeURIComponent(this.responseText));
  loadexe(content);
  $("#LOADnote").html("Session was loaded<br>");//LOADplotRd
  if($("input[name='LOADplotRd']:checked").val()=='Yes'){
    DEGfind();
    randomPlot('LOADnote',bSaved=false);
  }
}
function load(){
  var reader = new FileReader();
  reader.onload=function(ev){
    var content = JSON.parse(decodeURIComponent(ev.target.result));
    loadexe(content);
    $("#LOADnote").html("Session was loaded<br>");
    $("#loadfile").val('');
    if($("input[name='LOADplotRd']:checked").val()=='Yes'){
      DEGfind();
      randomPlot('LOADnote',bSaved=false);
    }
  };
  reader.readAsText($("#loadfile")[0].files[0]);
}
function loadexe(content){
  if(content['saveName']!='cellxgene.'+window.store.getState().config.displayNames.dataset){
    if(!confirm("Saved information is NOT matching with the current data! Continue?")){
      return;
    }
  }
  cellLoad(content['selCells']);
  geneLoad(content['selGenes']);
  sync();
  selectLoad(content['dropdown']);
  checkLoad(content['checkbox']);
  radioLoad(content['radio']);
  numberLoad(content['number']);
  sortableLoad(content['sortable']);
  eIDload(content['eID']);
}
// test the VIP
function tVIPcreate(act,eID){
  $(".tVIPbt").prop('disabled',true);
  $("#"+eID).html("");

  if(act=='random'){
    createTest($("#tVIPgrp").val(),'tVIPnote');
  }else if(act=='cells'){
    randomInitDEG($("#tVIPgrp").val(),'tVIPnote');
  }else if(act=='selections'){
    randomPlot('tVIPnote');
  }
}
// ignore global definition for stackedBar legend circle
function GetStyle(CLASSname){
    var styleSheets = document.styleSheets;
    var styleSheetsLength = styleSheets.length;
    for(var i = 0; i < styleSheetsLength; i++){
        if (styleSheets[i].rules ) { var classes = styleSheets[i].rules; }
        else {
            try {  if(!styleSheets[i].cssRules) {continue;} }
            //Note that SecurityError exception is specific to Firefox.
            catch(e) { if(e.name == 'SecurityError') { console.log("SecurityError. Cant readd: "+ styleSheets[i].href);  continue; }}
            var classes = styleSheets[i].cssRules ;
        }
        for (var x = 0; x < classes.length; x++) {
            if (classes[x].selectorText == CLASSname) {
                return classes[x];
            }
        }
    }
}
GetStyle('circle').style = {};
</script>

<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/DataTables/datatables.min.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/ace/ace.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/script.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/select.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/endAll.js"></script>
<script type="text/javascript" src="https://interactivereport.github.io/cellxgene_VIP/static/stackedbar/plot-transform.js"></script>
<script type="text/javascript" src="https://baohongz.github.io/figureComposer/dist/svg-inject.js"></script>
<div id="CLIVignetteCode1" style="display:none;">
import warnings
warnings.filterwarnings('ignore')
import scanpy as sc
from matplotlib import pyplot as plt
%config InlineBackend.figure_formats = ['svg']
#
sc.pp.scale(adata, zero_center=False)
sc.tl.rank_genes_groups(adata,groupby='cell_type',groups=['OPC','Astrocytes'],method='t-test')
marker_genes=['LHFPL3', 'PCDH15', 'DSCAM', 'TNR', 'NLGN1','AQP4', 'VCAN','CUX2']
geneGrpPos = [(0,2), (3,4), (5,7)]
geneGrpName = ['set1', 'set2', 'set3']
fig,axs = plt.subplots(1,3,figsize=(15,5))
sc.pl.rank_genes_groups_violin(adata,groups='OPC',n_genes=8,ax=axs[0],show=False,strip=False)
sc.pl.stacked_violin(adata,marker_genes,groupby='cell_type',var_group_positions=geneGrpPos,var_group_labels=geneGrpName,row_palette='muted',rotation=90,yticklabels=False,ax=axs[1],show=False)
sc.pl.matrixplot(adata,marker_genes,groupby='cell_type',var_group_positions=geneGrpPos,var_group_labels=geneGrpName,ax=axs[2],show=False)
fig.tight_layout()
</div>

<div id="CLIVignetteCode2" style="display:none;">
from pyarrow import feather
import pandas as pd
X = adata.to_df()
meta = pd.concat([adata.obs,adata.obsm.to_df()],axis=1)
meta[meta.select_dtypes(['category']).columns] = meta.select_dtypes(['category']).apply(lambda x: x.astype('object'))
feather.write_feather(X,strPath+'.feather')
feather.write_feather(meta,strPath+".obs.feather")

%%R
library(Seurat)
library(patchwork)
library(ggplot2)
D <- t(tibble::column_to_rownames(arrow::read_feather(paste(strPath,".feather",sep="")), var="__index_level_0__"))
meta <- as.data.frame(arrow::read_feather(paste(strPath,".obs.feather",sep="")))
colnames(D) <- meta[,"name_0"]
X <- CreateSeuratObject(counts=D)
X@meta.data <- cbind(X@meta.data,meta)
# For seurat < ver.5, use X@assays$RNA@data <- X@assays$RNA@counts
X[['RNA']]['data'] <- X[['RNA']]['counts']

%Rdevice png

%%R -r 600 -w 7200 -h 3600
X[["percent.mt"]] <- PercentageFeatureSet(X, pattern = "^MT-")
VlnPlot(X, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(X, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(X, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

%%R -r 300 -w 3000 -h 1600
X <- subset(X, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
X <- FindVariableFeatures(X, selection.method = "vst", nfeatures = 2000,verbose=F)
top10 <- head(VariableFeatures(X,verbose=F), 10)
plot1 <- VariableFeaturePlot(X)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2

%Rdevice svg

%%R -w 8 -h 4
all.genes <- rownames(X)
X <- ScaleData(X, features = all.genes,verbose=F)
X <- RunPCA(X, features = VariableFeatures(object = X))
print(X[["pca"]], dims = 1:5, nfeatures = 5)
DimPlot(X, reduction = "pca")

%Rdevice png

%%R -r 150 -w 1000 -h 1000
DimHeatmap(X, dims = 1, cells = 500, balanced = TRUE)

%%R -r 150 -w 2000 -h 2400
DimHeatmap(X, dims = 1:15, cells = 500, balanced = TRUE)

%Rdevice svg

%%R -w 6 -h 8
sink(stdout(),type="message")
suppressMessages(library(dplyr))
X <- FindNeighbors(X, dims = 1:10, verbose=FALSE)
X <- FindClusters(X, resolution = 0.5)
X.markers <- FindAllMarkers(X, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
top10 <- X.markers %>% group_by(cluster) %>% top_n(n = 10, wt = avg_log2FC)
DoHeatmap(X, features = top10$gene, label=FALSE) + theme(text = element_text(size = 8))
sessionInfo()
</div>

<div id="CLIVignetteCode3" style="display:none;">
```{python}
from pyarrow import feather
import pandas as pd
from anndata import read_h5ad

strData = r.strPath + '.h5ad'
adata=read_h5ad(strData)
X = adata.to_df()
meta = pd.concat([adata.obs,adata.obsm.to_df()],axis=1)
meta[meta.select_dtypes(['category']).columns] = meta.select_dtypes(['category']).apply(lambda x: x.astype('object'))
feather.write_feather(X,r.strPath+'.feather')
feather.write_feather(meta,r.strPath+".obs.feather")
```

```{r block1, message=FALSE, fig.width=14, fig.height=6, dpi=300}
library(Seurat)
library(patchwork)
library(ggplot2)
D <- t(tibble::column_to_rownames(arrow::read_feather(paste(strPath,".feather",sep="")), var="__index_level_0__"))
meta <- as.data.frame(arrow::read_feather(paste(strPath,".obs.feather",sep="")))
colnames(D) <- meta[,"name_0"]
X <- CreateSeuratObject(counts=D)
X@meta.data <- cbind(X@meta.data,meta)
# For seurat < ver.5, use X@assays$RNA@data <- X@assays$RNA@counts
X[['RNA']]['data'] <- X[['RNA']]['counts']


X[["percent.mt"]] <- PercentageFeatureSet(X, pattern = "^MT-")
VlnPlot(X, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
plot1 <- FeatureScatter(X, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(X, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2


X <- subset(X, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
X <- FindVariableFeatures(X, selection.method = "vst", nfeatures = 2000,verbose=F)
top10 <- head(VariableFeatures(X,verbose=F), 10)
plot1 <- VariableFeaturePlot(X)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

```{r block2, message=FALSE, fig.width=8, fig.height=6, dev='svg'}
all.genes <- rownames(X)
X <- ScaleData(X, features = all.genes,verbose=F)
X <- RunPCA(X, features = VariableFeatures(object = X))
print(X[["pca"]], dims = 1:5, nfeatures = 5)
DimPlot(X, reduction = "pca")
```
</div>
</body>
</html>
