<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Supplemental Tutorial: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data</title>
  <meta name="description" content="Supplemental Tutorial: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Supplemental Tutorial: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Supplemental Tutorial: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="web-source.html"/>
<script src="libs/header-attrs-2.11/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">CellXgene VIP tutorial</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Getting started with cellxgene VIP</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#why-use-cellxgene-vip"><i class="fa fa-check"></i><b>1.1</b> Why use cellxgene VIP?</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#getting-set-up"><i class="fa fa-check"></i><b>1.2</b> Getting Set up</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="index.html"><a href="index.html#execute-anaconda"><i class="fa fa-check"></i><b>1.2.1</b> Execute anaconda</a></li>
<li class="chapter" data-level="1.2.2" data-path="index.html"><a href="index.html#cellxgene-installation"><i class="fa fa-check"></i><b>1.2.2</b> Cellxgene installation</a></li>
<li class="chapter" data-level="1.2.3" data-path="index.html"><a href="index.html#r-dependencies"><i class="fa fa-check"></i><b>1.2.3</b> R dependencies</a></li>
<li class="chapter" data-level="1.2.4" data-path="index.html"><a href="index.html#run-cellxgene-by-h5ad-file"><i class="fa fa-check"></i><b>1.2.4</b> Run cellxgene by h5ad file</a></li>
<li class="chapter" data-level="1.2.5" data-path="index.html"><a href="index.html#cellxgene-on-web-browser"><i class="fa fa-check"></i><b>1.2.5</b> Cellxgene on web browser</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#authors"><i class="fa fa-check"></i><b>1.3</b> Authors</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="web-source.html"><a href="web-source.html"><i class="fa fa-check"></i><b>2</b> Web source</a>
<ul>
<li class="chapter" data-level="" data-path="web-source.html"><a href="web-source.html#cellxgene-vip"><i class="fa fa-check"></i>Cellxgene VIP</a></li>
<li class="chapter" data-level="" data-path="web-source.html"><a href="web-source.html#cellxgene"><i class="fa fa-check"></i>Cellxgene</a></li>
<li class="chapter" data-level="" data-path="web-source.html"><a href="web-source.html#python-packages"><i class="fa fa-check"></i>Python packages</a></li>
<li class="chapter" data-level="" data-path="web-source.html"><a href="web-source.html#others"><i class="fa fa-check"></i>Others</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html"><i class="fa fa-check"></i><b>3</b> How to use Cellxgene VIP</a>
<ul>
<li class="chapter" data-level="3.1" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#graphical-user-interface-of-cellxgene-and-vip"><i class="fa fa-check"></i><b>3.1</b> Graphical user interface of cellxgene and VIP</a></li>
<li class="chapter" data-level="3.2" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#cell-selection-by-categorial-annotations"><i class="fa fa-check"></i><b>3.2</b> Cell selection by categorial annotations</a></li>
<li class="chapter" data-level="3.3" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#cell-selection-by-brushing-on-distribution-of-continues-variables"><i class="fa fa-check"></i><b>3.3</b> Cell selection by brushing on distribution of continues variables</a></li>
<li class="chapter" data-level="3.4" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#free-hand-lasso-selection-on-dots-representing-cells"><i class="fa fa-check"></i><b>3.4</b> Free hand Lasso selection on dots representing cells</a></li>
<li class="chapter" data-level="3.5" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-figure-option"><i class="fa fa-check"></i><b>3.5</b> VIP – Figure Option</a></li>
<li class="chapter" data-level="3.6" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-add-genes-gene-sets"><i class="fa fa-check"></i><b>3.6</b> VIP – Add Genes / Gene Sets</a></li>
<li class="chapter" data-level="3.7" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-violin-plot"><i class="fa fa-check"></i><b>3.7</b> VIP – Violin Plot</a></li>
<li class="chapter" data-level="3.8" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-stacked-violin"><i class="fa fa-check"></i><b>3.8</b> VIP – Stacked Violin</a></li>
<li class="chapter" data-level="3.9" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-heatmap"><i class="fa fa-check"></i><b>3.9</b> VIP – Heatmap</a></li>
<li class="chapter" data-level="3.10" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-umaptsne"><i class="fa fa-check"></i><b>3.10</b> VIP – UMAP/tSNE</a></li>
<li class="chapter" data-level="3.11" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-dot-plot"><i class="fa fa-check"></i><b>3.11</b> VIP – Dot Plot</a></li>
<li class="chapter" data-level="3.12" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-track-plot"><i class="fa fa-check"></i><b>3.12</b> VIP – Track Plot</a></li>
<li class="chapter" data-level="3.13" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-density-plot"><i class="fa fa-check"></i><b>3.13</b> VIP – Density Plot</a></li>
<li class="chapter" data-level="3.14" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-density-scatter-plot"><i class="fa fa-check"></i><b>3.14</b> VIP – Density Scatter Plot</a></li>
<li class="chapter" data-level="3.15" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-dual-genes"><i class="fa fa-check"></i><b>3.15</b> VIP – Dual Genes</a></li>
<li class="chapter" data-level="3.16" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-sankey-diagram"><i class="fa fa-check"></i><b>3.16</b> VIP – Sankey Diagram</a></li>
<li class="chapter" data-level="3.17" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-stacked-barplot"><i class="fa fa-check"></i><b>3.17</b> VIP – Stacked Barplot</a></li>
<li class="chapter" data-level="3.18" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-gene-detected"><i class="fa fa-check"></i><b>3.18</b> VIP – Gene Detected</a></li>
<li class="chapter" data-level="3.19" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-deg-differential-expressed-genes"><i class="fa fa-check"></i><b>3.19</b> VIP – DEG (Differential Expressed Genes)</a></li>
<li class="chapter" data-level="3.20" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip---pre-computed-deg"><i class="fa fa-check"></i><b>3.20</b> VIP - Pre-computed DEG</a></li>
<li class="chapter" data-level="3.21" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-marker-genes"><i class="fa fa-check"></i><b>3.21</b> VIP – Marker Genes</a></li>
<li class="chapter" data-level="3.22" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip---spatial-transcriptomics"><i class="fa fa-check"></i><b>3.22</b> VIP - Spatial Transcriptomics</a></li>
<li class="chapter" data-level="3.23" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-command-line-interface"><i class="fa fa-check"></i><b>3.23</b> VIP – Command Line Interface</a></li>
<li class="chapter" data-level="3.24" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-comb.-abbr."><i class="fa fa-check"></i><b>3.24</b> VIP – Comb. &amp; Abbr.</a></li>
<li class="chapter" data-level="3.25" data-path="how-to-use-cellxgene-vip.html"><a href="how-to-use-cellxgene-vip.html#vip-other-functions"><i class="fa fa-check"></i><b>3.25</b> VIP – Other Functions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="methods.html"><a href="methods.html"><i class="fa fa-check"></i><b>4</b> Methods</a>
<ul>
<li class="chapter" data-level="" data-path="methods.html"><a href="methods.html#client-side-integration-by-a-jspanel-window-vip"><i class="fa fa-check"></i>Client-side Integration by a jsPanel Window (VIP)</a></li>
<li class="chapter" data-level="" data-path="methods.html"><a href="methods.html#server-side-integration"><i class="fa fa-check"></i>Server-side Integration</a></li>
<li class="chapter" data-level="" data-path="methods.html"><a href="methods.html#communication-between-vip-and-cellxgene-web-gui"><i class="fa fa-check"></i>Communication between VIP and cellxgene web GUI</a></li>
<li class="chapter" data-level="4.1" data-path="methods.html"><a href="methods.html#diffxpy-integration"><i class="fa fa-check"></i><b>4.1</b> Diffxpy Integration</a></li>
<li class="chapter" data-level="4.2" data-path="methods.html"><a href="methods.html#create-h5ad-file-from-seurat-object"><i class="fa fa-check"></i><b>4.2</b> Create h5ad file from Seurat object</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="helpful-tips.html"><a href="helpful-tips.html"><i class="fa fa-check"></i><b>5</b> Helpful Tips</a>
<ul>
<li class="chapter" data-level="5.1" data-path="helpful-tips.html"><a href="helpful-tips.html#handle-nulls-in-categorical-annotation"><i class="fa fa-check"></i><b>5.1</b> Handle nulls in categorical annotation</a></li>
<li class="chapter" data-level="5.2" data-path="helpful-tips.html"><a href="helpful-tips.html#display-full-traceback-stack-for-debugging-in-vip"><i class="fa fa-check"></i><b>5.2</b> Display full traceback stack for debugging in VIP</a></li>
<li class="chapter" data-level="5.3" data-path="helpful-tips.html"><a href="helpful-tips.html#pitfall-of-using-special-characters"><i class="fa fa-check"></i><b>5.3</b> Pitfall of using special characters</a></li>
<li class="chapter" data-level="5.4" data-path="helpful-tips.html"><a href="helpful-tips.html#potential-use-for-bulk-or-pseudo-bulk-sample-dataset"><i class="fa fa-check"></i><b>5.4</b> Potential use for bulk or pseudo bulk sample dataset</a></li>
<li class="chapter" data-level="5.5" data-path="helpful-tips.html"><a href="helpful-tips.html#common-mistakes-in-naming"><i class="fa fa-check"></i><b>5.5</b> Common mistakes in naming</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Supplemental Tutorial: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="header">
<h1 class="title">Supplemental Tutorial: Cellxgene VIP unleashes full power of interactive visualization and integrative analysis of scRNA-seq, spatial transcriptomics, and multiome data</h1>
</div>
<div id="getting-started-with-cellxgene-vip" class="section level1" number="1">
<h1><span class="header-section-number">Chapter 1</span> Getting started with cellxgene VIP</h1>
<p>This is a cellxgene VIP tutorial book written in <strong>Markdown</strong>.</p>
<div id="why-use-cellxgene-vip" class="section level2" number="1.1">
<h2><span class="header-section-number">1.1</span> Why use cellxgene VIP?</h2>
<p>To meet the growing demands from scientists to effectively extract deep insights from single cell RNA-seq datasets, we developed cellxgene VIP, a frontend interactive visualization plugin to cellxgene framework, which directly interacts with in-memory data to generate a comprehensive set of plots in high resolution, perform advanced analysis, and make data downloadable for further analysis. It makes large scale scRNA-seq data visualization and analysis more accessible and reproducible with the potential to become an ecosystem for the scientific community to contribute even more modules to the Swiss knife of scRNA-seq data exploration tool.</p>
</div>
<div id="getting-set-up" class="section level2" number="1.2">
<h2><span class="header-section-number">1.2</span> Getting Set up</h2>
<div id="execute-anaconda" class="section level3" number="1.2.1">
<h3><span class="header-section-number">1.2.1</span> Execute anaconda</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="index.html#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">bash</span> ~/Downloads/Anaconda3-2020.02-Linux-x86_64.sh</span></code></pre></div>
<p>If anaconda is not installed on server, you can install it following anaconda documentation (<a href="https://docs.anaconda.com/anaconda/install/linux/" class="uri">https://docs.anaconda.com/anaconda/install/linux/</a>)
### Create and enable conda environment</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="index.html#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># clone repo from cellxgene VIP github</span></span>
<span id="cb2-2"><a href="index.html#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">git</span> clone https://github.com/interactivereport/cellxgene_VIP.git</span>
<span id="cb2-3"><a href="index.html#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="bu">cd</span> cellxgene_VIP</span>
<span id="cb2-4"><a href="index.html#cb2-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-5"><a href="index.html#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># conda environment</span></span>
<span id="cb2-6"><a href="index.html#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> <span class="op">&lt;</span>path to Anaconda3<span class="op">&gt;</span>/etc/profile.d/conda.sh <span class="er">(</span><span class="ex">Default:</span> /opt/anaconda3/etc/profile.d/conda.sh<span class="kw">)</span></span>
<span id="cb2-7"><a href="index.html#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> config <span class="at">--set</span> channel_priority flexible</span>
<span id="cb2-8"><a href="index.html#cb2-8" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> env create <span class="at">-n</span> <span class="op">&lt;</span>env name, such as: VIP<span class="op">&gt;</span> -f VIP.yml <span class="er">(</span><span class="ex">system-wide</span> R<span class="kw">)</span> <span class="ex">or</span> VIP_conda_R.yml <span class="er">(</span><span class="bu">local</span> <span class="va">R</span> <span class="va">under</span> <span class="va">conda</span>, <span class="va">no</span> <span class="va">root</span> <span class="va">privilege</span> <span class="va">needed</span><span class="kw">)</span></span></code></pre></div>
<p>Activate conda environment</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="index.html#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate <span class="op">&lt;</span>env name, such as: VIP<span class="op">&gt;</span></span></code></pre></div>
<p>or</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="index.html#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="bu">source</span> activate <span class="op">&lt;</span>env name<span class="op">&gt;</span></span></code></pre></div>
</div>
<div id="cellxgene-installation" class="section level3" number="1.2.2">
<h3><span class="header-section-number">1.2.2</span> Cellxgene installation</h3>
<p>Install cellxgene by running config.sh in “cellxgene_VIP” directory</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb5-1"><a href="index.html#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="ex">./config.sh</span></span></code></pre></div>
</div>
<div id="r-dependencies" class="section level3" number="1.2.3">
<h3><span class="header-section-number">1.2.3</span> R dependencies</h3>
<p>Install all required R packages on linux:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="index.html#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">LIBARROW_MINIMAL</span><span class="op">=</span>false</span>
<span id="cb6-2"><a href="index.html#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="co">#  ensure that the right instance of R is used. e.g. system-wide: /bin/R or /usr/bin/R ; local R under conda: ~/.conda/envs/VIP_conda_R/bin/R</span></span>
<span id="cb6-3"><a href="index.html#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span> R</span>
<span id="cb6-4"><a href="index.html#cb6-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-5"><a href="index.html#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(devtools)) install.packages(&quot;devtools&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-6"><a href="index.html#cb6-6" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(Cairo)) devtools::install_version(&quot;Cairo&quot;,version=&quot;1.5-12&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-7"><a href="index.html#cb6-7" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(foreign)) devtools::install_version(&quot;foreign&quot;,version=&quot;0.8-76&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-8"><a href="index.html#cb6-8" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(ggpubr)) devtools::install_version(&quot;ggpubr&quot;,version=&quot;0.3.0&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-9"><a href="index.html#cb6-9" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(ggrastr)) devtools::install_version(&quot;ggrastr&quot;,version=&quot;0.1.9&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-10"><a href="index.html#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(arrow)) devtools::install_version(&quot;arrow&quot;,version=&quot;2.0.0&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-11"><a href="index.html#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(Seurat)) devtools::install_version(&quot;Seurat&quot;,version=&quot;3.2.3&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-12"><a href="index.html#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(rmarkdown)) devtools::install_version(&quot;rmarkdown&quot;,version=&quot;2.5&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-13"><a href="index.html#cb6-13" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(tidyverse)) devtools::install_version(&quot;tidyverse&quot;,version=&quot;1.3.0&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-14"><a href="index.html#cb6-14" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(viridis)) devtools::install_version(&quot;viridis&quot;,version=&quot;0.5.1&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-15"><a href="index.html#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(BiocManager)) devtools::install_version(&quot;BiocManager&quot;,version=&quot;1.30.10&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-16"><a href="index.html#cb6-16" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(fgsea)) BiocManager::install(&quot;fgsea&quot;)&#39;</span></span>
<span id="cb6-17"><a href="index.html#cb6-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-18"><a href="index.html#cb6-18" aria-hidden="true" tabindex="-1"></a><span class="co"># These should be already installed as dependencies of above packages</span></span>
<span id="cb6-19"><a href="index.html#cb6-19" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(dbplyr)) devtools::install_version(&quot;dbplyr&quot;,version=&quot;1.0.2&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-20"><a href="index.html#cb6-20" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(RColorBrewer)) devtools::install_version(&quot;RColorBrewer&quot;,version=&quot;1.1-2&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-21"><a href="index.html#cb6-21" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(glue)) devtools::install_version(&quot;glue&quot;,version=&quot;1.4.2&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-22"><a href="index.html#cb6-22" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(gridExtra)) devtools::install_version(&quot;gridExtra&quot;,version=&quot;2.3&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-23"><a href="index.html#cb6-23" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(ggrepel)) devtools::install_version(&quot;ggrepel&quot;,version=&quot;0.8.2&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-24"><a href="index.html#cb6-24" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(MASS)) devtools::install_version(&quot;MASS&quot;,version=&quot;7.3-51.6&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span>
<span id="cb6-25"><a href="index.html#cb6-25" aria-hidden="true" tabindex="-1"></a><span class="ex">R</span> <span class="at">-q</span> <span class="at">-e</span> <span class="st">&#39;if(!require(data.table)) devtools::install_version(&quot;data.table&quot;,version=&quot;1.13.0&quot;,repos = &quot;http://cran.us.r-project.org&quot;)&#39;</span></span></code></pre></div>
</div>
<div id="run-cellxgene-by-h5ad-file" class="section level3" number="1.2.4">
<h3><span class="header-section-number">1.2.4</span> Run cellxgene by h5ad file</h3>
<p>You can aslo run cellxgene by specifying a h5ad file, which stores scRNA-seq data along with a host and a port.
Use ‘ps’ to find used ports to spare. Please see <a href="https://chanzuckerberg.github.io/cellxgene/posts/launch" class="uri">https://chanzuckerberg.github.io/cellxgene/posts/launch</a> for details</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb7-1"><a href="index.html#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ps</span> <span class="at">-ef</span> <span class="kw">|</span> <span class="fu">grep</span> cellxgene</span>
<span id="cb7-2"><a href="index.html#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="at">-e</span> <span class="st">&#39;reticulate::py_config()&#39;</span></span>
<span id="cb7-3"><a href="index.html#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Run the following command if the output of the above command doesn&#39;t point to the Python in your env.</span></span>
<span id="cb7-4"><a href="index.html#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="bu">export</span> <span class="va">RETICULATE_PYTHON</span><span class="op">=</span><span class="kw">`</span><span class="fu">which</span> python<span class="kw">`</span></span>
<span id="cb7-5"><a href="index.html#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="ex">cellxgene</span> launch <span class="at">--host</span> <span class="op">&lt;</span>xxx<span class="op">&gt;</span> --port <span class="op">&lt;</span>xxx<span class="op">&gt;</span> --disable-annotations <span class="at">--verbose</span> <span class="op">&lt;</span>h5ad file<span class="op">&gt;</span></span></code></pre></div>
</div>
<div id="cellxgene-on-web-browser" class="section level3" number="1.2.5">
<h3><span class="header-section-number">1.2.5</span> Cellxgene on web browser</h3>
<p>chrome is preferred, version 87.0.4280.88 or 87.0.4280.141 is used. Users can access .
Following screenshot is what you should be able to see in console of chrome developer tools.
<img src="cellonweb.png" alt="Alt text" /></p>
</div>
</div>
<div id="authors" class="section level2" number="1.3">
<h2><span class="header-section-number">1.3</span> Authors</h2>
<ul>
<li>Keijie Li (<a href="mailto:kejie.li@biogen.com" class="email">kejie.li@biogen.com</a>), Associate Director at Biogen in the research department. Main author and content wrangler.</li>
<li>Zhengyu Ouyang, Associate Director of Bioingormatics at BioinfoRx. Content wrangler.</li>
<li>Baohong Zhang (<a href="mailto:baohong.zhang@biogen.com" class="email">baohong.zhang@biogen.com</a>), Head of Genome Informatics at Biogen in the research department. Corresponding author and content wrangler.</li>
</ul>
<p>The development and delivery of this material has also contributed by:</p>
<ul>
<li>Yirui Chen (<a href="mailto:yirui.chen@bigen.com" class="email">yirui.chen@bigen.com</a>), Biogen Inc.</li>
<li>Dongdong Lin (<a href="mailto:dongdong.lin@biogen.com" class="email">dongdong.lin@biogen.com</a>), Biogen Inc.</li>
<li>Michael Mingueneau (<a href="mailto:michael.mingueneau@bioigen.com" class="email">michael.mingueneau@bioigen.com</a>), Biogen Inc.</li>
<li>Jake Gagnon (<a href="mailto:jake.gagnon@biogen.com" class="email">jake.gagnon@biogen.com</a>), Biogen Inc.</li>
<li>Will Chen (<a href="mailto:wwchen@post.harvard.edu" class="email">wwchen@post.harvard.edu</a>), Biogen Inc.</li>
<li>David Sexton (<a href="mailto:david.sexton@biogen.com" class="email">david.sexton@biogen.com</a>), Biogen Inc.</li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="web-source.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
